<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Anemonefish Mutualism Analysis Results</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/quarto.js"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="index_files/libs/quarto-html/anchor.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="index_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="index_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="index_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Anemonefish Mutualism Analysis Results</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="drivers-of-host-sea-anemone-and-anemonefish-richness" class="level1">
<h1>Drivers of Host Sea Anemone and Anemonefish Richness</h1>
<p>This section replicates the initial data loading and processing steps from the desert ant paperâ€™s results section, adapted for the anemonefish-anemone mutualism. We load the current predicted suitability rasters for host sea anemones and anemonefish (environmental-only models), calculate community richness, and crop them to the Indo-Pacific study area.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Using pacman for streamlined package management</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"pacman"</span>)) <span class="fu">install.packages</span>(<span class="st">"pacman"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>pacman<span class="sc">::</span><span class="fu">p_load</span>(here, dplyr, terra, sf, stringr, ggplot2, readr, tools)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Load project configuration</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure your working directory is the root of your sdm_anemonefish project</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(<span class="st">"scripts/config.R"</span>)) {</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">source</span>(<span class="st">"scripts/config.R"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">file.exists</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"scripts/config.R"</span>))) {</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">source</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"scripts/config.R"</span>))</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"FATAL: Configuration file 'scripts/config.R' not found. Please set your working directory to the project root."</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Configuration loaded and bundled into 'config' list.
Base directory: /Users/chrisrauch/anemonefish_post_analysis 
Intermediate SDM Output Dir: /Users/chrisrauch/anemonefish_post_analysis/data/sdm_output_intermediate 
Intermediate Models Dir: /Users/chrisrauch/anemonefish_post_analysis/data/sdm_output_intermediate/models 
Intermediate Results Dir: /Users/chrisrauch/anemonefish_post_analysis/data/sdm_output_intermediate/results 
Target Prediction Base: /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions 
Target Results Base: /Users/chrisrauch/anemonefish_post_analysis/predictions 
Logging to file: /Users/chrisrauch/anemonefish_post_analysis/data/sdm_output_intermediate/logs_and_analysis/sdm_run_20250528_170555.log (Level: INFO Append: TRUE )
Logging to console: TRUE (Level: INFO )
Parallel execution: TRUE with 7 cores.
Using predictors: PCA </code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"config"</span>)) <span class="fu">stop</span>(<span class="st">"FATAL: 'config' list not found after sourcing config.R"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Helper function for constructing prediction filenames (already in your sdm_modeling_helpers.R)</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Make sure this helper is sourced or defined if not running the full pipeline before this</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">file.path</span>(config<span class="sc">$</span>helpers_dir, <span class="st">"sdm_modeling_helpers.R"</span>)) <span class="co"># If needed</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Define Parameters ---</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>current_scenario_name <span class="ot">&lt;-</span> <span class="st">"current"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the suffix based on config (assuming this is for env-only models)</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>predictor_suffix <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(config<span class="sc">$</span>use_pca_predictors, <span class="st">"_pca"</span>, <span class="st">"_vif"</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 1. Load Host Sea Anemone Data ---</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"--- Loading Host Sea Anemone Predictions (Current Scenario) ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Loading Host Sea Anemone Predictions (Current Scenario) ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(config<span class="sc">$</span>anemone_species_list_file)) {</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"Anemone species list CSV not found: "</span>, config<span class="sc">$</span>anemone_species_list_file)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>anemone_species_df <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(config<span class="sc">$</span>anemone_species_list_file, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>host_raster_files <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>host_short_names <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(anemone_species_df)) {</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  sp_name_sanitized <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">" "</span>, <span class="st">"_"</span>, anemone_species_df<span class="sc">$</span>scientificName[i])</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Use the construct_prediction_filename helper if it's available and loaded</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Otherwise, construct manually (ensure it matches your file naming)</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  pred_file_path <span class="ot">&lt;-</span> <span class="fu">construct_prediction_filename</span>(</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">species_name_sanitized =</span> sp_name_sanitized,</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">scenario_name =</span> current_scenario_name,</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">predictor_type_suffix =</span> predictor_suffix, <span class="co"># Suffix for anemone env-only models</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">config =</span> config</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">file.exists</span>(pred_file_path)) {</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>    host_raster_files <span class="ot">&lt;-</span> <span class="fu">c</span>(host_raster_files, pred_file_path)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>    host_short_names <span class="ot">&lt;-</span> <span class="fu">c</span>(host_short_names, sp_name_sanitized) <span class="co"># Or create 2-letter codes</span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Warning: Host prediction file not found for"</span>, sp_name_sanitized, <span class="st">"at"</span>, pred_file_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Warning: Host prediction file not found for Radianthus_aurora at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/mean_pred_Radianthus_aurora_pca.tif </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>host_pred_stack <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(host_raster_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  host_pred_stack <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">rast</span>(host_raster_files)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(host_pred_stack) <span class="ot">&lt;-</span> host_short_names</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Loaded"</span>, terra<span class="sc">::</span><span class="fu">nlyr</span>(host_pred_stack), <span class="st">"host anemone prediction rasters.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  host_richness_sum <span class="ot">&lt;-</span> <span class="fu">sum</span>(host_pred_stack, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(host_richness_sum) <span class="ot">&lt;-</span> <span class="st">"HostAnemoneRichness"</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Calculated host anemone summed richness.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Optional: Plot individual host rasters and summed richness before cropping</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># plot(host_pred_stack, main = "Individual Host Anemone Suitability (Current)")</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># plot(host_richness_sum, main = "Summed Host Anemone Richness (Current - Uncropped)")</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Error: No host anemone prediction rasters found. Cannot proceed with host richness.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  host_richness_sum <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="co"># Ensure it's NULL if no files found</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded 9 host anemone prediction rasters.

|---------|---------|---------|---------|
=========================================
                                          
Calculated host anemone summed richness.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 2. Load Anemonefish (Environmental-Only Models) Data ---</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Loading Anemonefish (Env-Only) Predictions (Current Scenario) ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Loading Anemonefish (Env-Only) Predictions (Current Scenario) ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(config<span class="sc">$</span>anemonefish_species_list_file)) {</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"Anemonefish species list CSV not found: "</span>, config<span class="sc">$</span>anemonefish_species_list_file)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>anemonefish_species_df <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(config<span class="sc">$</span>anemonefish_species_list_file, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>fish_raster_files <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>fish_short_names <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(anemonefish_species_df)) {</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  sp_name_sanitized <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">" "</span>, <span class="st">"_"</span>, anemonefish_species_df<span class="sc">$</span>scientificName[i])</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Suffix for anemonefish env-only models (from 06b)</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  pred_file_path <span class="ot">&lt;-</span> <span class="fu">construct_prediction_filename</span>(</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">species_name_sanitized =</span> sp_name_sanitized,</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">scenario_name =</span> current_scenario_name,</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">predictor_type_suffix =</span> predictor_suffix, <span class="co"># Same suffix if using same method (PCA/VIF)</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">config =</span> config</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">file.exists</span>(pred_file_path)) {</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>    fish_raster_files <span class="ot">&lt;-</span> <span class="fu">c</span>(fish_raster_files, pred_file_path)</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>    fish_short_names <span class="ot">&lt;-</span> <span class="fu">c</span>(fish_short_names, sp_name_sanitized)</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Warning: Anemonefish (env-only) prediction file not found for"</span>, sp_name_sanitized, <span class="st">"at"</span>, pred_file_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Warning: Anemonefish (env-only) prediction file not found for Amphiprion_akindynos at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/mean_pred_Amphiprion_akindynos_pca.tif 
Warning: Anemonefish (env-only) prediction file not found for Amphiprion_allardi at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/mean_pred_Amphiprion_allardi_pca.tif 
Warning: Anemonefish (env-only) prediction file not found for Amphiprion_bicinctus at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/mean_pred_Amphiprion_bicinctus_pca.tif 
Warning: Anemonefish (env-only) prediction file not found for Amphiprion_chrysopterus at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/mean_pred_Amphiprion_chrysopterus_pca.tif 
Warning: Anemonefish (env-only) prediction file not found for Amphiprion_melanopus at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/mean_pred_Amphiprion_melanopus_pca.tif 
Warning: Anemonefish (env-only) prediction file not found for Amphiprion_percula at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/mean_pred_Amphiprion_percula_pca.tif 
Warning: Anemonefish (env-only) prediction file not found for Premnas_biaculeatus at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/mean_pred_Premnas_biaculeatus_pca.tif </code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>fish_pred_stack <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(fish_raster_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  fish_pred_stack <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">rast</span>(fish_raster_files)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(fish_pred_stack) <span class="ot">&lt;-</span> fish_short_names</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Loaded"</span>, terra<span class="sc">::</span><span class="fu">nlyr</span>(fish_pred_stack), <span class="st">"anemonefish (env-only) prediction rasters.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  fish_richness_sum <span class="ot">&lt;-</span> <span class="fu">sum</span>(fish_pred_stack, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(fish_richness_sum) <span class="ot">&lt;-</span> <span class="st">"AnemonefishRichness_EnvOnly"</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Calculated anemonefish (env-only) summed richness.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Optional: Plot individual fish rasters and summed richness before cropping</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># plot(fish_pred_stack, main = "Individual Anemonefish Suitability (Current, Env-Only)")</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># plot(fish_richness_sum, main = "Summed Anemonefish Richness (Current, Env-Only - Uncropped)")</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Error: No anemonefish (env-only) prediction rasters found. Cannot proceed with fish richness.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>  fish_richness_sum <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="co"># Ensure it's NULL</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded 5 anemonefish (env-only) prediction rasters.

|---------|---------|---------|---------|
=========================================
                                          
Calculated anemonefish (env-only) summed richness.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 3. Crop to Indo-Pacific Extent (if rasters were successfully loaded) ---</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Cropping Rasters to Indo-Pacific Extent ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Cropping Rasters to Indo-Pacific Extent ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (config<span class="sc">$</span>apply_indo_pacific_crop) {</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  ip_extent <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">ext</span>(config<span class="sc">$</span>indo_pacific_bbox)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Using Indo-Pacific Bounding Box for cropping:"</span>, </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>      <span class="fu">paste</span>(config<span class="sc">$</span>indo_pacific_bbox, <span class="at">collapse=</span><span class="st">", "</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(host_pred_stack)) {</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    host_pred_stack_cropped <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>      terra<span class="sc">::</span><span class="fu">crop</span>(host_pred_stack, ip_extent)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Warning: Failed to crop host_pred_stack:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); host_pred_stack</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> { host_pred_stack_cropped <span class="ot">&lt;-</span> <span class="cn">NULL</span> }</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(host_richness_sum)) {</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>    host_richness_sum_cropped <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>      terra<span class="sc">::</span><span class="fu">crop</span>(host_richness_sum, ip_extent)</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Warning: Failed to crop host_richness_sum:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); host_richness_sum</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> { host_richness_sum_cropped <span class="ot">&lt;-</span> <span class="cn">NULL</span> }</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(fish_pred_stack)) {</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>    fish_pred_stack_cropped <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>      terra<span class="sc">::</span><span class="fu">crop</span>(fish_pred_stack, ip_extent)</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Warning: Failed to crop fish_pred_stack:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); fish_pred_stack</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> { fish_pred_stack_cropped <span class="ot">&lt;-</span> <span class="cn">NULL</span> }</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(fish_richness_sum)) {</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>    fish_richness_sum_cropped <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a>      terra<span class="sc">::</span><span class="fu">crop</span>(fish_richness_sum, ip_extent)</span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Warning: Failed to crop fish_richness_sum:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); fish_richness_sum</span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> { fish_richness_sum_cropped <span class="ot">&lt;-</span> <span class="cn">NULL</span> }</span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Cropping complete.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Skipping Indo-Pacific cropping based on config.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a>  host_pred_stack_cropped <span class="ot">&lt;-</span> host_pred_stack</span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a>  host_richness_sum_cropped <span class="ot">&lt;-</span> host_richness_sum</span>
<span id="cb18-43"><a href="#cb18-43" aria-hidden="true" tabindex="-1"></a>  fish_pred_stack_cropped <span class="ot">&lt;-</span> fish_pred_stack</span>
<span id="cb18-44"><a href="#cb18-44" aria-hidden="true" tabindex="-1"></a>  fish_richness_sum_cropped <span class="ot">&lt;-</span> fish_richness_sum</span>
<span id="cb18-45"><a href="#cb18-45" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Using Indo-Pacific Bounding Box for cropping: 30, 180, -50, 50 

|---------|---------|---------|---------|
=========================================
                                          
Cropping complete.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 4. Plot Cropped Richness Maps ---</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Plotting Cropped Richness Maps ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Plotting Cropped Richness Maps ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(host_richness_sum_cropped)) {</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(host_richness_sum_cropped, <span class="at">main =</span> <span class="st">"Summed Host Anemone Richness (Current - Cropped)"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-and-process-rasters-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(fish_richness_sum_cropped)) {</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(fish_richness_sum_cropped, <span class="at">main =</span> <span class="st">"Summed Anemonefish Richness (Current, Env-Only - Cropped)"</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-and-process-rasters-2.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Optional: Plot cropped individual stacks if needed for visual check</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># if (!is.null(host_pred_stack_cropped)) {</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   plot(host_pred_stack_cropped, main = "Individual Host Anemone Suitability (Cropped)")</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co"># if (!is.null(fish_pred_stack_cropped)) {</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   plot(fish_pred_stack_cropped, main = "Individual Anemonefish (Env-Only) Suitability (Cropped)")</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- First section of results processing finished. ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- First section of results processing finished. ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The objects host_pred_stack_cropped, host_richness_sum_cropped, </span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># fish_pred_stack_cropped, and fish_richness_sum_cropped are now available.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="evaluate-models" class="level2">
<h2 class="anchored" data-anchor-id="evaluate-models">Evaluate models</h2>
<p>This section loads the cross-validation (CV) results from the various SDM runs for host sea anemones and different types of anemonefish models (environmental-only, biotic-only, and combined environmental + biotic), focusing on AUC.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr) <span class="co"># For map_df</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr) <span class="co"># For data manipulation</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr) <span class="co"># For read_csv</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 1. Load Host Sea Anemone CV Results ("Plants" equivalent) ---</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine suffix based on config (assuming this matches what was used for 06a)</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>host_predictor_suffix <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(config<span class="sc">$</span>use_pca_predictors, <span class="st">"_pca"</span>, <span class="st">"_vif"</span>)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>host_cv_results_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(config<span class="sc">$</span>target_results_base, <span class="fu">paste0</span>(<span class="st">"anemone"</span>, host_predictor_suffix))</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Loading Host Sea Anemone CV results from:"</span>, host_cv_results_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading Host Sea Anemone CV results from: /Users/chrisrauch/anemonefish_post_analysis/predictions/anemone_pca </code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>host_cv_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(host_cv_results_dir,</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">pattern =</span> <span class="st">"^CV_Results_.*</span><span class="sc">\\</span><span class="st">.csv$"</span>, <span class="co"># Files start with CV_Results_</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(host_cv_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  host_cv_data <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_df</span>(host_cv_files, </span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>                <span class="sc">~</span>readr<span class="sc">::</span><span class="fu">read_csv</span>(.x, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">mutate</span>(<span class="at">filename_full =</span> .x, </span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">filename =</span> <span class="fu">basename</span>(.x)))</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>  host_model_evals <span class="ot">&lt;-</span> host_cv_data <span class="sc">%&gt;%</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(filename) <span class="sc">%&gt;%</span></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summarise</span>(</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">mean_AUC_test_CV =</span> <span class="fu">mean</span>(AUC_test_CV, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="co"># SDMtune output column is AUC_test_CV</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">sd_AUC_test_CV =</span> <span class="fu">sd</span>(AUC_test_CV, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>      <span class="co"># mean_test_tss = mean(test_TSS, na.rm = TRUE), # SDMtune output column is test_TSS</span></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>      <span class="co"># sd_test_tss = sd(test_TSS, na.rm = TRUE)</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">where</span>(is.numeric), <span class="sc">~</span><span class="fu">round</span>(., <span class="dv">3</span>))) </span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">First few rows of Host Sea Anemone Model Evaluations (AUC only):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(host_model_evals))</span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>  host_cv_summary_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(config<span class="sc">$</span>base_dir, <span class="st">"outputs_for_analysis"</span>, <span class="st">"host_anemone_CV_summary_auc_only.csv"</span>)</span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">dirname</span>(host_cv_summary_path), <span class="at">recursive =</span> <span class="cn">TRUE</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(host_model_evals, host_cv_summary_path, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Host Sea Anemone CV summary (AUC only) saved to:"</span>, host_cv_summary_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb29-28"><a href="#cb29-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb29-29"><a href="#cb29-29" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb29-30"><a href="#cb29-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: No CV_Results CSV files found for host sea anemones in"</span>, host_cv_results_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb29-31"><a href="#cb29-31" aria-hidden="true" tabindex="-1"></a>  host_cv_data <span class="ot">&lt;-</span> <span class="cn">NULL</span> </span>
<span id="cb29-32"><a href="#cb29-32" aria-hidden="true" tabindex="-1"></a>  host_model_evals <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb29-33"><a href="#cb29-33" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
First few rows of Host Sea Anemone Model Evaluations (AUC only):
# A tibble: 6 Ã— 3
  filename                              mean_AUC_test_CV sd_AUC_test_CV
  &lt;chr&gt;                                            &lt;dbl&gt;          &lt;dbl&gt;
1 CV_Results_Entacmaea_quadricolor.csv             0.76           0.019
2 CV_Results_Macrodactyla_doreensis.csv            0.771          0.024
3 CV_Results_Radianthus_aurora.csv                 0.76           0.017
4 CV_Results_Radianthus_crispa.csv                 0.836          0.013
5 CV_Results_Radianthus_magnifica.csv              0.777          0.022
6 CV_Results_Radianthus_malu.csv                 Inf            NaN    

Host Sea Anemone CV summary (AUC only) saved to: /Users/chrisrauch/anemonefish_post_analysis/outputs_for_analysis/host_anemone_CV_summary_auc_only.csv </code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 2. Load Anemonefish Environmental-Only CV Results ("Ants Climate-Only" equivalent) ---</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>fish_env_predictor_suffix <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(config<span class="sc">$</span>use_pca_predictors, <span class="st">"_pca"</span>, <span class="st">"_vif"</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>fish_env_cv_results_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(config<span class="sc">$</span>target_results_base, <span class="fu">paste0</span>(<span class="st">"anemonefish"</span>, fish_env_predictor_suffix))</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Loading Anemonefish (Environmental-Only) CV results from:"</span>, fish_env_cv_results_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Loading Anemonefish (Environmental-Only) CV results from: /Users/chrisrauch/anemonefish_post_analysis/predictions/anemonefish_pca </code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>fish_env_cv_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(fish_env_cv_results_dir,</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">pattern =</span> <span class="st">"^CV_Results_.*</span><span class="sc">\\</span><span class="st">.csv$"</span>,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(fish_env_cv_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  fish_env_cv_data <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_df</span>(fish_env_cv_files, </span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>                  <span class="sc">~</span>readr<span class="sc">::</span><span class="fu">read_csv</span>(.x, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>                     <span class="fu">mutate</span>(<span class="at">filename_full =</span> .x,</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>                            <span class="at">filename =</span> <span class="fu">basename</span>(.x)))</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Loaded"</span>, <span class="fu">nrow</span>(fish_env_cv_data), <span class="st">"rows from"</span>, <span class="fu">length</span>(fish_env_cv_files), <span class="st">"anemonefish (env-only) CV files.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(fish_env_cv_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(filename, AUC_test_CV) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(filename) <span class="sc">%&gt;%</span> <span class="fu">slice_head</span>(<span class="at">n=</span><span class="dv">1</span>), <span class="at">n=</span><span class="dv">3</span>)) <span class="co"># Show a sample (test_TSS removed for now)</span></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: No CV_Results CSV files found for anemonefish (env-only) in"</span>, fish_env_cv_results_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>  fish_env_cv_data <span class="ot">&lt;-</span> <span class="cn">NULL</span> </span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded 213 rows from 12 anemonefish (env-only) CV files.
# A tibble: 3 Ã— 2
# Groups:   filename [3]
  filename                            AUC_test_CV
  &lt;chr&gt;                                     &lt;dbl&gt;
1 CV_Results_Amphiprion_akindynos.csv       0.880
2 CV_Results_Amphiprion_allardi.csv         0.923
3 CV_Results_Amphiprion_bicinctus.csv       0.862</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 3. Load Anemonefish Biotic Model CV Results (Equivalent to "Ant Contrasts") ---</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   3a. Biotic-Only CV Results (from 06c)</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>fish_biotic_only_suffix_config <span class="ot">&lt;-</span> <span class="st">"_biotic_only"</span> </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>fish_biotic_only_cv_results_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(config<span class="sc">$</span>target_results_base, <span class="fu">paste0</span>(<span class="st">"anemonefish"</span>, fish_biotic_only_suffix_config))</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Loading Anemonefish (Biotic-Only) CV results from:"</span>, fish_biotic_only_cv_results_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Loading Anemonefish (Biotic-Only) CV results from: /Users/chrisrauch/anemonefish_post_analysis/predictions/anemonefish_biotic_only </code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>fish_biotic_only_cv_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(fish_biotic_only_cv_results_dir,</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">pattern =</span> <span class="st">"^CV_Results_.*</span><span class="sc">\\</span><span class="st">.csv$"</span>,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(fish_biotic_only_cv_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  anemonefish_biotic_cv_data <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_df</span>(fish_biotic_only_cv_files, </span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>                <span class="sc">~</span>readr<span class="sc">::</span><span class="fu">read_csv</span>(.x, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">mutate</span>(<span class="at">filename_full =</span> .x,</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">filename =</span> <span class="fu">basename</span>(.x)))</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Loaded"</span>, <span class="fu">nrow</span>(anemonefish_biotic_cv_data), <span class="st">"rows from"</span>, <span class="fu">length</span>(fish_biotic_only_cv_files), <span class="st">"anemonefish (biotic-only) CV files.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(anemonefish_biotic_cv_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(filename, AUC_test_CV) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(filename) <span class="sc">%&gt;%</span> <span class="fu">slice_head</span>(<span class="at">n=</span><span class="dv">1</span>), <span class="at">n=</span><span class="dv">3</span>)) <span class="co"># test_TSS removed for now</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: No CV_Results CSV files found for anemonefish (biotic-only) in"</span>, fish_biotic_only_cv_results_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>  anemonefish_biotic_cv_data <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded 236 rows from 12 anemonefish (biotic-only) CV files.
# A tibble: 3 Ã— 2
# Groups:   filename [3]
  filename                            AUC_test_CV
  &lt;chr&gt;                                     &lt;dbl&gt;
1 CV_Results_Amphiprion_akindynos.csv       0.809
2 CV_Results_Amphiprion_allardi.csv         0.958
3 CV_Results_Amphiprion_bicinctus.csv       0.847</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#   3b. Combined Env+Biotic CV Results (from 06d)</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>fish_combined_suffix_config <span class="ot">&lt;-</span> <span class="st">"_combined_pca"</span> </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>fish_combined_cv_results_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(config<span class="sc">$</span>target_results_base, <span class="fu">paste0</span>(<span class="st">"anemonefish"</span>, fish_combined_suffix_config))</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Loading Anemonefish (Combined Env+Biotic) CV results from:"</span>, fish_combined_cv_results_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Loading Anemonefish (Combined Env+Biotic) CV results from: /Users/chrisrauch/anemonefish_post_analysis/predictions/anemonefish_combined_pca </code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>fish_combined_cv_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(fish_combined_cv_results_dir,</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">pattern =</span> <span class="st">"^CV_Results_.*</span><span class="sc">\\</span><span class="st">.csv$"</span>,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(fish_combined_cv_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  anemonefish_combined_cv_data <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_df</span>(fish_combined_cv_files, </span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>                <span class="sc">~</span>readr<span class="sc">::</span><span class="fu">read_csv</span>(.x, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">mutate</span>(<span class="at">filename_full =</span> .x,</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">filename =</span> <span class="fu">basename</span>(.x)))</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Loaded"</span>, <span class="fu">nrow</span>(anemonefish_combined_cv_data), <span class="st">"rows from"</span>, <span class="fu">length</span>(fish_combined_cv_files), <span class="st">"anemonefish (combined) CV files.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(anemonefish_combined_cv_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(filename, AUC_test_CV) <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(filename) <span class="sc">%&gt;%</span> <span class="fu">slice_head</span>(<span class="at">n=</span><span class="dv">1</span>), <span class="at">n=</span><span class="dv">3</span>)) <span class="co"># test_TSS removed for now</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: No CV_Results CSV files found for anemonefish (combined) in"</span>, fish_combined_cv_results_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>  anemonefish_combined_cv_data <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded 217 rows from 12 anemonefish (combined) CV files.
# A tibble: 3 Ã— 2
# Groups:   filename [3]
  filename                            AUC_test_CV
  &lt;chr&gt;                                     &lt;dbl&gt;
1 CV_Results_Amphiprion_akindynos.csv       0.892
2 CV_Results_Amphiprion_allardi.csv         0.957
3 CV_Results_Amphiprion_bicinctus.csv       0.879</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The data frames:</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co"># - host_cv_data (and host_model_evals)</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="co"># - fish_env_cv_data</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="co"># - anemonefish_biotic_cv_data</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="co"># - anemonefish_combined_cv_data</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="co"># are now available for the next steps.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="collect-and-prepare-anemonefish-model-evaluation-data" class="level2">
<h2 class="anchored" data-anchor-id="collect-and-prepare-anemonefish-model-evaluation-data">Collect and Prepare Anemonefish Model Evaluation Data</h2>
<p>This section gathers the CV results from the different anemonefish model types (environmental-only, biotic-only, and combined host+environment), cleans the species names, and standardizes a â€˜model_typeâ€™ column for comparison.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr) <span class="co"># For separate</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr) <span class="co"># For string manipulation</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 1. Prepare Anemonefish Environmental-Only Data ---</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(fish_env_cv_data) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(fish_env_cv_data) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  fish_env_processed <span class="ot">&lt;-</span> fish_env_cv_data <span class="sc">%&gt;%</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>      <span class="co"># # Extract species name: Remove "CV_Results_" prefix and ".csv" suffix</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>      <span class="co"># species = str_remove(filename, "^CV_Results_"),</span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>      <span class="co"># species = str_remove(species, "\\.csv$"),</span></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>      <span class="co"># # Add model type</span></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">species =</span> SpeciesName,</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">model_type =</span> <span class="st">"env_only"</span> </span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(species, model_type, AUC_test_CV, filename_full) <span class="co"># Keep relevant columns, filename_full for debugging</span></span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Processed anemonefish environmental-only CV data.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: fish_env_cv_data is empty or NULL. Cannot process environmental-only models.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a>  fish_env_processed <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Processed anemonefish environmental-only CV data.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 2. Prepare Anemonefish Biotic-Only Data ---</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(anemonefish_biotic_cv_data) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(anemonefish_biotic_cv_data) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  fish_biotic_processed <span class="ot">&lt;-</span> anemonefish_biotic_cv_data <span class="sc">%&gt;%</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Extract species name: Remove "CV_Results_" prefix and the known suffix "_biotic_only.csv" (or _biotic_pc4 if that was kept)</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Assuming filenames are like: CV_Results_Amphiprion_clarkii_biotic_only.csv or CV_Results_Amphiprion_clarkii_biotic_pc4.csv</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Adjust the str_remove pattern if your biotic_only filenames have a different consistent suffix</span></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>      <span class="co"># species = str_remove(filename, "^CV_Results_"),</span></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>      <span class="co"># species = str_remove(species, paste0(config$model_output_subdir_map[["_biotic_only"]] %||% "_biotic_only", "\\.csv$")), # Use suffix from map or default</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>      <span class="co"># species = str_remove(species, paste0(config$model_output_subdir_map[["_biotic_pc4"]] %||% "_biotic_pc4", "\\.csv$")), # If _biotic_pc4 was used</span></span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">species =</span> SpeciesName,</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">model_type =</span> <span class="st">"biotic_only"</span></span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(species, model_type, AUC_test_CV, filename_full)</span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Processed anemonefish biotic-only CV data.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb46-17"><a href="#cb46-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: anemonefish_biotic_cv_data is empty or NULL. Cannot process biotic-only models.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb46-18"><a href="#cb46-18" aria-hidden="true" tabindex="-1"></a>  fish_biotic_processed <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb46-19"><a href="#cb46-19" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Processed anemonefish biotic-only CV data.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 3. Prepare Anemonefish Combined (Host + Env) Data ---</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(anemonefish_combined_cv_data) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(anemonefish_combined_cv_data) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  fish_combined_processed <span class="ot">&lt;-</span> anemonefish_combined_cv_data <span class="sc">%&gt;%</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>      <span class="co"># # Extract species name: Remove "CV_Results_" and the known suffix "_combined_pca.csv"</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>      <span class="co"># species = str_remove(filename, "^CV_Results_"),</span></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>      <span class="co"># species = str_remove(species, paste0(config$model_output_subdir_map[["_combined_pca"]] %||% "_combined_pca", "\\.csv$")), # Use suffix from map or default</span></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">species =</span> SpeciesName,</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">model_type =</span> <span class="st">"combined_host_env"</span></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(species, model_type, AUC_test_CV, filename_full)</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Processed anemonefish combined (host+env) CV data.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: anemonefish_combined_cv_data is empty or NULL. Cannot process combined models.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>  fish_combined_processed <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Processed anemonefish combined (host+env) CV data.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 4. Combine all processed anemonefish model data ---</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Use bind_rows which handles NULL data frames gracefully</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>df_fish_comparison <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  fish_env_processed,</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>  fish_biotic_processed,</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>  fish_combined_processed</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(df_fish_comparison) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(df_fish_comparison) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Optional: Filter out species that might have been excluded or problematic in the ant paper's example</span></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># For now, we'll keep all successfully processed species.</span></span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># df_fish_comparison &lt;- df_fish_comparison %&gt;%</span></span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   filter(!species %in% c("Problem_Species1", "Problem_Species2"))</span></span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Combined data frame 'df_fish_comparison' created with"</span>, <span class="fu">nrow</span>(df_fish_comparison), <span class="st">"rows.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Summary of model types and counts:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb50-17"><a href="#cb50-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">table</span>(df_fish_comparison<span class="sc">$</span>model_type))</span>
<span id="cb50-18"><a href="#cb50-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">First few rows of combined data:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb50-19"><a href="#cb50-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(df_fish_comparison))</span>
<span id="cb50-20"><a href="#cb50-20" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb50-21"><a href="#cb50-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Error: 'df_fish_comparison' is empty. No anemonefish CV data was successfully processed.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb50-22"><a href="#cb50-22" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Combined data frame 'df_fish_comparison' created with 666 rows.
Summary of model types and counts:

      biotic_only combined_host_env          env_only 
              236               217               213 

First few rows of combined data:
# A tibble: 6 Ã— 4
  species              model_type AUC_test_CV filename_full                     
  &lt;chr&gt;                &lt;chr&gt;            &lt;dbl&gt; &lt;chr&gt;                             
1 Amphiprion_akindynos env_only         0.880 /Users/chrisrauch/anemonefish_posâ€¦
2 Amphiprion_akindynos env_only         0.878 /Users/chrisrauch/anemonefish_posâ€¦
3 Amphiprion_akindynos env_only         0.910 /Users/chrisrauch/anemonefish_posâ€¦
4 Amphiprion_akindynos env_only         0.877 /Users/chrisrauch/anemonefish_posâ€¦
5 Amphiprion_akindynos env_only         0.877 /Users/chrisrauch/anemonefish_posâ€¦
6 Amphiprion_akindynos env_only         0.877 /Users/chrisrauch/anemonefish_posâ€¦</code></pre>
</div>
</div>
</section>
<section id="compare-anemonefish-model-performance" class="level2">
<h2 class="anchored" data-anchor-id="compare-anemonefish-model-performance">Compare anemonefish model performance</h2>
<p>This section statistically compares the performance (test AUC) of the different anemonefish model types (environmental-only, biotic-only, combined host+environment) using a linear mixed model. It also visualizes these comparisons.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glmmTMB)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co"># library(performance) # Optional, for model_performance() if needed later</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(df_fish_comparison) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(df_fish_comparison) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> <span class="st">"AUC_test_CV"</span> <span class="sc">%in%</span> <span class="fu">names</span>(df_fish_comparison)) {</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure model_type is a factor for the GLMM</span></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>  df_fish_comparison<span class="sc">$</span>model_type <span class="ot">&lt;-</span> <span class="fu">factor</span>(df_fish_comparison<span class="sc">$</span>model_type, </span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"env_only"</span>, <span class="st">"biotic_only"</span>, <span class="st">"combined_host_env"</span>))</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Statistical Comparison using GLMM ---</span></span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Fitting GLMM to compare AUC_test_CV across model types ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Model: AUC_test_CV ~ model_type + (1|species)</span></span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># This tests if model_type has a significant effect on AUC, </span></span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># accounting for random variation between species.</span></span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 'env_only' will be the reference level by default due to factor level ordering.</span></span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a>  m1_fish_auc <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">glmmTMB</span>(AUC_test_CV <span class="sc">~</span> model_type <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>species), <span class="at">data =</span> df_fish_comparison)</span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a>  }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb52-20"><a href="#cb52-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Error fitting GLMM for AUC:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb52-21"><a href="#cb52-21" aria-hidden="true" tabindex="-1"></a>    <span class="cn">NULL</span></span>
<span id="cb52-22"><a href="#cb52-22" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb52-23"><a href="#cb52-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-24"><a href="#cb52-24" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(m1_fish_auc)) {</span>
<span id="cb52-25"><a href="#cb52-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Summary of GLMM for AUC_test_CV:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb52-26"><a href="#cb52-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">summary</span>(m1_fish_auc))</span>
<span id="cb52-27"><a href="#cb52-27" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-28"><a href="#cb52-28" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Optional: Check model performance diagnostics if library(performance) is loaded</span></span>
<span id="cb52-29"><a href="#cb52-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># cat("\nGLMM Performance Metrics (AUC model):\n")</span></span>
<span id="cb52-30"><a href="#cb52-30" aria-hidden="true" tabindex="-1"></a>    <span class="co"># print(performance::model_performance(m1_fish_auc))</span></span>
<span id="cb52-31"><a href="#cb52-31" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-32"><a href="#cb52-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Optional: Plot residuals</span></span>
<span id="cb52-33"><a href="#cb52-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot(resid(m1_fish_auc), main = "Residuals of GLMM for AUC")</span></span>
<span id="cb52-34"><a href="#cb52-34" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb52-35"><a href="#cb52-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-36"><a href="#cb52-36" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Visualization ---</span></span>
<span id="cb52-37"><a href="#cb52-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Generating boxplot of AUC_test_CV by model type ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb52-38"><a href="#cb52-38" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-39"><a href="#cb52-39" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Basic boxplot comparing model types</span></span>
<span id="cb52-40"><a href="#cb52-40" aria-hidden="true" tabindex="-1"></a>  plot_auc_comparison <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_fish_comparison, <span class="fu">aes</span>(<span class="at">x =</span> model_type, <span class="at">y =</span> AUC_test_CV, <span class="at">fill =</span> model_type)) <span class="sc">+</span></span>
<span id="cb52-41"><a href="#cb52-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb52-42"><a href="#cb52-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_summary</span>(<span class="at">fun =</span> mean, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">18</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">color =</span> <span class="st">"black"</span>,</span>
<span id="cb52-43"><a href="#cb52-43" aria-hidden="true" tabindex="-1"></a>                 <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.75</span>)) <span class="sc">+</span> <span class="co"># Add mean points</span></span>
<span id="cb52-44"><a href="#cb52-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb52-45"><a href="#cb52-45" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"Anemonefish Model Performance Comparison"</span>,</span>
<span id="cb52-46"><a href="#cb52-46" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"Model Type"</span>,</span>
<span id="cb52-47"><a href="#cb52-47" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Test AUC (Predictive Accuracy)"</span></span>
<span id="cb52-48"><a href="#cb52-48" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb52-49"><a href="#cb52-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb52-50"><a href="#cb52-50" aria-hidden="true" tabindex="-1"></a>      <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"env_only"</span> <span class="ot">=</span> <span class="st">"grey70"</span>, <span class="st">"biotic_only"</span> <span class="ot">=</span> <span class="st">"skyblue"</span>, <span class="st">"combined_host_env"</span> <span class="ot">=</span> <span class="st">"salmon"</span>),</span>
<span id="cb52-51"><a href="#cb52-51" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> <span class="st">"Model Type"</span>,</span>
<span id="cb52-52"><a href="#cb52-52" aria-hidden="true" tabindex="-1"></a>      <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Environmental Only"</span>, <span class="st">"Biotic (Host) Only"</span>, <span class="st">"Combined (Host+Env)"</span>)</span>
<span id="cb52-53"><a href="#cb52-53" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb52-54"><a href="#cb52-54" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb52-55"><a href="#cb52-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb52-56"><a href="#cb52-56" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb52-57"><a href="#cb52-57" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.minor.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb52-58"><a href="#cb52-58" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.position =</span> <span class="st">"bottom"</span>,</span>
<span id="cb52-59"><a href="#cb52-59" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>) <span class="co"># Improve x-axis label readability</span></span>
<span id="cb52-60"><a href="#cb52-60" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb52-61"><a href="#cb52-61" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-62"><a href="#cb52-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(plot_auc_comparison)</span>
<span id="cb52-63"><a href="#cb52-63" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-64"><a href="#cb52-64" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Detailed boxplot per species, similar to ant paper (Fig 2 in ant PDF)</span></span>
<span id="cb52-65"><a href="#cb52-65" aria-hidden="true" tabindex="-1"></a>  <span class="co"># We need to make species names more readable for the plot</span></span>
<span id="cb52-66"><a href="#cb52-66" aria-hidden="true" tabindex="-1"></a>  df_fish_comparison_plot <span class="ot">&lt;-</span> df_fish_comparison <span class="sc">%&gt;%</span></span>
<span id="cb52-67"><a href="#cb52-67" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">species_display =</span> <span class="fu">str_replace</span>(species, <span class="st">"_"</span>, <span class="st">" "</span>)) <span class="co"># Replace underscore with space for display</span></span>
<span id="cb52-68"><a href="#cb52-68" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-69"><a href="#cb52-69" aria-hidden="true" tabindex="-1"></a>  plot_auc_per_species <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_fish_comparison_plot, <span class="fu">aes</span>(<span class="at">x =</span> species_display, <span class="at">y =</span> AUC_test_CV, <span class="at">fill =</span> model_type)) <span class="sc">+</span></span>
<span id="cb52-70"><a href="#cb52-70" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">preserve =</span> <span class="st">"single"</span>)) <span class="sc">+</span> <span class="co"># preserve = "single" helps alignment</span></span>
<span id="cb52-71"><a href="#cb52-71" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_summary</span>(<span class="at">fun =</span> mean, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">23</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">fill =</span> <span class="st">"red"</span>, <span class="at">color =</span> <span class="st">"black"</span>,</span>
<span id="cb52-72"><a href="#cb52-72" aria-hidden="true" tabindex="-1"></a>                 <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.75</span>, <span class="at">preserve =</span> <span class="st">"single"</span>), </span>
<span id="cb52-73"><a href="#cb52-73" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">aes</span>(<span class="at">group =</span> model_type)) <span class="sc">+</span> <span class="co"># Add mean points as red diamonds</span></span>
<span id="cb52-74"><a href="#cb52-74" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb52-75"><a href="#cb52-75" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"Anemonefish Model Performance by Species and Type"</span>,</span>
<span id="cb52-76"><a href="#cb52-76" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"Anemonefish Species"</span>,</span>
<span id="cb52-77"><a href="#cb52-77" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Test AUC (Predictive Accuracy)"</span></span>
<span id="cb52-78"><a href="#cb52-78" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb52-79"><a href="#cb52-79" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(</span>
<span id="cb52-80"><a href="#cb52-80" aria-hidden="true" tabindex="-1"></a>      <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"env_only"</span> <span class="ot">=</span> <span class="st">"grey70"</span>, <span class="st">"biotic_only"</span> <span class="ot">=</span> <span class="st">"skyblue"</span>, <span class="st">"combined_host_env"</span> <span class="ot">=</span> <span class="st">"salmon"</span>),</span>
<span id="cb52-81"><a href="#cb52-81" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> <span class="st">"Model Type:"</span>,</span>
<span id="cb52-82"><a href="#cb52-82" aria-hidden="true" tabindex="-1"></a>      <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Environmental Only"</span>, <span class="st">"Biotic (Host) Only"</span>, <span class="st">"Combined (Host+Env)"</span>)</span>
<span id="cb52-83"><a href="#cb52-83" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb52-84"><a href="#cb52-84" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb52-85"><a href="#cb52-85" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb52-86"><a href="#cb52-86" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.major.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb52-87"><a href="#cb52-87" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb52-88"><a href="#cb52-88" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.position =</span> <span class="st">"top"</span>,</span>
<span id="cb52-89"><a href="#cb52-89" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">60</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">8</span>), <span class="co"># Rotate and adjust x-axis labels</span></span>
<span id="cb52-90"><a href="#cb52-90" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb52-91"><a href="#cb52-91" aria-hidden="true" tabindex="-1"></a>      <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)</span>
<span id="cb52-92"><a href="#cb52-92" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb52-93"><a href="#cb52-93" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">1</span>) <span class="co"># Ensure AUC scale is from 0 to 1</span></span>
<span id="cb52-94"><a href="#cb52-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-95"><a href="#cb52-95" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(plot_auc_per_species)</span>
<span id="cb52-96"><a href="#cb52-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-97"><a href="#cb52-97" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print mean AUC per species per model type</span></span>
<span id="cb52-98"><a href="#cb52-98" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Mean AUC_test_CV per species and model type:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb52-99"><a href="#cb52-99" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(</span>
<span id="cb52-100"><a href="#cb52-100" aria-hidden="true" tabindex="-1"></a>    df_fish_comparison <span class="sc">%&gt;%</span></span>
<span id="cb52-101"><a href="#cb52-101" aria-hidden="true" tabindex="-1"></a>      <span class="fu">group_by</span>(species, model_type) <span class="sc">%&gt;%</span></span>
<span id="cb52-102"><a href="#cb52-102" aria-hidden="true" tabindex="-1"></a>      dplyr<span class="sc">::</span><span class="fu">summarize</span>(<span class="at">mean_auc =</span> <span class="fu">mean</span>(AUC_test_CV, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">.groups =</span> <span class="st">'drop'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-103"><a href="#cb52-103" aria-hidden="true" tabindex="-1"></a>      <span class="fu">arrange</span>(species, model_type)</span>
<span id="cb52-104"><a href="#cb52-104" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb52-105"><a href="#cb52-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-106"><a href="#cb52-106" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb52-107"><a href="#cb52-107" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Skipping model comparison as 'df_fish_comparison' is empty or AUC_test_CV is missing.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb52-108"><a href="#cb52-108" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Fitting GLMM to compare AUC_test_CV across model types ---

Summary of GLMM for AUC_test_CV:
 Family: gaussian  ( identity )
Formula:          AUC_test_CV ~ model_type + (1 | species)
Data: df_fish_comparison

     AIC      BIC   logLik deviance df.resid 
 -2543.9  -2521.3   1276.9  -2553.9      661 

Random effects:

Conditional model:
 Groups   Name        Variance Std.Dev.
 species  (Intercept) 0.003404 0.05835 
 Residual             0.001154 0.03397 
Number of obs: 666, groups:  species, 12

Dispersion estimate for gaussian family (sigma^2): 0.00115 

Conditional model:
                             Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)                  0.841875   0.017006   49.51  &lt; 2e-16 ***
model_typebiotic_only       -0.025187   0.003224   -7.81 5.63e-15 ***
model_typecombined_host_env  0.013379   0.003279    4.08 4.50e-05 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

--- Generating boxplot of AUC_test_CV by model type ---</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/compare-fish-model-performance-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/compare-fish-model-performance-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Mean AUC_test_CV per species and model type:
# A tibble: 36 Ã— 3
   species                 model_type        mean_auc
   &lt;chr&gt;                   &lt;fct&gt;                &lt;dbl&gt;
 1 Amphiprion_akindynos    env_only             0.891
 2 Amphiprion_akindynos    biotic_only          0.765
 3 Amphiprion_akindynos    combined_host_env    0.896
 4 Amphiprion_allardi      env_only             0.946
 5 Amphiprion_allardi      biotic_only          0.959
 6 Amphiprion_allardi      combined_host_env    0.963
 7 Amphiprion_bicinctus    env_only             0.842
 8 Amphiprion_bicinctus    biotic_only          0.912
 9 Amphiprion_bicinctus    combined_host_env    0.916
10 Amphiprion_chrysopterus env_only             0.891
# â„¹ 26 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># (TTS analysis section will be added here once TSS data is available)</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ... (Code for GLMM and plots for TSS, similar to AUC, but commented out) ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="species-specific-anemonefish-model-performance-comparison" class="level2">
<h2 class="anchored" data-anchor-id="species-specific-anemonefish-model-performance-comparison">Species-Specific Anemonefish Model Performance Comparison</h2>
<p>This section performs a model comparison (environmental-only vs.&nbsp;biotic-only vs.&nbsp;combined) for each anemonefish species individually. For each species, an ANOVA is used to test for significant differences in mean Test AUC values across the model types, followed by Tukeyâ€™s HSD post-hoc tests if the ANOVA is significant. Boxplots illustrate these comparisons for each species.</p>
<div class="cell" data-cap="Comparison of Anemonefish Model Performance (Test AUC) for Each Species. For each species, boxplots show the distribution of Test AUC values from replicate runs for each model type (environmental-only, biotic-only, combined host+environment). Individual points represent single model runs (jittered), and large black points indicate the mean AUC for each model type. ANOVA and Tukey's HSD tests (where appropriate) are performed to assess significant differences between model types within each species.">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure df_fish_comparison is available from the previous chunk</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(df_fish_comparison) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(df_fish_comparison) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> <span class="st">"AUC_test_CV"</span> <span class="sc">%in%</span> <span class="fu">names</span>(df_fish_comparison)) {</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create a working version of the dataframe with a cleaned species column for this specific analysis.</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># This avoids modifying the original df_fish_comparison if it's used later in its original state.</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># It also handles cases where species names might have .csv suffixes.</span></span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>  df_fish_comparison_for_species_analysis <span class="ot">&lt;-</span> df_fish_comparison <span class="sc">%&gt;%</span></span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">species_cleaned_for_loop =</span> stringr<span class="sc">::</span><span class="fu">str_remove</span>(species, <span class="st">"</span><span class="sc">\\</span><span class="st">.csv$"</span>)) <span class="co"># Remove .csv suffix</span></span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get unique cleaned species names for the loop</span></span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>  all_anemonefish_species_for_loop <span class="ot">&lt;-</span> <span class="fu">unique</span>(df_fish_comparison_for_species_analysis<span class="sc">$</span>species_cleaned_for_loop)</span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Performing species-specific model performance comparisons for anemonefish...</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Loop through each unique anemonefish species</span></span>
<span id="cb56-15"><a href="#cb56-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (current_species_name_sanitized <span class="cf">in</span> all_anemonefish_species_for_loop) {</span>
<span id="cb56-16"><a href="#cb56-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-17"><a href="#cb56-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Create a more readable species name for output</span></span>
<span id="cb56-18"><a href="#cb56-18" aria-hidden="true" tabindex="-1"></a>    current_species_display_name <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, current_species_name_sanitized)</span>
<span id="cb56-19"><a href="#cb56-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-20"><a href="#cb56-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n\n</span><span class="st">---</span><span class="sc">\n</span><span class="st">"</span>) <span class="co"># Horizontal rule for separation in output</span></span>
<span id="cb56-21"><a href="#cb56-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"### Statistical Comparison for: *"</span>, current_species_display_name, <span class="st">"*</span><span class="sc">\n</span><span class="st">"</span>)) <span class="co"># Markdown sub-header</span></span>
<span id="cb56-22"><a href="#cb56-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-23"><a href="#cb56-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Subset data for the current species</span></span>
<span id="cb56-24"><a href="#cb56-24" aria-hidden="true" tabindex="-1"></a>    species_data_subset <span class="ot">&lt;-</span> df_fish_comparison <span class="sc">%&gt;%</span></span>
<span id="cb56-25"><a href="#cb56-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>(species <span class="sc">==</span> current_species_name_sanitized) <span class="sc">%&gt;%</span></span>
<span id="cb56-26"><a href="#cb56-26" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">model_type =</span> <span class="fu">factor</span>(model_type, </span>
<span id="cb56-27"><a href="#cb56-27" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"env_only"</span>, <span class="st">"biotic_only"</span>, <span class="st">"combined_host_env"</span>))) <span class="co"># Ensure factor levels</span></span>
<span id="cb56-28"><a href="#cb56-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-29"><a href="#cb56-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Check if there's enough data to proceed</span></span>
<span id="cb56-30"><a href="#cb56-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">nrow</span>(species_data_subset) <span class="sc">&lt;</span> <span class="dv">3</span> <span class="sc">||</span> <span class="fu">length</span>(<span class="fu">unique</span>(species_data_subset<span class="sc">$</span>model_type)) <span class="sc">&lt;</span> <span class="dv">2</span>) {</span>
<span id="cb56-31"><a href="#cb56-31" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"Not enough data or distinct model types for "</span>, current_species_display_name, <span class="st">" to perform comparison. Skipping.</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb56-32"><a href="#cb56-32" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb56-33"><a href="#cb56-33" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb56-34"><a href="#cb56-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-35"><a href="#cb56-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Check for sufficient replicates per model type</span></span>
<span id="cb56-36"><a href="#cb56-36" aria-hidden="true" tabindex="-1"></a>    replicates_summary <span class="ot">&lt;-</span> species_data_subset <span class="sc">%&gt;%</span></span>
<span id="cb56-37"><a href="#cb56-37" aria-hidden="true" tabindex="-1"></a>        <span class="fu">group_by</span>(model_type) <span class="sc">%&gt;%</span></span>
<span id="cb56-38"><a href="#cb56-38" aria-hidden="true" tabindex="-1"></a>        <span class="fu">summarise</span>(<span class="at">n_runs =</span> <span class="fu">n</span>(), <span class="at">.groups =</span> <span class="st">'drop'</span>)</span>
<span id="cb56-39"><a href="#cb56-39" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-40"><a href="#cb56-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Number of runs per model type:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb56-41"><a href="#cb56-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(knitr<span class="sc">::</span><span class="fu">kable</span>(replicates_summary, <span class="at">caption =</span> <span class="fu">paste</span>(<span class="st">"Replicates for"</span>, current_species_display_name)))</span>
<span id="cb56-42"><a href="#cb56-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-43"><a href="#cb56-43" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">any</span>(replicates_summary<span class="sc">$</span>n_runs <span class="sc">&lt;</span> <span class="dv">2</span>) <span class="sc">&amp;&amp;</span> <span class="fu">length</span>(<span class="fu">unique</span>(replicates_summary<span class="sc">$</span>model_type)) <span class="sc">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb56-44"><a href="#cb56-44" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">*Warning: At least one model type for "</span>, current_species_display_name, <span class="st">" has fewer than 2 replicate runs. ANOVA results may not be robust.*</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb56-45"><a href="#cb56-45" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb56-46"><a href="#cb56-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(<span class="fu">unique</span>(replicates_summary<span class="sc">$</span>model_type[replicates_summary<span class="sc">$</span>n_runs <span class="sc">&gt;</span> <span class="dv">0</span>])) <span class="sc">&lt;</span> <span class="dv">2</span>) { <span class="co"># Check model types with actual data</span></span>
<span id="cb56-47"><a href="#cb56-47" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">*Warning: "</span>, current_species_display_name, <span class="st">" does not have at least two different model types with data. Skipping ANOVA for this species.*</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb56-48"><a href="#cb56-48" aria-hidden="true" tabindex="-1"></a>        <span class="cf">next</span></span>
<span id="cb56-49"><a href="#cb56-49" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb56-50"><a href="#cb56-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-51"><a href="#cb56-51" aria-hidden="true" tabindex="-1"></a>    <span class="co"># --- Statistical Comparison using ANOVA for the current species ---</span></span>
<span id="cb56-52"><a href="#cb56-52" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">**ANOVA: Test AUC vs. Model Type**</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb56-53"><a href="#cb56-53" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-54"><a href="#cb56-54" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Check if AUC_test_CV has any variance for this species</span></span>
<span id="cb56-55"><a href="#cb56-55" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">na.omit</span>(species_data_subset<span class="sc">$</span>AUC_test_CV))) <span class="sc">&lt;</span> <span class="dv">2</span>) {</span>
<span id="cb56-56"><a href="#cb56-56" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"AUC_test_CV values are constant (or all NA) for "</span>, current_species_display_name, <span class="st">". Skipping ANOVA.</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb56-57"><a href="#cb56-57" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb56-58"><a href="#cb56-58" aria-hidden="true" tabindex="-1"></a>        anova_model_species <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb56-59"><a href="#cb56-59" aria-hidden="true" tabindex="-1"></a>          <span class="fu">aov</span>(AUC_test_CV <span class="sc">~</span> model_type, <span class="at">data =</span> species_data_subset)</span>
<span id="cb56-60"><a href="#cb56-60" aria-hidden="true" tabindex="-1"></a>        }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb56-61"><a href="#cb56-61" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"Error performing ANOVA for "</span>, current_species_display_name, <span class="st">": "</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb56-62"><a href="#cb56-62" aria-hidden="true" tabindex="-1"></a>          <span class="cn">NULL</span></span>
<span id="cb56-63"><a href="#cb56-63" aria-hidden="true" tabindex="-1"></a>        })</span>
<span id="cb56-64"><a href="#cb56-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-65"><a href="#cb56-65" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(anova_model_species)) {</span>
<span id="cb56-66"><a href="#cb56-66" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Print ANOVA summary</span></span>
<span id="cb56-67"><a href="#cb56-67" aria-hidden="true" tabindex="-1"></a>          <span class="fu">print</span>(<span class="fu">summary</span>(anova_model_species))</span>
<span id="cb56-68"><a href="#cb56-68" aria-hidden="true" tabindex="-1"></a>          </span>
<span id="cb56-69"><a href="#cb56-69" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Perform Tukey's HSD for pairwise comparisons if ANOVA is significant</span></span>
<span id="cb56-70"><a href="#cb56-70" aria-hidden="true" tabindex="-1"></a>          anova_summary_table <span class="ot">&lt;-</span> <span class="fu">summary</span>(anova_model_species)[[<span class="dv">1</span>]] <span class="co"># This is a list containing the table</span></span>
<span id="cb56-71"><a href="#cb56-71" aria-hidden="true" tabindex="-1"></a>          p_value_from_anova <span class="ot">&lt;-</span> anova_summary_table<span class="sc">$</span><span class="st">"Pr(&gt;F)"</span>[<span class="dv">1</span>] <span class="co"># Get the p-value for model_type</span></span>
<span id="cb56-72"><a href="#cb56-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-73"><a href="#cb56-73" aria-hidden="true" tabindex="-1"></a>          <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.na</span>(p_value_from_anova) <span class="sc">&amp;&amp;</span> p_value_from_anova <span class="sc">&lt;</span> <span class="fl">0.05</span>) {</span>
<span id="cb56-74"><a href="#cb56-74" aria-hidden="true" tabindex="-1"></a>            <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb56-75"><a href="#cb56-75" aria-hidden="true" tabindex="-1"></a>            tukey_hsd_results <span class="ot">&lt;-</span> <span class="fu">TukeyHSD</span>(anova_model_species)</span>
<span id="cb56-76"><a href="#cb56-76" aria-hidden="true" tabindex="-1"></a>            <span class="fu">print</span>(tukey_hsd_results)</span>
<span id="cb56-77"><a href="#cb56-77" aria-hidden="true" tabindex="-1"></a>          } <span class="cf">else</span> {</span>
<span id="cb56-78"><a href="#cb56-78" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.na</span>(p_value_from_anova)) {</span>
<span id="cb56-79"><a href="#cb56-79" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">ANOVA for model_type not significant (p = "</span>, <span class="fu">round</span>(p_value_from_anova, <span class="dv">4</span>), <span class="st">"), skipping Tukey's HSD.</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb56-80"><a href="#cb56-80" aria-hidden="true" tabindex="-1"></a>            } <span class="cf">else</span> {</span>
<span id="cb56-81"><a href="#cb56-81" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Could not determine ANOVA significance, skipping Tukey's HSD.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb56-82"><a href="#cb56-82" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb56-83"><a href="#cb56-83" aria-hidden="true" tabindex="-1"></a>          }</span>
<span id="cb56-84"><a href="#cb56-84" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb56-85"><a href="#cb56-85" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb56-86"><a href="#cb56-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-87"><a href="#cb56-87" aria-hidden="true" tabindex="-1"></a>    <span class="co"># --- Visualization for the current species ---</span></span>
<span id="cb56-88"><a href="#cb56-88" aria-hidden="true" tabindex="-1"></a>    <span class="co"># cat("\n**Boxplot: Test AUC by Model Type**\n") # This title is now part of the plot</span></span>
<span id="cb56-89"><a href="#cb56-89" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-90"><a href="#cb56-90" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Determine dynamic y-axis limits, ensuring they are sensible for AUC</span></span>
<span id="cb56-91"><a href="#cb56-91" aria-hidden="true" tabindex="-1"></a>    min_auc_val <span class="ot">&lt;-</span> <span class="fu">min</span>(species_data_subset<span class="sc">$</span>AUC_test_CV, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb56-92"><a href="#cb56-92" aria-hidden="true" tabindex="-1"></a>    max_auc_val <span class="ot">&lt;-</span> <span class="fu">max</span>(species_data_subset<span class="sc">$</span>AUC_test_CV, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb56-93"><a href="#cb56-93" aria-hidden="true" tabindex="-1"></a>    y_lower_limit <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="dv">0</span>, <span class="fu">min</span>(<span class="fl">0.4</span>, min_auc_val <span class="sc">-</span> <span class="fl">0.05</span>)) <span class="co"># Not below 0, and give some space</span></span>
<span id="cb56-94"><a href="#cb56-94" aria-hidden="true" tabindex="-1"></a>    y_upper_limit <span class="ot">&lt;-</span> <span class="fu">min</span>(<span class="dv">1</span>, max_auc_val <span class="sc">+</span> <span class="fl">0.05</span>)         <span class="co"># Not above 1, and give some space</span></span>
<span id="cb56-95"><a href="#cb56-95" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">is.infinite</span>(y_lower_limit) <span class="sc">||</span> <span class="fu">is.infinite</span>(y_upper_limit)) { <span class="co"># Fallback if all NAs</span></span>
<span id="cb56-96"><a href="#cb56-96" aria-hidden="true" tabindex="-1"></a>        y_lower_limit <span class="ot">&lt;-</span> <span class="fl">0.4</span></span>
<span id="cb56-97"><a href="#cb56-97" aria-hidden="true" tabindex="-1"></a>        y_upper_limit <span class="ot">&lt;-</span> <span class="fl">1.0</span></span>
<span id="cb56-98"><a href="#cb56-98" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb56-99"><a href="#cb56-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-100"><a href="#cb56-100" aria-hidden="true" tabindex="-1"></a>    plot_auc_species_specific <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(species_data_subset, <span class="fu">aes</span>(<span class="at">x =</span> model_type, <span class="at">y =</span> AUC_test_CV, <span class="at">fill =</span> model_type)) <span class="sc">+</span></span>
<span id="cb56-101"><a href="#cb56-101" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>, <span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span> <span class="co"># Hiding default outliers as geom_jitter shows all points</span></span>
<span id="cb56-102"><a href="#cb56-102" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_jitter</span>(<span class="at">width =</span> <span class="fl">0.15</span>, <span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">height =</span> <span class="dv">0</span>, <span class="at">size =</span> <span class="fl">1.5</span>) <span class="sc">+</span> <span class="co"># Show individual run points</span></span>
<span id="cb56-103"><a href="#cb56-103" aria-hidden="true" tabindex="-1"></a>      <span class="fu">stat_summary</span>(<span class="at">fun =</span> mean, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">23</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">fill =</span> <span class="st">"white"</span>, <span class="at">color=</span><span class="st">"black"</span>, <span class="co"># Diamond for mean</span></span>
<span id="cb56-104"><a href="#cb56-104" aria-hidden="true" tabindex="-1"></a>                   <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.75</span>)) <span class="sc">+</span></span>
<span id="cb56-105"><a href="#cb56-105" aria-hidden="true" tabindex="-1"></a>      <span class="fu">labs</span>(</span>
<span id="cb56-106"><a href="#cb56-106" aria-hidden="true" tabindex="-1"></a>        <span class="at">title =</span> <span class="fu">paste</span>(<span class="st">"Model Performance (Test AUC) for"</span>, current_species_display_name),</span>
<span id="cb56-107"><a href="#cb56-107" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">"Model Type"</span>,</span>
<span id="cb56-108"><a href="#cb56-108" aria-hidden="true" tabindex="-1"></a>        <span class="at">y =</span> <span class="st">"Test AUC Score"</span></span>
<span id="cb56-109"><a href="#cb56-109" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb56-110"><a href="#cb56-110" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_fill_manual</span>(</span>
<span id="cb56-111"><a href="#cb56-111" aria-hidden="true" tabindex="-1"></a>        <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"env_only"</span> <span class="ot">=</span> <span class="st">"grey70"</span>, <span class="st">"biotic_only"</span> <span class="ot">=</span> <span class="st">"skyblue"</span>, <span class="st">"combined_host_env"</span> <span class="ot">=</span> <span class="st">"salmon"</span>),</span>
<span id="cb56-112"><a href="#cb56-112" aria-hidden="true" tabindex="-1"></a>        <span class="at">name =</span> <span class="st">"Model Type:"</span>,</span>
<span id="cb56-113"><a href="#cb56-113" aria-hidden="true" tabindex="-1"></a>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Environmental Only"</span>, <span class="st">"Biotic (Host) Only"</span>, <span class="st">"Combined (Host+Env)"</span>)</span>
<span id="cb56-114"><a href="#cb56-114" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb56-115"><a href="#cb56-115" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb56-116"><a href="#cb56-116" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(</span>
<span id="cb56-117"><a href="#cb56-117" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"top"</span>, </span>
<span id="cb56-118"><a href="#cb56-118" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">9</span>),</span>
<span id="cb56-119"><a href="#cb56-119" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">13</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb56-120"><a href="#cb56-120" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>()</span>
<span id="cb56-121"><a href="#cb56-121" aria-hidden="true" tabindex="-1"></a>      ) <span class="sc">+</span></span>
<span id="cb56-122"><a href="#cb56-122" aria-hidden="true" tabindex="-1"></a>      <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(y_lower_limit, y_upper_limit)) <span class="co"># Apply dynamic y-limits</span></span>
<span id="cb56-123"><a href="#cb56-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-124"><a href="#cb56-124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(plot_auc_species_specific)</span>
<span id="cb56-125"><a href="#cb56-125" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb56-126"><a href="#cb56-126" aria-hidden="true" tabindex="-1"></a>  } <span class="co"># End loop through species</span></span>
<span id="cb56-127"><a href="#cb56-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-128"><a href="#cb56-128" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb56-129"><a href="#cb56-129" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Skipping species-specific anemonefish model comparison as 'df_fish_comparison' is not available, empty, or 'AUC_test_CV' column is missing.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb56-130"><a href="#cb56-130" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Performing species-specific model performance comparisons for anemonefish...


---
### Statistical Comparison for: *Amphiprion akindynos*

Number of runs per model type:


Table: Replicates for Amphiprion akindynos

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     14|
|biotic_only       |     20|
|combined_host_env |     15|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq Mean Sq F value Pr(&gt;F)    
model_type   2 0.19650 0.09825   104.3 &lt;2e-16 ***
Residuals   46 0.04335 0.00094                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                      diff         lwr         upr     p adj
biotic_only-env_only          -0.126565714 -0.15247178 -0.10065965 0.0000000
combined_host_env-env_only     0.004320952 -0.02330583  0.03194774 0.9240926
combined_host_env-biotic_only  0.130886667  0.10549364  0.15627969 0.0000000</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion allardi*

Number of runs per model type:


Table: Replicates for Amphiprion allardi

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |      9|
|biotic_only       |     20|
|combined_host_env |     12|

**ANOVA: Test AUC vs. Model Type**
            Df   Sum Sq  Mean Sq F value   Pr(&gt;F)    
model_type   2 0.001532 7.66e-04    11.4 0.000132 ***
Residuals   38 0.002552 6.72e-05                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                     diff          lwr        upr     p adj
biotic_only-env_only          0.013267778  0.005244974 0.02129058 0.0007330
combined_host_env-env_only    0.016344444  0.007530685 0.02515820 0.0001691
combined_host_env-biotic_only 0.003076667 -0.004221819 0.01037515 0.5639999</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion bicinctus*

Number of runs per model type:


Table: Replicates for Amphiprion bicinctus

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     14|
|biotic_only       |     20|
|combined_host_env |     17|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq  Mean Sq F value   Pr(&gt;F)    
model_type   2 0.05219 0.026094   42.05 2.81e-11 ***
Residuals   48 0.02979 0.000621                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                     diff         lwr        upr    p adj
biotic_only-env_only          0.069657857  0.04866412 0.09065159 0.000000
combined_host_env-env_only    0.073822269  0.05207916 0.09556538 0.000000
combined_host_env-biotic_only 0.004164412 -0.01570983 0.02403866 0.868409</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion chrysopterus*

Number of runs per model type:


Table: Replicates for Amphiprion chrysopterus

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     20|
|biotic_only       |     20|
|combined_host_env |     20|

**ANOVA: Test AUC vs. Model Type**
            Df   Sum Sq  Mean Sq F value Pr(&gt;F)    
model_type   2 0.021049 0.010525   77.32 &lt;2e-16 ***
Residuals   57 0.007759 0.000136                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                   diff          lwr          upr     p adj
biotic_only-env_only          -0.039815 -0.048693393 -0.030936607 0.0000000
combined_host_env-env_only    -0.000165 -0.009043393  0.008713393 0.9988979
combined_host_env-biotic_only  0.039650  0.030771607  0.048528393 0.0000000</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion clarkii*

Number of runs per model type:


Table: Replicates for Amphiprion clarkii

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     20|
|biotic_only       |     20|
|combined_host_env |     20|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq   Mean Sq F value Pr(&gt;F)
model_type   2 0.00028 0.0001412   0.231  0.794
Residuals   57 0.03484 0.0006112               

ANOVA for model_type not significant (p = 0.7944), skipping Tukey's HSD.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion melanopus*

Number of runs per model type:


Table: Replicates for Amphiprion melanopus

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     20|
|biotic_only       |     20|
|combined_host_env |     20|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq  Mean Sq F value   Pr(&gt;F)    
model_type   2 0.00664 0.003320   13.06 2.13e-05 ***
Residuals   57 0.01448 0.000254                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                   diff           lwr          upr     p adj
biotic_only-env_only          -0.013230 -0.0253604436 -0.001099556 0.0294763
combined_host_env-env_only     0.012535  0.0004045564  0.024665444 0.0413318
combined_host_env-biotic_only  0.025765  0.0136345564  0.037895444 0.0000115</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-6.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion ocellaris*

Number of runs per model type:


Table: Replicates for Amphiprion ocellaris

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     19|
|biotic_only       |     20|
|combined_host_env |     19|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq  Mean Sq F value   Pr(&gt;F)    
model_type   2 0.01453 0.007263   8.701 0.000521 ***
Residuals   55 0.04591 0.000835                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                      diff          lwr        upr     p adj
biotic_only-env_only          -0.009586579 -0.031881021 0.01270786 0.5577299
combined_host_env-env_only     0.027747368  0.005168909 0.05032583 0.0123967
combined_host_env-biotic_only  0.037333947  0.015039505 0.05962839 0.0004949</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-7.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion percula*

Number of runs per model type:


Table: Replicates for Amphiprion percula

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     19|
|biotic_only       |     20|
|combined_host_env |     19|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq  Mean Sq F value Pr(&gt;F)  
model_type   2 0.01019 0.005096   2.785 0.0704 .
Residuals   55 0.10064 0.001830                 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

ANOVA for model_type not significant (p = 0.0704), skipping Tukey's HSD.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-8.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion perideraion*

Number of runs per model type:


Table: Replicates for Amphiprion perideraion

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     20|
|biotic_only       |     20|
|combined_host_env |     20|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq  Mean Sq F value   Pr(&gt;F)    
model_type   2 0.05940 0.029700   63.94 2.73e-15 ***
Residuals   57 0.02647 0.000464                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                   diff        lwr        upr     p adj
biotic_only-env_only          -0.066245 -0.0826452 -0.0498448 0.0000000
combined_host_env-env_only     0.000990 -0.0154102  0.0173902 0.9884364
combined_host_env-biotic_only  0.067235  0.0508348  0.0836352 0.0000000</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-9.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion polymnus*

Number of runs per model type:


Table: Replicates for Amphiprion polymnus

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     20|
|biotic_only       |     20|
|combined_host_env |     20|

**ANOVA: Test AUC vs. Model Type**
            Df   Sum Sq   Mean Sq F value Pr(&gt;F)  
model_type   2 0.004107 0.0020533   4.784  0.012 *
Residuals   57 0.024465 0.0004292                 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                   diff          lwr         upr     p adj
biotic_only-env_only          -0.006660 -0.022425377 0.009105377 0.5695440
combined_host_env-env_only     0.013245 -0.002520377 0.029010377 0.1162085
combined_host_env-biotic_only  0.019905  0.004139623 0.035670377 0.0098875</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-10.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Amphiprion sandaracinos*

Number of runs per model type:


Table: Replicates for Amphiprion sandaracinos

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     20|
|biotic_only       |     20|
|combined_host_env |     20|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq  Mean Sq F value Pr(&gt;F)  
model_type   2 0.02273 0.011365   4.317  0.018 *
Residuals   57 0.15006 0.002633                 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                   diff          lwr           upr     p adj
biotic_only-env_only          -0.039745 -0.078790163 -0.0006998368 0.0451638
combined_host_env-env_only     0.002930 -0.036115163  0.0419751632 0.9821890
combined_host_env-biotic_only  0.042675  0.003629837  0.0817201632 0.0290671</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-11.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

---
### Statistical Comparison for: *Premnas biaculeatus*

Number of runs per model type:


Table: Replicates for Premnas biaculeatus

|model_type        | n_runs|
|:-----------------|------:|
|env_only          |     18|
|biotic_only       |     16|
|combined_host_env |     15|

**ANOVA: Test AUC vs. Model Type**
            Df  Sum Sq  Mean Sq F value   Pr(&gt;F)    
model_type   2 0.03610 0.018050   35.44 4.85e-10 ***
Residuals   46 0.02343 0.000509                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

**Tukey's HSD Post-Hoc Test (Pairwise Comparisons)**
  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = AUC_test_CV ~ model_type, data = species_data_subset)

$model_type
                                     diff          lwr         upr     p adj
biotic_only-env_only          -0.04929722 -0.068077682 -0.03051676 0.0000003
combined_host_env-env_only     0.01541111 -0.003697895  0.03452012 0.1355503
combined_host_env-biotic_only  0.06470833  0.045063958  0.08435271 0.0000000</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/species-specific-fish-model-performance-12.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="current-environmental-predictors-pca" class="level2">
<h2 class="anchored" data-anchor-id="current-environmental-predictors-pca">Current Environmental Predictors (PCA)</h2>
<p>This section loads the Principal Component Analysis (PCA) rasters for the current environmental conditions. These PCA rasters were generated by script <code>05_preprocess_env_pca_only.R</code> (if PCA was used) and represent the primary environmental gradients used in the SDMs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (config<span class="sc">$</span>use_pca_predictors) {</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"--- Loading Current Environmental PCA Rasters ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># The rds file stores a list of paths, one for each scenario's PCA .tif file</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(config<span class="sc">$</span>pca_raster_paths_rds_path)) {</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"PCA raster paths RDS file not found: "</span>, config<span class="sc">$</span>pca_raster_paths_rds_path, </span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>         <span class="st">"</span><span class="sc">\n</span><span class="st">Please ensure script 05 (PCA preprocessing) ran successfully."</span>)</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>  all_pca_raster_paths <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(config<span class="sc">$</span>pca_raster_paths_rds_path)</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>  current_pca_path <span class="ot">&lt;-</span> all_pca_raster_paths[[<span class="st">"current"</span>]]</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(current_pca_path) <span class="sc">||</span> <span class="sc">!</span><span class="fu">file.exists</span>(current_pca_path)) {</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"Path for 'current' PCA raster not found in RDS or file does not exist: "</span>, current_pca_path <span class="sc">%||%</span> <span class="st">"NULL"</span>)</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a>  env_pca_current <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a>    terra<span class="sc">::</span><span class="fu">rast</span>(current_pca_path)</span>
<span id="cb69-17"><a href="#cb69-17" aria-hidden="true" tabindex="-1"></a>  }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb69-18"><a href="#cb69-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Error loading current PCA raster from:"</span>, current_pca_path, <span class="st">"</span><span class="sc">\n</span><span class="st">Error:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb69-19"><a href="#cb69-19" aria-hidden="true" tabindex="-1"></a>    <span class="cn">NULL</span></span>
<span id="cb69-20"><a href="#cb69-20" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb69-21"><a href="#cb69-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb69-22"><a href="#cb69-22" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(env_pca_current)) {</span>
<span id="cb69-23"><a href="#cb69-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(env_pca_current) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span>terra<span class="sc">::</span><span class="fu">nlyr</span>(env_pca_current)) <span class="co"># Ensure standard names</span></span>
<span id="cb69-24"><a href="#cb69-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Loaded current environmental PCA rasters. Layers:"</span>, <span class="fu">paste</span>(<span class="fu">names</span>(env_pca_current), <span class="at">collapse=</span><span class="st">", "</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb69-25"><a href="#cb69-25" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb69-26"><a href="#cb69-26" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Crop if configured (should match how other current rasters were handled)</span></span>
<span id="cb69-27"><a href="#cb69-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (config<span class="sc">$</span>apply_indo_pacific_crop) {</span>
<span id="cb69-28"><a href="#cb69-28" aria-hidden="true" tabindex="-1"></a>      ip_extent <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">ext</span>(config<span class="sc">$</span>indo_pacific_bbox)</span>
<span id="cb69-29"><a href="#cb69-29" aria-hidden="true" tabindex="-1"></a>      env_pca_current_cropped <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">crop</span>(env_pca_current, ip_extent)</span>
<span id="cb69-30"><a href="#cb69-30" aria-hidden="true" tabindex="-1"></a>      <span class="fu">plot</span>(env_pca_current_cropped, <span class="at">main =</span> <span class="st">"Current Environmental PCA (Cropped)"</span>)</span>
<span id="cb69-31"><a href="#cb69-31" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb69-32"><a href="#cb69-32" aria-hidden="true" tabindex="-1"></a>      env_pca_current_cropped <span class="ot">&lt;-</span> env_pca_current <span class="co"># Use uncropped if not configured</span></span>
<span id="cb69-33"><a href="#cb69-33" aria-hidden="true" tabindex="-1"></a>      <span class="fu">plot</span>(env_pca_current_cropped, <span class="at">main =</span> <span class="st">"Current Environmental PCA (Uncropped)"</span>)</span>
<span id="cb69-34"><a href="#cb69-34" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb69-35"><a href="#cb69-35" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb69-36"><a href="#cb69-36" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Error: Could not load current environmental PCA rasters.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb69-37"><a href="#cb69-37" aria-hidden="true" tabindex="-1"></a>    env_pca_current_cropped <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb69-38"><a href="#cb69-38" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb69-39"><a href="#cb69-39" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb69-40"><a href="#cb69-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"--- PCA predictors not used. Skipping loading of current PCA rasters. ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb69-41"><a href="#cb69-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"    You might need to load your VIF-selected current environmental rasters here if needed for specific analyses.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb69-42"><a href="#cb69-42" aria-hidden="true" tabindex="-1"></a>  env_pca_current_cropped <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="co"># Ensure this object exists as NULL if PCA not used</span></span>
<span id="cb69-43"><a href="#cb69-43" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Loading Current Environmental PCA Rasters ---
Loaded current environmental PCA rasters. Layers: PC1, PC2, PC3, PC4 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-current-env-pca-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="future-species-richness-projections" class="level2">
<h2 class="anchored" data-anchor-id="future-species-richness-projections">Future Species Richness Projections</h2>
<p>This section loads the predicted future suitability rasters for host sea anemones and anemonefish (from environmental-only models) for each future scenario and time step, and calculates the summed richness.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the future scenarios and time steps from your config</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Your config$env_scenarios already includes "current", so we filter that out</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>future_scenarios_to_load <span class="ot">&lt;-</span> config<span class="sc">$</span>env_scenarios[config<span class="sc">$</span>env_scenarios <span class="sc">!=</span> <span class="st">"current"</span>]</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(future_scenarios_to_load) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"No future scenarios defined in config$env_scenarios to load predictions for.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Lists to store the summed richness rasters for each future scenario</span></span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>host_richness_future_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>fish_richness_future_list <span class="ot">&lt;-</span> <span class="fu">list</span>() <span class="co"># For env-only fish models</span></span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each future scenario</span></span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (scenario_name <span class="cf">in</span> future_scenarios_to_load) {</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Processing Future Scenario:"</span>, scenario_name, <span class="st">"---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- 1. Load Host Sea Anemone Future Predictions ---</span></span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"  Loading Host Sea Anemone predictions for"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-18"><a href="#cb71-18" aria-hidden="true" tabindex="-1"></a>  host_future_raster_files <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb71-19"><a href="#cb71-19" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(anemone_species_df)) { <span class="co"># Uses anemone_species_df from previous chunk</span></span>
<span id="cb71-20"><a href="#cb71-20" aria-hidden="true" tabindex="-1"></a>    sp_name_sanitized <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">" "</span>, <span class="st">"_"</span>, anemone_species_df<span class="sc">$</span>scientificName[i])</span>
<span id="cb71-21"><a href="#cb71-21" aria-hidden="true" tabindex="-1"></a>    pred_file_path <span class="ot">&lt;-</span> <span class="fu">construct_prediction_filename</span>(</span>
<span id="cb71-22"><a href="#cb71-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">species_name_sanitized =</span> sp_name_sanitized,</span>
<span id="cb71-23"><a href="#cb71-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">scenario_name =</span> scenario_name,</span>
<span id="cb71-24"><a href="#cb71-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">predictor_type_suffix =</span> predictor_suffix, <span class="co"># Suffix for anemone env-only models</span></span>
<span id="cb71-25"><a href="#cb71-25" aria-hidden="true" tabindex="-1"></a>      <span class="at">config =</span> config</span>
<span id="cb71-26"><a href="#cb71-26" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb71-27"><a href="#cb71-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">file.exists</span>(pred_file_path)) {</span>
<span id="cb71-28"><a href="#cb71-28" aria-hidden="true" tabindex="-1"></a>      host_future_raster_files <span class="ot">&lt;-</span> <span class="fu">c</span>(host_future_raster_files, pred_file_path)</span>
<span id="cb71-29"><a href="#cb71-29" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb71-30"><a href="#cb71-30" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Host prediction file not found for"</span>, sp_name_sanitized, <span class="st">"in"</span>, scenario_name, <span class="st">"at"</span>, pred_file_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-31"><a href="#cb71-31" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb71-32"><a href="#cb71-32" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb71-33"><a href="#cb71-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-34"><a href="#cb71-34" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(host_future_raster_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb71-35"><a href="#cb71-35" aria-hidden="true" tabindex="-1"></a>    host_pred_stack_future <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">rast</span>(host_future_raster_files)</span>
<span id="cb71-36"><a href="#cb71-36" aria-hidden="true" tabindex="-1"></a>    host_richness_sum_future <span class="ot">&lt;-</span> <span class="fu">sum</span>(host_pred_stack_future, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb71-37"><a href="#cb71-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(host_richness_sum_future) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"HostAnemoneRichness_"</span>, scenario_name)</span>
<span id="cb71-38"><a href="#cb71-38" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-39"><a href="#cb71-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Crop to Indo-Pacific extent</span></span>
<span id="cb71-40"><a href="#cb71-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (config<span class="sc">$</span>apply_indo_pacific_crop) {</span>
<span id="cb71-41"><a href="#cb71-41" aria-hidden="true" tabindex="-1"></a>      ip_extent <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">ext</span>(config<span class="sc">$</span>indo_pacific_bbox)</span>
<span id="cb71-42"><a href="#cb71-42" aria-hidden="true" tabindex="-1"></a>      host_richness_future_list[[scenario_name]] <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">crop</span>(host_richness_sum_future, ip_extent)</span>
<span id="cb71-43"><a href="#cb71-43" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb71-44"><a href="#cb71-44" aria-hidden="true" tabindex="-1"></a>      host_richness_future_list[[scenario_name]] <span class="ot">&lt;-</span> host_richness_sum_future</span>
<span id="cb71-45"><a href="#cb71-45" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb71-46"><a href="#cb71-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Processed host anemone richness for"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-47"><a href="#cb71-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot(host_richness_future_list[[scenario_name]], main = paste("Host Richness -", scenario_name))</span></span>
<span id="cb71-48"><a href="#cb71-48" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb71-49"><a href="#cb71-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Warning: No host prediction files found for scenario"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-50"><a href="#cb71-50" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb71-51"><a href="#cb71-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-52"><a href="#cb71-52" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- 2. Load Anemonefish (Environmental-Only) Future Predictions ---</span></span>
<span id="cb71-53"><a href="#cb71-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"  Loading Anemonefish (Env-Only) predictions for"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-54"><a href="#cb71-54" aria-hidden="true" tabindex="-1"></a>  fish_future_raster_files <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb71-55"><a href="#cb71-55" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(anemonefish_species_df)) { <span class="co"># Uses anemonefish_species_df from previous chunk</span></span>
<span id="cb71-56"><a href="#cb71-56" aria-hidden="true" tabindex="-1"></a>    sp_name_sanitized <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">" "</span>, <span class="st">"_"</span>, anemonefish_species_df<span class="sc">$</span>scientificName[i])</span>
<span id="cb71-57"><a href="#cb71-57" aria-hidden="true" tabindex="-1"></a>    pred_file_path <span class="ot">&lt;-</span> <span class="fu">construct_prediction_filename</span>(</span>
<span id="cb71-58"><a href="#cb71-58" aria-hidden="true" tabindex="-1"></a>      <span class="at">species_name_sanitized =</span> sp_name_sanitized,</span>
<span id="cb71-59"><a href="#cb71-59" aria-hidden="true" tabindex="-1"></a>      <span class="at">scenario_name =</span> scenario_name,</span>
<span id="cb71-60"><a href="#cb71-60" aria-hidden="true" tabindex="-1"></a>      <span class="at">predictor_type_suffix =</span> predictor_suffix, <span class="co"># Suffix for anemonefish env-only models</span></span>
<span id="cb71-61"><a href="#cb71-61" aria-hidden="true" tabindex="-1"></a>      <span class="at">config =</span> config</span>
<span id="cb71-62"><a href="#cb71-62" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb71-63"><a href="#cb71-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">file.exists</span>(pred_file_path)) {</span>
<span id="cb71-64"><a href="#cb71-64" aria-hidden="true" tabindex="-1"></a>      fish_future_raster_files <span class="ot">&lt;-</span> <span class="fu">c</span>(fish_future_raster_files, pred_file_path)</span>
<span id="cb71-65"><a href="#cb71-65" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb71-66"><a href="#cb71-66" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Anemonefish (env-only) prediction file not found for"</span>, sp_name_sanitized, <span class="st">"in"</span>, scenario_name, <span class="st">"at"</span>, pred_file_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-67"><a href="#cb71-67" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb71-68"><a href="#cb71-68" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb71-69"><a href="#cb71-69" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-70"><a href="#cb71-70" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(fish_future_raster_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb71-71"><a href="#cb71-71" aria-hidden="true" tabindex="-1"></a>    fish_pred_stack_future <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">rast</span>(fish_future_raster_files)</span>
<span id="cb71-72"><a href="#cb71-72" aria-hidden="true" tabindex="-1"></a>    fish_richness_sum_future <span class="ot">&lt;-</span> <span class="fu">sum</span>(fish_pred_stack_future, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb71-73"><a href="#cb71-73" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(fish_richness_sum_future) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"AnemonefishRichness_EnvOnly_"</span>, scenario_name)</span>
<span id="cb71-74"><a href="#cb71-74" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-75"><a href="#cb71-75" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (config<span class="sc">$</span>apply_indo_pacific_crop) {</span>
<span id="cb71-76"><a href="#cb71-76" aria-hidden="true" tabindex="-1"></a>      ip_extent <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">ext</span>(config<span class="sc">$</span>indo_pacific_bbox)</span>
<span id="cb71-77"><a href="#cb71-77" aria-hidden="true" tabindex="-1"></a>      fish_richness_future_list[[scenario_name]] <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">crop</span>(fish_richness_sum_future, ip_extent)</span>
<span id="cb71-78"><a href="#cb71-78" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb71-79"><a href="#cb71-79" aria-hidden="true" tabindex="-1"></a>      fish_richness_future_list[[scenario_name]] <span class="ot">&lt;-</span> fish_richness_sum_future</span>
<span id="cb71-80"><a href="#cb71-80" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb71-81"><a href="#cb71-81" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Processed anemonefish (env-only) richness for"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-82"><a href="#cb71-82" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(fish_richness_future_list[[scenario_name]], <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"Anemonefish (Env-Only) Richness -"</span>, scenario_name))</span>
<span id="cb71-83"><a href="#cb71-83" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb71-84"><a href="#cb71-84" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Warning: No anemonefish (env-only) prediction files found for scenario"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb71-85"><a href="#cb71-85" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb71-86"><a href="#cb71-86" aria-hidden="true" tabindex="-1"></a>} <span class="co"># End loop through future scenarios</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Processing Future Scenario: ssp119_2050 ---
  Loading Host Sea Anemone predictions for ssp119_2050 
  Warning: Host prediction file not found for Radianthus_aurora in ssp119_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Radianthus_aurora_ssp119_dec50_pca.tif 

|---------|---------|---------|---------|
=========================================
                                          
  Processed host anemone richness for ssp119_2050 
  Loading Anemonefish (Env-Only) predictions for ssp119_2050 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_akindynos in ssp119_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_akindynos_ssp119_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_allardi in ssp119_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_allardi_ssp119_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_bicinctus in ssp119_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_bicinctus_ssp119_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_chrysopterus in ssp119_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_chrysopterus_ssp119_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_melanopus in ssp119_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_melanopus_ssp119_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_percula in ssp119_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_percula_ssp119_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Premnas_biaculeatus in ssp119_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Premnas_biaculeatus_ssp119_dec50_pca.tif 

|---------|---------|---------|---------|
=========================================
                                          
  Processed anemonefish (env-only) richness for ssp119_2050 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-species-predictions-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Processing Future Scenario: ssp119_2100 ---
  Loading Host Sea Anemone predictions for ssp119_2100 
  Warning: Host prediction file not found for Radianthus_aurora in ssp119_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Radianthus_aurora_ssp119_dec100_pca.tif 

|---------|---------|---------|---------|
=========================================
                                          
  Processed host anemone richness for ssp119_2100 
  Loading Anemonefish (Env-Only) predictions for ssp119_2100 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_akindynos in ssp119_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_akindynos_ssp119_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_allardi in ssp119_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_allardi_ssp119_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_bicinctus in ssp119_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_bicinctus_ssp119_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_chrysopterus in ssp119_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_chrysopterus_ssp119_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_melanopus in ssp119_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_melanopus_ssp119_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_percula in ssp119_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Amphiprion_percula_ssp119_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Premnas_biaculeatus in ssp119_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp119/mean_pred_Premnas_biaculeatus_ssp119_dec100_pca.tif 

|---------|---------|---------|---------|
=========================================
                                          
  Processed anemonefish (env-only) richness for ssp119_2100 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-species-predictions-2.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Processing Future Scenario: ssp585_2050 ---
  Loading Host Sea Anemone predictions for ssp585_2050 
  Warning: Host prediction file not found for Radianthus_aurora in ssp585_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Radianthus_aurora_ssp585_dec50_pca.tif 

|---------|---------|---------|---------|
=========================================
                                          
  Processed host anemone richness for ssp585_2050 
  Loading Anemonefish (Env-Only) predictions for ssp585_2050 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_akindynos in ssp585_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_akindynos_ssp585_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_allardi in ssp585_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_allardi_ssp585_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_bicinctus in ssp585_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_bicinctus_ssp585_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_chrysopterus in ssp585_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_chrysopterus_ssp585_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_melanopus in ssp585_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_melanopus_ssp585_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_percula in ssp585_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_percula_ssp585_dec50_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Premnas_biaculeatus in ssp585_2050 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Premnas_biaculeatus_ssp585_dec50_pca.tif 

|---------|---------|---------|---------|
=========================================
                                          
  Processed anemonefish (env-only) richness for ssp585_2050 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-species-predictions-3.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Processing Future Scenario: ssp585_2100 ---
  Loading Host Sea Anemone predictions for ssp585_2100 
  Warning: Host prediction file not found for Radianthus_aurora in ssp585_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Radianthus_aurora_ssp585_dec100_pca.tif 

|---------|---------|---------|---------|
=========================================
                                          
  Processed host anemone richness for ssp585_2100 
  Loading Anemonefish (Env-Only) predictions for ssp585_2100 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_akindynos in ssp585_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_akindynos_ssp585_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_allardi in ssp585_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_allardi_ssp585_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_bicinctus in ssp585_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_bicinctus_ssp585_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_chrysopterus in ssp585_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_chrysopterus_ssp585_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_melanopus in ssp585_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_melanopus_ssp585_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Amphiprion_percula in ssp585_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Amphiprion_percula_ssp585_dec100_pca.tif 
  Warning: Anemonefish (env-only) prediction file not found for Premnas_biaculeatus in ssp585_2100 at /Users/chrisrauch/anemonefish_post_analysis/data/output/predictions/Future/ssp585/mean_pred_Premnas_biaculeatus_ssp585_dec100_pca.tif 

|---------|---------|---------|---------|
=========================================
                                          
  Processed anemonefish (env-only) richness for ssp585_2100 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-species-predictions-4.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># At this point, host_richness_future_list and fish_richness_future_list </span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="co"># are lists of SpatRasters, named by scenario (e.g., host_richness_future_list[["ssp119_2050"]])</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Example: Plot one of the future richness maps</span></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(host_richness_future_list) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> <span class="st">"ssp119_2050"</span> <span class="sc">%in%</span> <span class="fu">names</span>(host_richness_future_list)) {</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(host_richness_future_list[[<span class="st">"ssp119_2050"</span>]], <span class="at">main =</span> <span class="st">"Host Richness - SSP1-1.9 (2050) - Cropped"</span>)</span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-species-predictions-5.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(fish_richness_future_list) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> <span class="st">"ssp585_2100"</span> <span class="sc">%in%</span> <span class="fu">names</span>(fish_richness_future_list)) {</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(fish_richness_future_list[[<span class="st">"ssp585_2100"</span>]], <span class="at">main =</span> <span class="st">"Anemonefish (Env-Only) Richness - SSP5-8.5 (2100) - Cropped"</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-species-predictions-6.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Future species prediction loading and processing finished. ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Future species prediction loading and processing finished. ---</code></pre>
</div>
</div>
</section>
<section id="future-environmental-predictors-pca" class="level2">
<h2 class="anchored" data-anchor-id="future-environmental-predictors-pca">Future Environmental Predictors (PCA)</h2>
<p>This section loads the PCA rasters for the future environmental conditions, corresponding to each future scenario and time step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (config<span class="sc">$</span>use_pca_predictors) {</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"--- Loading Future Environmental PCA Rasters ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"all_pca_raster_paths"</span>) <span class="sc">||</span> <span class="fu">is.null</span>(all_pca_raster_paths)) {</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Warning: 'all_pca_raster_paths' not found from previous chunk. Attempting to reload.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(config<span class="sc">$</span>pca_raster_paths_rds_path)) {</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">stop</span>(<span class="st">"PCA raster paths RDS file not found: "</span>, config<span class="sc">$</span>pca_raster_paths_rds_path)</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>    all_pca_raster_paths <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(config<span class="sc">$</span>pca_raster_paths_rds_path)</span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>  env_pca_future_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (scenario_name <span class="cf">in</span> future_scenarios_to_load) { <span class="co"># future_scenarios_to_load from previous chunk</span></span>
<span id="cb80-14"><a href="#cb80-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Loading PCA for future scenario:"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb80-15"><a href="#cb80-15" aria-hidden="true" tabindex="-1"></a>    future_pca_path <span class="ot">&lt;-</span> all_pca_raster_paths[[scenario_name]]</span>
<span id="cb80-16"><a href="#cb80-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb80-17"><a href="#cb80-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">is.null</span>(future_pca_path) <span class="sc">||</span> <span class="sc">!</span><span class="fu">file.exists</span>(future_pca_path)) {</span>
<span id="cb80-18"><a href="#cb80-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Path for PCA raster not found for scenario:"</span>, scenario_name, <span class="st">"Path:"</span>, future_pca_path <span class="sc">%||%</span> <span class="st">"NULL"</span>, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb80-19"><a href="#cb80-19" aria-hidden="true" tabindex="-1"></a>      <span class="cf">next</span></span>
<span id="cb80-20"><a href="#cb80-20" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb80-21"><a href="#cb80-21" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb80-22"><a href="#cb80-22" aria-hidden="true" tabindex="-1"></a>    env_pca_future_scenario <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb80-23"><a href="#cb80-23" aria-hidden="true" tabindex="-1"></a>      terra<span class="sc">::</span><span class="fu">rast</span>(future_pca_path)</span>
<span id="cb80-24"><a href="#cb80-24" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb80-25"><a href="#cb80-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Error loading PCA raster for"</span>, scenario_name, <span class="st">"from:"</span>, future_pca_path, <span class="st">"</span><span class="sc">\n</span><span class="st">Error:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb80-26"><a href="#cb80-26" aria-hidden="true" tabindex="-1"></a>      <span class="cn">NULL</span></span>
<span id="cb80-27"><a href="#cb80-27" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb80-28"><a href="#cb80-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb80-29"><a href="#cb80-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(env_pca_future_scenario)) {</span>
<span id="cb80-30"><a href="#cb80-30" aria-hidden="true" tabindex="-1"></a>      <span class="fu">names</span>(env_pca_future_scenario) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span>terra<span class="sc">::</span><span class="fu">nlyr</span>(env_pca_future_scenario)) <span class="co"># Ensure standard names</span></span>
<span id="cb80-31"><a href="#cb80-31" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb80-32"><a href="#cb80-32" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (config<span class="sc">$</span>apply_indo_pacific_crop) {</span>
<span id="cb80-33"><a href="#cb80-33" aria-hidden="true" tabindex="-1"></a>        ip_extent <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">ext</span>(config<span class="sc">$</span>indo_pacific_bbox)</span>
<span id="cb80-34"><a href="#cb80-34" aria-hidden="true" tabindex="-1"></a>        env_pca_future_list[[scenario_name]] <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">crop</span>(env_pca_future_scenario, ip_extent)</span>
<span id="cb80-35"><a href="#cb80-35" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb80-36"><a href="#cb80-36" aria-hidden="true" tabindex="-1"></a>        env_pca_future_list[[scenario_name]] <span class="ot">&lt;-</span> env_pca_future_scenario</span>
<span id="cb80-37"><a href="#cb80-37" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb80-38"><a href="#cb80-38" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"    Loaded and processed PCA for"</span>, scenario_name, <span class="st">". Layers:"</span>, <span class="fu">paste</span>(<span class="fu">names</span>(env_pca_future_list[[scenario_name]]), <span class="at">collapse=</span><span class="st">", "</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb80-39"><a href="#cb80-39" aria-hidden="true" tabindex="-1"></a>      <span class="fu">plot</span>(env_pca_future_list[[scenario_name]], <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"Future Env PCA -"</span>, scenario_name))</span>
<span id="cb80-40"><a href="#cb80-40" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb80-41"><a href="#cb80-41" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb80-42"><a href="#cb80-42" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb80-43"><a href="#cb80-43" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Example: Plot one of the future PCA stacks</span></span>
<span id="cb80-44"><a href="#cb80-44" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(env_pca_future_list) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> <span class="st">"ssp119_2050"</span> <span class="sc">%in%</span> <span class="fu">names</span>(env_pca_future_list)) {</span>
<span id="cb80-45"><a href="#cb80-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(env_pca_future_list[[<span class="st">"ssp119_2050"</span>]], <span class="at">main =</span> <span class="st">"Future Env PCA - SSP1-1.9 (2050) - Cropped"</span>)</span>
<span id="cb80-46"><a href="#cb80-46" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb80-47"><a href="#cb80-47" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb80-48"><a href="#cb80-48" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb80-49"><a href="#cb80-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"--- PCA predictors not used. Skipping loading of future PCA rasters. ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb80-50"><a href="#cb80-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"    You might need to load your VIF-selected future environmental rasters here if needed for specific analyses.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb80-51"><a href="#cb80-51" aria-hidden="true" tabindex="-1"></a>  env_pca_future_list <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="co"># Ensure this object exists as NULL</span></span>
<span id="cb80-52"><a href="#cb80-52" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Loading Future Environmental PCA Rasters ---
  Loading PCA for future scenario: ssp119_2050 
    Loaded and processed PCA for ssp119_2050 . Layers: PC1, PC2, PC3, PC4 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-env-pca-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Loading PCA for future scenario: ssp119_2100 
    Loaded and processed PCA for ssp119_2100 . Layers: PC1, PC2, PC3, PC4 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-env-pca-2.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Loading PCA for future scenario: ssp585_2050 
    Loaded and processed PCA for ssp585_2050 . Layers: PC1, PC2, PC3, PC4 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-env-pca-3.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Loading PCA for future scenario: ssp585_2100 
    Loaded and processed PCA for ssp585_2100 . Layers: PC1, PC2, PC3, PC4 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-env-pca-4.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/load-future-env-pca-5.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The following objects are now available for subsequent analyses:</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="co"># - env_pca_current_cropped: SpatRaster of current PCA env predictors</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co"># - host_richness_future_list: List of SpatRasters for host future richness, named by scenario</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a><span class="co"># - fish_richness_future_list: List of SpatRasters for anemonefish (env-only) future richness, named by scenario</span></span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a><span class="co"># - env_pca_future_list: List of SpatRasters for future PCA env predictors, named by scenario</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="environmental-shifts-pca-components" class="level2">
<h2 class="anchored" data-anchor-id="environmental-shifts-pca-components">Environmental Shifts (PCA Components)</h2>
<p>This section examines the projected changes in the principal environmental gradients (PCA components) between the current conditions and future climate scenarios. This helps understand how the fundamental environmental space is predicted to change.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span>config<span class="sc">$</span>use_pca_predictors) {</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"PCA predictors not used in this configuration. Skipping PCA environmental shift analysis.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.null</span>(env_pca_current_cropped) <span class="sc">||</span> <span class="fu">is.null</span>(env_pca_future_list) <span class="sc">||</span> <span class="fu">length</span>(env_pca_future_list) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Warning: Current or future PCA environmental rasters are not available. Cannot calculate shifts.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"--- Calculating and Plotting Shifts in PCA Environmental Gradients ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Ensure current PCA stack is SpatRaster for subtraction</span></span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>    current_pca_terra <span class="ot">&lt;-</span> <span class="cf">if</span>(<span class="fu">inherits</span>(env_pca_current_cropped, <span class="st">"SpatRaster"</span>)) env_pca_current_cropped <span class="cf">else</span> terra<span class="sc">::</span><span class="fu">rast</span>(env_pca_current_cropped)</span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (scenario_name <span class="cf">in</span> <span class="fu">names</span>(env_pca_future_list)) {</span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Processing shifts for scenario:"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a>      future_pca_stack <span class="ot">&lt;-</span> env_pca_future_list[[scenario_name]]</span>
<span id="cb86-15"><a href="#cb86-15" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb86-16"><a href="#cb86-16" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">is.null</span>(future_pca_stack)) {</span>
<span id="cb86-17"><a href="#cb86-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"  Skipping scenario"</span>, scenario_name, <span class="st">"- future PCA stack is NULL.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb86-18"><a href="#cb86-18" aria-hidden="true" tabindex="-1"></a>        <span class="cf">next</span></span>
<span id="cb86-19"><a href="#cb86-19" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb86-20"><a href="#cb86-20" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb86-21"><a href="#cb86-21" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Ensure layers match for subtraction (e.g., up to n_pca_components)</span></span>
<span id="cb86-22"><a href="#cb86-22" aria-hidden="true" tabindex="-1"></a>      <span class="co"># It's crucial that both current and future PCA stacks were generated with the same number of components</span></span>
<span id="cb86-23"><a href="#cb86-23" aria-hidden="true" tabindex="-1"></a>      <span class="co"># and represent the same underlying variables in the same order before PCA.</span></span>
<span id="cb86-24"><a href="#cb86-24" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb86-25"><a href="#cb86-25" aria-hidden="true" tabindex="-1"></a>      <span class="co"># If layer names don't match exactly but number of layers does for the first N components:</span></span>
<span id="cb86-26"><a href="#cb86-26" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (terra<span class="sc">::</span><span class="fu">nlyr</span>(current_pca_terra) <span class="sc">&gt;=</span> config<span class="sc">$</span>n_pca_components <span class="sc">&amp;&amp;</span> terra<span class="sc">::</span><span class="fu">nlyr</span>(future_pca_stack) <span class="sc">&gt;=</span> config<span class="sc">$</span>n_pca_components) {</span>
<span id="cb86-27"><a href="#cb86-27" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb86-28"><a href="#cb86-28" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Select the first N components (defined in config)</span></span>
<span id="cb86-29"><a href="#cb86-29" aria-hidden="true" tabindex="-1"></a>        current_pca_subset <span class="ot">&lt;-</span> current_pca_terra[[<span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components]]</span>
<span id="cb86-30"><a href="#cb86-30" aria-hidden="true" tabindex="-1"></a>        future_pca_subset <span class="ot">&lt;-</span> future_pca_stack[[<span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components]]</span>
<span id="cb86-31"><a href="#cb86-31" aria-hidden="true" tabindex="-1"></a>        <span class="fu">names</span>(current_pca_subset) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components) <span class="co"># Standardize names for safety</span></span>
<span id="cb86-32"><a href="#cb86-32" aria-hidden="true" tabindex="-1"></a>        <span class="fu">names</span>(future_pca_subset) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components)</span>
<span id="cb86-33"><a href="#cb86-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-34"><a href="#cb86-34" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate the difference (shift)</span></span>
<span id="cb86-35"><a href="#cb86-35" aria-hidden="true" tabindex="-1"></a>        <span class="co"># terra::`-.SpatRaster` works element-wise if names match or by layer order if names don't.</span></span>
<span id="cb86-36"><a href="#cb86-36" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Better to ensure consistent naming (done above)</span></span>
<span id="cb86-37"><a href="#cb86-37" aria-hidden="true" tabindex="-1"></a>        env_shift_scenario <span class="ot">&lt;-</span> future_pca_subset <span class="sc">-</span> current_pca_subset </span>
<span id="cb86-38"><a href="#cb86-38" aria-hidden="true" tabindex="-1"></a>        <span class="fu">names</span>(env_shift_scenario) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Shift_PC"</span>, <span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components)</span>
<span id="cb86-39"><a href="#cb86-39" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb86-40"><a href="#cb86-40" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"    Calculated PCA shifts for scenario:"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb86-41"><a href="#cb86-41" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plot</span>(env_shift_scenario, <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"Environmental PCA Shifts (Future - Current):"</span>, scenario_name), <span class="at">nc =</span> <span class="dv">2</span>)</span>
<span id="cb86-42"><a href="#cb86-42" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb86-43"><a href="#cb86-43" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb86-44"><a href="#cb86-44" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"  Warning: Layer mismatch or insufficient layers for PCA shift calculation in scenario"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb86-45"><a href="#cb86-45" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"    Current PCA layers:"</span>, terra<span class="sc">::</span><span class="fu">nlyr</span>(current_pca_terra), <span class="st">" Future PCA layers:"</span>, terra<span class="sc">::</span><span class="fu">nlyr</span>(future_pca_stack), <span class="st">" Needed:"</span>, config<span class="sc">$</span>n_pca_components, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb86-46"><a href="#cb86-46" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb86-47"><a href="#cb86-47" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb86-48"><a href="#cb86-48" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb86-49"><a href="#cb86-49" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Calculating and Plotting Shifts in PCA Environmental Gradients ---
  Processing shifts for scenario: ssp119_2050 
    Calculated PCA shifts for scenario: ssp119_2050 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/plot-environmental-pca-shifts-1.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing shifts for scenario: ssp119_2100 
    Calculated PCA shifts for scenario: ssp119_2100 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/plot-environmental-pca-shifts-2.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing shifts for scenario: ssp585_2050 
    Calculated PCA shifts for scenario: ssp585_2050 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/plot-environmental-pca-shifts-3.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing shifts for scenario: ssp585_2100 
    Calculated PCA shifts for scenario: ssp585_2100 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/plot-environmental-pca-shifts-4.png" class="img-fluid figure-img" width="1152"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="load-future-individual-species-prediction-rasters" class="level2">
<h2 class="anchored" data-anchor-id="load-future-individual-species-prediction-rasters">Load Future Individual Species Prediction Rasters</h2>
<p>This section loads the individual species prediction rasters for host sea anemones and anemonefish (environmental-only models) for each future scenario. These are needed for calculating Schoenerâ€™s D overlap and individual suitability shifts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># future_scenarios_to_load was defined in the "Future Species Richness Projections" chunk</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"future_scenarios_to_load"</span>)) {</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  future_scenarios_to_load <span class="ot">&lt;-</span> config<span class="sc">$</span>env_scenarios[config<span class="sc">$</span>env_scenarios <span class="sc">!=</span> <span class="st">"current"</span>]</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a><span class="co"># predictor_suffix was defined in the "Drivers of Richness" chunk</span></span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"predictor_suffix"</span>)) {</span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>  predictor_suffix <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(config<span class="sc">$</span>use_pca_predictors, <span class="st">"_pca"</span>, <span class="st">"_vif"</span>)</span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb91-10"><a href="#cb91-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-11"><a href="#cb91-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-12"><a href="#cb91-12" aria-hidden="true" tabindex="-1"></a>host_pred_stacks_future_individual <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb91-13"><a href="#cb91-13" aria-hidden="true" tabindex="-1"></a>fish_pred_stacks_future_individual <span class="ot">&lt;-</span> <span class="fu">list</span>() <span class="co"># For env-only fish models</span></span>
<span id="cb91-14"><a href="#cb91-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-15"><a href="#cb91-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (scenario_name <span class="cf">in</span> future_scenarios_to_load) {</span>
<span id="cb91-16"><a href="#cb91-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Loading individual future predictions for Scenario:"</span>, scenario_name, <span class="st">"---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb91-17"><a href="#cb91-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb91-18"><a href="#cb91-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Host Anemones ---</span></span>
<span id="cb91-19"><a href="#cb91-19" aria-hidden="true" tabindex="-1"></a>  current_host_future_files <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb91-20"><a href="#cb91-20" aria-hidden="true" tabindex="-1"></a>  current_host_future_names <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb91-21"><a href="#cb91-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">exists</span>(<span class="st">"anemone_species_df"</span>) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(anemone_species_df) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb91-22"><a href="#cb91-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(anemone_species_df)) {</span>
<span id="cb91-23"><a href="#cb91-23" aria-hidden="true" tabindex="-1"></a>      sp_name_sanitized <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">" "</span>, <span class="st">"_"</span>, anemone_species_df<span class="sc">$</span>scientificName[i])</span>
<span id="cb91-24"><a href="#cb91-24" aria-hidden="true" tabindex="-1"></a>      pred_file_path <span class="ot">&lt;-</span> <span class="fu">construct_prediction_filename</span>(</span>
<span id="cb91-25"><a href="#cb91-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">species_name_sanitized =</span> sp_name_sanitized,</span>
<span id="cb91-26"><a href="#cb91-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">scenario_name =</span> scenario_name,</span>
<span id="cb91-27"><a href="#cb91-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">predictor_type_suffix =</span> predictor_suffix,</span>
<span id="cb91-28"><a href="#cb91-28" aria-hidden="true" tabindex="-1"></a>        <span class="at">config =</span> config</span>
<span id="cb91-29"><a href="#cb91-29" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb91-30"><a href="#cb91-30" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">file.exists</span>(pred_file_path)) {</span>
<span id="cb91-31"><a href="#cb91-31" aria-hidden="true" tabindex="-1"></a>        current_host_future_files <span class="ot">&lt;-</span> <span class="fu">c</span>(current_host_future_files, pred_file_path)</span>
<span id="cb91-32"><a href="#cb91-32" aria-hidden="true" tabindex="-1"></a>        current_host_future_names <span class="ot">&lt;-</span> <span class="fu">c</span>(current_host_future_names, sp_name_sanitized)</span>
<span id="cb91-33"><a href="#cb91-33" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb91-34"><a href="#cb91-34" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"  Warning: Host future prediction file not found for"</span>, sp_name_sanitized, <span class="st">"in"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb91-35"><a href="#cb91-35" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb91-36"><a href="#cb91-36" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb91-37"><a href="#cb91-37" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(current_host_future_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb91-38"><a href="#cb91-38" aria-hidden="true" tabindex="-1"></a>      temp_stack <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">rast</span>(current_host_future_files)</span>
<span id="cb91-39"><a href="#cb91-39" aria-hidden="true" tabindex="-1"></a>      <span class="fu">names</span>(temp_stack) <span class="ot">&lt;-</span> current_host_future_names</span>
<span id="cb91-40"><a href="#cb91-40" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (config<span class="sc">$</span>apply_indo_pacific_crop) {</span>
<span id="cb91-41"><a href="#cb91-41" aria-hidden="true" tabindex="-1"></a>        ip_extent <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">ext</span>(config<span class="sc">$</span>indo_pacific_bbox)</span>
<span id="cb91-42"><a href="#cb91-42" aria-hidden="true" tabindex="-1"></a>        host_pred_stacks_future_individual[[scenario_name]] <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">crop</span>(temp_stack, ip_extent)</span>
<span id="cb91-43"><a href="#cb91-43" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb91-44"><a href="#cb91-44" aria-hidden="true" tabindex="-1"></a>        host_pred_stacks_future_individual[[scenario_name]] <span class="ot">&lt;-</span> temp_stack</span>
<span id="cb91-45"><a href="#cb91-45" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb91-46"><a href="#cb91-46" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Loaded and cropped"</span>, terra<span class="sc">::</span><span class="fu">nlyr</span>(host_pred_stacks_future_individual[[scenario_name]]), <span class="st">"host future predictions for"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb91-47"><a href="#cb91-47" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb91-48"><a href="#cb91-48" aria-hidden="true" tabindex="-1"></a>       host_pred_stacks_future_individual[[scenario_name]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb91-49"><a href="#cb91-49" aria-hidden="true" tabindex="-1"></a>       <span class="fu">cat</span>(<span class="st">"  No host future prediction files found for"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb91-50"><a href="#cb91-50" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb91-51"><a href="#cb91-51" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb91-52"><a href="#cb91-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-53"><a href="#cb91-53" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Anemonefish (Environmental-Only) ---</span></span>
<span id="cb91-54"><a href="#cb91-54" aria-hidden="true" tabindex="-1"></a>  current_fish_future_files <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb91-55"><a href="#cb91-55" aria-hidden="true" tabindex="-1"></a>  current_fish_future_names <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb91-56"><a href="#cb91-56" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">exists</span>(<span class="st">"anemonefish_species_df"</span>) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(anemonefish_species_df) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb91-57"><a href="#cb91-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(anemonefish_species_df)) {</span>
<span id="cb91-58"><a href="#cb91-58" aria-hidden="true" tabindex="-1"></a>      sp_name_sanitized <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">" "</span>, <span class="st">"_"</span>, anemonefish_species_df<span class="sc">$</span>scientificName[i])</span>
<span id="cb91-59"><a href="#cb91-59" aria-hidden="true" tabindex="-1"></a>      pred_file_path <span class="ot">&lt;-</span> <span class="fu">construct_prediction_filename</span>(</span>
<span id="cb91-60"><a href="#cb91-60" aria-hidden="true" tabindex="-1"></a>        <span class="at">species_name_sanitized =</span> sp_name_sanitized,</span>
<span id="cb91-61"><a href="#cb91-61" aria-hidden="true" tabindex="-1"></a>        <span class="at">scenario_name =</span> scenario_name,</span>
<span id="cb91-62"><a href="#cb91-62" aria-hidden="true" tabindex="-1"></a>        <span class="at">predictor_type_suffix =</span> predictor_suffix, </span>
<span id="cb91-63"><a href="#cb91-63" aria-hidden="true" tabindex="-1"></a>        <span class="at">config =</span> config</span>
<span id="cb91-64"><a href="#cb91-64" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb91-65"><a href="#cb91-65" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">file.exists</span>(pred_file_path)) {</span>
<span id="cb91-66"><a href="#cb91-66" aria-hidden="true" tabindex="-1"></a>        current_fish_future_files <span class="ot">&lt;-</span> <span class="fu">c</span>(current_fish_future_files, pred_file_path)</span>
<span id="cb91-67"><a href="#cb91-67" aria-hidden="true" tabindex="-1"></a>        current_fish_future_names <span class="ot">&lt;-</span> <span class="fu">c</span>(current_fish_future_names, sp_name_sanitized)</span>
<span id="cb91-68"><a href="#cb91-68" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb91-69"><a href="#cb91-69" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"  Warning: Anemonefish (env-only) future prediction file not found for"</span>, sp_name_sanitized, <span class="st">"in"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb91-70"><a href="#cb91-70" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb91-71"><a href="#cb91-71" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb91-72"><a href="#cb91-72" aria-hidden="true" tabindex="-1"></a>     <span class="cf">if</span> (<span class="fu">length</span>(current_fish_future_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb91-73"><a href="#cb91-73" aria-hidden="true" tabindex="-1"></a>      temp_stack_fish <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">rast</span>(current_fish_future_files)</span>
<span id="cb91-74"><a href="#cb91-74" aria-hidden="true" tabindex="-1"></a>      <span class="fu">names</span>(temp_stack_fish) <span class="ot">&lt;-</span> current_fish_future_names</span>
<span id="cb91-75"><a href="#cb91-75" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (config<span class="sc">$</span>apply_indo_pacific_crop) {</span>
<span id="cb91-76"><a href="#cb91-76" aria-hidden="true" tabindex="-1"></a>        ip_extent <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">ext</span>(config<span class="sc">$</span>indo_pacific_bbox)</span>
<span id="cb91-77"><a href="#cb91-77" aria-hidden="true" tabindex="-1"></a>        fish_pred_stacks_future_individual[[scenario_name]] <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">crop</span>(temp_stack_fish, ip_extent)</span>
<span id="cb91-78"><a href="#cb91-78" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb91-79"><a href="#cb91-79" aria-hidden="true" tabindex="-1"></a>        fish_pred_stacks_future_individual[[scenario_name]] <span class="ot">&lt;-</span> temp_stack_fish</span>
<span id="cb91-80"><a href="#cb91-80" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb91-81"><a href="#cb91-81" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Loaded and cropped"</span>, terra<span class="sc">::</span><span class="fu">nlyr</span>(fish_pred_stacks_future_individual[[scenario_name]]), <span class="st">"anemonefish (env-only) future predictions for"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb91-82"><a href="#cb91-82" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb91-83"><a href="#cb91-83" aria-hidden="true" tabindex="-1"></a>       fish_pred_stacks_future_individual[[scenario_name]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb91-84"><a href="#cb91-84" aria-hidden="true" tabindex="-1"></a>       <span class="fu">cat</span>(<span class="st">"  No anemonefish (env-only) future prediction files found for"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb91-85"><a href="#cb91-85" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb91-86"><a href="#cb91-86" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb91-87"><a href="#cb91-87" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Loading individual future predictions for Scenario: ssp119_2050 ---
  Warning: Host future prediction file not found for Radianthus_aurora in ssp119_2050 

|---------|---------|---------|---------|
=========================================
                                          
  Loaded and cropped 9 host future predictions for ssp119_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_akindynos in ssp119_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_allardi in ssp119_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_bicinctus in ssp119_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_chrysopterus in ssp119_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_melanopus in ssp119_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_percula in ssp119_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Premnas_biaculeatus in ssp119_2050 
  Loaded and cropped 5 anemonefish (env-only) future predictions for ssp119_2050 

--- Loading individual future predictions for Scenario: ssp119_2100 ---
  Warning: Host future prediction file not found for Radianthus_aurora in ssp119_2100 

|---------|---------|---------|---------|
=========================================
                                          
  Loaded and cropped 9 host future predictions for ssp119_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_akindynos in ssp119_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_allardi in ssp119_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_bicinctus in ssp119_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_chrysopterus in ssp119_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_melanopus in ssp119_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_percula in ssp119_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Premnas_biaculeatus in ssp119_2100 
  Loaded and cropped 5 anemonefish (env-only) future predictions for ssp119_2100 

--- Loading individual future predictions for Scenario: ssp585_2050 ---
  Warning: Host future prediction file not found for Radianthus_aurora in ssp585_2050 

|---------|---------|---------|---------|
=========================================
                                          
  Loaded and cropped 9 host future predictions for ssp585_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_akindynos in ssp585_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_allardi in ssp585_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_bicinctus in ssp585_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_chrysopterus in ssp585_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_melanopus in ssp585_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_percula in ssp585_2050 
  Warning: Anemonefish (env-only) future prediction file not found for Premnas_biaculeatus in ssp585_2050 
  Loaded and cropped 5 anemonefish (env-only) future predictions for ssp585_2050 

--- Loading individual future predictions for Scenario: ssp585_2100 ---
  Warning: Host future prediction file not found for Radianthus_aurora in ssp585_2100 

|---------|---------|---------|---------|
=========================================
                                          
  Loaded and cropped 9 host future predictions for ssp585_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_akindynos in ssp585_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_allardi in ssp585_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_bicinctus in ssp585_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_chrysopterus in ssp585_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_melanopus in ssp585_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Amphiprion_percula in ssp585_2100 
  Warning: Anemonefish (env-only) future prediction file not found for Premnas_biaculeatus in ssp585_2100 
  Loaded and cropped 5 anemonefish (env-only) future predictions for ssp585_2100 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Result: host_pred_stacks_future_individual and fish_pred_stacks_future_individual</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="co"># are lists of SpatRasters, named by scenario, containing individual species layers.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="individual-species-suitability-shifts" class="level2">
<h2 class="anchored" data-anchor-id="individual-species-suitability-shifts">Individual Species Suitability Shifts</h2>
<p>This section calculates the mean change in environmental suitability for each individual host sea anemone and anemonefish species between the current period and each future climate scenario. These shift values will be used to explore correlations with niche breadth.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prerequisites from previous chunks:</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="co"># - host_pred_stack_cropped: Current individual host predictions (cropped)</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a><span class="co"># - fish_pred_stack_cropped: Current individual fish (env-only) predictions (cropped)</span></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a><span class="co"># - host_pred_stacks_future_individual: List of future individual host predictions (cropped), named by scenario</span></span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a><span class="co"># - fish_pred_stacks_future_individual: List of future individual fish (env-only) predictions (cropped), named by scenario</span></span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a><span class="co"># - future_scenarios_to_load: Vector of future scenario names (e.g., "ssp119_2050", "ssp585_2100")</span></span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"host_pred_stack_cropped"</span>) <span class="sc">||</span> <span class="sc">!</span><span class="fu">exists</span>(<span class="st">"fish_pred_stack_cropped"</span>) <span class="sc">||</span></span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a>    <span class="sc">!</span><span class="fu">exists</span>(<span class="st">"host_pred_stacks_future_individual"</span>) <span class="sc">||</span> <span class="sc">!</span><span class="fu">exists</span>(<span class="st">"fish_pred_stacks_future_individual"</span>) <span class="sc">||</span></span>
<span id="cb94-10"><a href="#cb94-10" aria-hidden="true" tabindex="-1"></a>    <span class="sc">!</span><span class="fu">exists</span>(<span class="st">"future_scenarios_to_load"</span>)) {</span>
<span id="cb94-11"><a href="#cb94-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"Prerequisite data for calculating suitability shifts is missing. Ensure previous chunks (loading current and future individual predictions) have run successfully."</span>)</span>
<span id="cb94-12"><a href="#cb94-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb94-13"><a href="#cb94-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-14"><a href="#cb94-14" aria-hidden="true" tabindex="-1"></a>suitability_shifts_list <span class="ot">&lt;-</span> <span class="fu">list</span>() <span class="co"># Initialize list to store results for all species and scenarios</span></span>
<span id="cb94-15"><a href="#cb94-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-16"><a href="#cb94-16" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"--- Calculating Mean Suitability Shifts for Individual Species ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Calculating Mean Suitability Shifts for Individual Species ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (scenario_name_fut <span class="cf">in</span> future_scenarios_to_load) {</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">  Processing Scenario:"</span>, scenario_name_fut, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Host Anemones Suitability Shifts ---</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"    Calculating shifts for Host Anemones...</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>  current_hosts_for_diff <span class="ot">&lt;-</span> host_pred_stack_cropped</span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a>  future_hosts_for_diff_raw <span class="ot">&lt;-</span> host_pred_stacks_future_individual[[scenario_name_fut]]</span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb96-9"><a href="#cb96-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(current_hosts_for_diff) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(future_hosts_for_diff_raw) <span class="sc">&amp;&amp;</span> </span>
<span id="cb96-10"><a href="#cb96-10" aria-hidden="true" tabindex="-1"></a>      terra<span class="sc">::</span><span class="fu">nlyr</span>(current_hosts_for_diff) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> terra<span class="sc">::</span><span class="fu">nlyr</span>(future_hosts_for_diff_raw) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb96-11"><a href="#cb96-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb96-12"><a href="#cb96-12" aria-hidden="true" tabindex="-1"></a>    common_host_species <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">names</span>(current_hosts_for_diff), <span class="fu">names</span>(future_hosts_for_diff_raw))</span>
<span id="cb96-13"><a href="#cb96-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb96-14"><a href="#cb96-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(common_host_species) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb96-15"><a href="#cb96-15" aria-hidden="true" tabindex="-1"></a>      current_hosts_subset <span class="ot">&lt;-</span> current_hosts_for_diff[[common_host_species]]</span>
<span id="cb96-16"><a href="#cb96-16" aria-hidden="true" tabindex="-1"></a>      future_hosts_subset_raw <span class="ot">&lt;-</span> future_hosts_for_diff_raw[[common_host_species]]</span>
<span id="cb96-17"><a href="#cb96-17" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb96-18"><a href="#cb96-18" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Ensure geometry matches for subtraction by resampling future to current if needed</span></span>
<span id="cb96-19"><a href="#cb96-19" aria-hidden="true" tabindex="-1"></a>      future_hosts_subset_aligned <span class="ot">&lt;-</span> future_hosts_subset_raw</span>
<span id="cb96-20"><a href="#cb96-20" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="sc">!</span>terra<span class="sc">::</span><span class="fu">compareGeom</span>(current_hosts_subset, future_hosts_subset_raw, <span class="at">stopOnError=</span><span class="cn">FALSE</span>, <span class="at">res=</span><span class="cn">TRUE</span>, <span class="at">crs=</span><span class="cn">TRUE</span>, <span class="at">ext=</span><span class="cn">TRUE</span>)){</span>
<span id="cb96-21"><a href="#cb96-21" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"      Resampling future host predictions for scenario '"</span>, scenario_name_fut, <span class="st">"' to match current geometry for shift calculation.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-22"><a href="#cb96-22" aria-hidden="true" tabindex="-1"></a>          future_hosts_subset_aligned <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>(</span>
<span id="cb96-23"><a href="#cb96-23" aria-hidden="true" tabindex="-1"></a>            terra<span class="sc">::</span><span class="fu">resample</span>(future_hosts_subset_raw, current_hosts_subset, <span class="at">method=</span><span class="st">"bilinear"</span>),</span>
<span id="cb96-24"><a href="#cb96-24" aria-hidden="true" tabindex="-1"></a>            <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb96-25"><a href="#cb96-25" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"      ERROR resampling future host stack for"</span>, scenario_name_fut, <span class="st">":"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span></span>
<span id="cb96-26"><a href="#cb96-26" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb96-27"><a href="#cb96-27" aria-hidden="true" tabindex="-1"></a>          )</span>
<span id="cb96-28"><a href="#cb96-28" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb96-29"><a href="#cb96-29" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb96-30"><a href="#cb96-30" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(future_hosts_subset_aligned)) {</span>
<span id="cb96-31"><a href="#cb96-31" aria-hidden="true" tabindex="-1"></a>        diff_hosts_scenario_indiv <span class="ot">&lt;-</span> future_hosts_subset_aligned <span class="sc">-</span> current_hosts_subset</span>
<span id="cb96-32"><a href="#cb96-32" aria-hidden="true" tabindex="-1"></a>        mean_shifts_hosts_indiv <span class="ot">&lt;-</span> <span class="fu">global</span>(diff_hosts_scenario_indiv, <span class="st">"mean"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb96-33"><a href="#cb96-33" aria-hidden="true" tabindex="-1"></a>        <span class="fu">rownames</span>(mean_shifts_hosts_indiv) <span class="ot">&lt;-</span> <span class="fu">names</span>(diff_hosts_scenario_indiv) <span class="co"># Should be common_host_species</span></span>
<span id="cb96-34"><a href="#cb96-34" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb96-35"><a href="#cb96-35" aria-hidden="true" tabindex="-1"></a>        suitability_shifts_list[[<span class="fu">paste0</span>(<span class="st">"hosts_"</span>, scenario_name_fut)]] <span class="ot">&lt;-</span> mean_shifts_hosts_indiv</span>
<span id="cb96-36"><a href="#cb96-36" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"      Mean suitability shifts for hosts in"</span>, scenario_name_fut, <span class="st">"calculated.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-37"><a href="#cb96-37" aria-hidden="true" tabindex="-1"></a>        <span class="co"># print(head(mean_shifts_hosts_indiv)) </span></span>
<span id="cb96-38"><a href="#cb96-38" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb96-39"><a href="#cb96-39" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb96-40"><a href="#cb96-40" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"      No common host species found between current and future stacks for scenario"</span>, scenario_name_fut, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-41"><a href="#cb96-41" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb96-42"><a href="#cb96-42" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb96-43"><a href="#cb96-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"    Skipping host anemones for scenario"</span>, scenario_name_fut, <span class="st">"- current or future individual prediction stack missing or empty.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-44"><a href="#cb96-44" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb96-45"><a href="#cb96-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-46"><a href="#cb96-46" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Anemonefish (Environmental-Only Models) Suitability Shifts ---</span></span>
<span id="cb96-47"><a href="#cb96-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"    Calculating shifts for Anemonefish (Env-Only)...</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-48"><a href="#cb96-48" aria-hidden="true" tabindex="-1"></a>  current_fish_for_diff <span class="ot">&lt;-</span> fish_pred_stack_cropped </span>
<span id="cb96-49"><a href="#cb96-49" aria-hidden="true" tabindex="-1"></a>  future_fish_for_diff_raw <span class="ot">&lt;-</span> fish_pred_stacks_future_individual[[scenario_name_fut]] </span>
<span id="cb96-50"><a href="#cb96-50" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb96-51"><a href="#cb96-51" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(current_fish_for_diff) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(future_fish_for_diff_raw) <span class="sc">&amp;&amp;</span></span>
<span id="cb96-52"><a href="#cb96-52" aria-hidden="true" tabindex="-1"></a>      terra<span class="sc">::</span><span class="fu">nlyr</span>(current_fish_for_diff) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> terra<span class="sc">::</span><span class="fu">nlyr</span>(future_fish_for_diff_raw) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb96-53"><a href="#cb96-53" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb96-54"><a href="#cb96-54" aria-hidden="true" tabindex="-1"></a>    common_fish_species <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">names</span>(current_fish_for_diff), <span class="fu">names</span>(future_fish_for_diff_raw))</span>
<span id="cb96-55"><a href="#cb96-55" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb96-56"><a href="#cb96-56" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(common_fish_species) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb96-57"><a href="#cb96-57" aria-hidden="true" tabindex="-1"></a>      current_fish_subset <span class="ot">&lt;-</span> current_fish_for_diff[[common_fish_species]]</span>
<span id="cb96-58"><a href="#cb96-58" aria-hidden="true" tabindex="-1"></a>      future_fish_subset_raw <span class="ot">&lt;-</span> future_fish_for_diff_raw[[common_fish_species]]</span>
<span id="cb96-59"><a href="#cb96-59" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb96-60"><a href="#cb96-60" aria-hidden="true" tabindex="-1"></a>      future_fish_subset_aligned <span class="ot">&lt;-</span> future_fish_subset_raw</span>
<span id="cb96-61"><a href="#cb96-61" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="sc">!</span>terra<span class="sc">::</span><span class="fu">compareGeom</span>(current_fish_subset, future_fish_subset_raw, <span class="at">stopOnError=</span><span class="cn">FALSE</span>, <span class="at">res=</span><span class="cn">TRUE</span>, <span class="at">crs=</span><span class="cn">TRUE</span>, <span class="at">ext=</span><span class="cn">TRUE</span>)){</span>
<span id="cb96-62"><a href="#cb96-62" aria-hidden="true" tabindex="-1"></a>         <span class="fu">cat</span>(<span class="st">"      Resampling future fish (env-only) predictions for scenario '"</span>, scenario_name_fut, <span class="st">"' to match current geometry.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-63"><a href="#cb96-63" aria-hidden="true" tabindex="-1"></a>         future_fish_subset_aligned <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>(</span>
<span id="cb96-64"><a href="#cb96-64" aria-hidden="true" tabindex="-1"></a>            terra<span class="sc">::</span><span class="fu">resample</span>(future_fish_subset_raw, current_fish_subset, <span class="at">method=</span><span class="st">"bilinear"</span>),</span>
<span id="cb96-65"><a href="#cb96-65" aria-hidden="true" tabindex="-1"></a>            <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb96-66"><a href="#cb96-66" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"      ERROR resampling future fish stack for"</span>, scenario_name_fut, <span class="st">":"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span></span>
<span id="cb96-67"><a href="#cb96-67" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb96-68"><a href="#cb96-68" aria-hidden="true" tabindex="-1"></a>          )</span>
<span id="cb96-69"><a href="#cb96-69" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb96-70"><a href="#cb96-70" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb96-71"><a href="#cb96-71" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(future_fish_subset_aligned)) {</span>
<span id="cb96-72"><a href="#cb96-72" aria-hidden="true" tabindex="-1"></a>        diff_fish_scenario_indiv <span class="ot">&lt;-</span> future_fish_subset_aligned <span class="sc">-</span> current_fish_subset</span>
<span id="cb96-73"><a href="#cb96-73" aria-hidden="true" tabindex="-1"></a>        mean_shifts_fish_indiv <span class="ot">&lt;-</span> <span class="fu">global</span>(diff_fish_scenario_indiv, <span class="st">"mean"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb96-74"><a href="#cb96-74" aria-hidden="true" tabindex="-1"></a>        <span class="fu">rownames</span>(mean_shifts_fish_indiv) <span class="ot">&lt;-</span> <span class="fu">names</span>(diff_fish_scenario_indiv)</span>
<span id="cb96-75"><a href="#cb96-75" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb96-76"><a href="#cb96-76" aria-hidden="true" tabindex="-1"></a>        suitability_shifts_list[[<span class="fu">paste0</span>(<span class="st">"fish_env_only_"</span>, scenario_name_fut)]] <span class="ot">&lt;-</span> mean_shifts_fish_indiv</span>
<span id="cb96-77"><a href="#cb96-77" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"      Mean suitability shifts for anemonefish (env-only) in"</span>, scenario_name_fut, <span class="st">"calculated.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-78"><a href="#cb96-78" aria-hidden="true" tabindex="-1"></a>        <span class="co"># print(head(mean_shifts_fish_indiv)) </span></span>
<span id="cb96-79"><a href="#cb96-79" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb96-80"><a href="#cb96-80" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb96-81"><a href="#cb96-81" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"      No common anemonefish species found between current and future stacks for scenario"</span>, scenario_name_fut, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-82"><a href="#cb96-82" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb96-83"><a href="#cb96-83" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb96-84"><a href="#cb96-84" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"    Skipping anemonefish (env-only) for scenario"</span>, scenario_name_fut, <span class="st">"- current or future individual prediction stack missing or empty.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb96-85"><a href="#cb96-85" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb96-86"><a href="#cb96-86" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb96-87"><a href="#cb96-87" aria-hidden="true" tabindex="-1"></a>} <span class="co"># End loop through future scenarios</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  Processing Scenario: ssp119_2050 
    Calculating shifts for Host Anemones...

|---------|---------|---------|---------|
=========================================
                                          
      Mean suitability shifts for hosts in ssp119_2050 calculated.
    Calculating shifts for Anemonefish (Env-Only)...
      Mean suitability shifts for anemonefish (env-only) in ssp119_2050 calculated.

  Processing Scenario: ssp119_2100 
    Calculating shifts for Host Anemones...

|---------|---------|---------|---------|
=========================================
                                          
      Mean suitability shifts for hosts in ssp119_2100 calculated.
    Calculating shifts for Anemonefish (Env-Only)...
      Mean suitability shifts for anemonefish (env-only) in ssp119_2100 calculated.

  Processing Scenario: ssp585_2050 
    Calculating shifts for Host Anemones...

|---------|---------|---------|---------|
=========================================
                                          
      Mean suitability shifts for hosts in ssp585_2050 calculated.
    Calculating shifts for Anemonefish (Env-Only)...
      Mean suitability shifts for anemonefish (env-only) in ssp585_2050 calculated.

  Processing Scenario: ssp585_2100 
    Calculating shifts for Host Anemones...

|---------|---------|---------|---------|
=========================================
                                          
      Mean suitability shifts for hosts in ssp585_2100 calculated.
    Calculating shifts for Anemonefish (Env-Only)...
      Mean suitability shifts for anemonefish (env-only) in ssp585_2100 calculated.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Individual species suitability shift calculations finished. ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Individual species suitability shift calculations finished. ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The `suitability_shifts_list` object is now populated.</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Example access: suitability_shifts_list[["hosts_ssp119_2050"]]</span></span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="co"># It will contain data frames with rownames = species_name_sanitized and a column "mean" for the shift.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="relationship-between-host-and-anemonefish-richness-current" class="level2">
<h2 class="anchored" data-anchor-id="relationship-between-host-and-anemonefish-richness-current">Relationship between Host and Anemonefish Richness (Current)</h2>
<p>This section explores the relationship between the current predicted richness of host sea anemones and their mutualistic anemonefish (based on environmental-only models). We use variance partitioning to determine the independent and shared contributions of host richness and environmental factors (PCA components) to anemonefish richness.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan) <span class="co"># For varpart</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">is.null</span>(host_richness_sum_cropped) <span class="sc">||</span> <span class="fu">is.null</span>(fish_richness_sum_cropped) <span class="sc">||</span> </span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>    (config<span class="sc">$</span>use_pca_predictors <span class="sc">&amp;&amp;</span> <span class="fu">is.null</span>(env_pca_current_cropped))) {</span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: Necessary raster data (host richness, fish richness, or current PCA env) is missing. Skipping variance partitioning.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"--- Preparing Data for Variance Partitioning ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure all layers have the same extent and resolution for stacking.</span></span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># The previous cropping step should have handled this for richness sums.</span></span>
<span id="cb101-11"><a href="#cb101-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># If PCA was used, env_pca_current_cropped should also match.</span></span>
<span id="cb101-12"><a href="#cb101-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb101-13"><a href="#cb101-13" aria-hidden="true" tabindex="-1"></a>  layers_for_sampling <span class="ot">&lt;-</span> <span class="fu">c</span>(host_richness_sum_cropped, fish_richness_sum_cropped)</span>
<span id="cb101-14"><a href="#cb101-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb101-15"><a href="#cb101-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (config<span class="sc">$</span>use_pca_predictors <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(env_pca_current_cropped)) {</span>
<span id="cb101-16"><a href="#cb101-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Ensure the env_pca_current_cropped is compatible. Resample if necessary.</span></span>
<span id="cb101-17"><a href="#cb101-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span>terra<span class="sc">::</span><span class="fu">compareGeom</span>(layers_for_sampling[[<span class="dv">1</span>]], env_pca_current_cropped[[<span class="dv">1</span>]], <span class="at">stopOnError=</span><span class="cn">FALSE</span>, <span class="at">res=</span><span class="cn">TRUE</span>, <span class="at">crs=</span><span class="cn">TRUE</span>)) {</span>
<span id="cb101-18"><a href="#cb101-18" aria-hidden="true" tabindex="-1"></a>       <span class="fu">cat</span>(<span class="st">"  Resampling current PCA environmental layers to match richness map geometry for sampling...</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-19"><a href="#cb101-19" aria-hidden="true" tabindex="-1"></a>       env_pca_current_resampled <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb101-20"><a href="#cb101-20" aria-hidden="true" tabindex="-1"></a>         terra<span class="sc">::</span><span class="fu">resample</span>(env_pca_current_cropped, layers_for_sampling[[<span class="dv">1</span>]], <span class="at">method=</span><span class="st">"bilinear"</span>)</span>
<span id="cb101-21"><a href="#cb101-21" aria-hidden="true" tabindex="-1"></a>       }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb101-22"><a href="#cb101-22" aria-hidden="true" tabindex="-1"></a>         <span class="fu">cat</span>(<span class="st">"  Warning: Failed to resample PCA env layers:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span></span>
<span id="cb101-23"><a href="#cb101-23" aria-hidden="true" tabindex="-1"></a>       })</span>
<span id="cb101-24"><a href="#cb101-24" aria-hidden="true" tabindex="-1"></a>       <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(env_pca_current_resampled)) {</span>
<span id="cb101-25"><a href="#cb101-25" aria-hidden="true" tabindex="-1"></a>         env_pca_current_for_sampling <span class="ot">&lt;-</span> env_pca_current_resampled</span>
<span id="cb101-26"><a href="#cb101-26" aria-hidden="true" tabindex="-1"></a>       } <span class="cf">else</span> {</span>
<span id="cb101-27"><a href="#cb101-27" aria-hidden="true" tabindex="-1"></a>         <span class="fu">cat</span>(<span class="st">"  Using un-resampled PCA env. May lead to issues if extents differ significantly.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-28"><a href="#cb101-28" aria-hidden="true" tabindex="-1"></a>         env_pca_current_for_sampling <span class="ot">&lt;-</span> env_pca_current_cropped</span>
<span id="cb101-29"><a href="#cb101-29" aria-hidden="true" tabindex="-1"></a>       }</span>
<span id="cb101-30"><a href="#cb101-30" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb101-31"><a href="#cb101-31" aria-hidden="true" tabindex="-1"></a>      env_pca_current_for_sampling <span class="ot">&lt;-</span> env_pca_current_cropped</span>
<span id="cb101-32"><a href="#cb101-32" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb101-33"><a href="#cb101-33" aria-hidden="true" tabindex="-1"></a>    layers_for_sampling <span class="ot">&lt;-</span> <span class="fu">c</span>(layers_for_sampling, env_pca_current_for_sampling[[<span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components]]) <span class="co"># Use first N components</span></span>
<span id="cb101-34"><a href="#cb101-34" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (<span class="sc">!</span>config<span class="sc">$</span>use_pca_predictors) {</span>
<span id="cb101-35"><a href="#cb101-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># </span><span class="al">TODO</span><span class="co">: If not using PCA, you'd need to load your VIF-selected CURRENT environmental rasters here,</span></span>
<span id="cb101-36"><a href="#cb101-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ensure they are cropped and resampled like the richness maps, and add them to layers_for_sampling.</span></span>
<span id="cb101-37"><a href="#cb101-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Warning: Variance partitioning without PCA environmental layers is not fully implemented here. Please adapt if using VIF-selected vars.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-38"><a href="#cb101-38" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb101-39"><a href="#cb101-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-40"><a href="#cb101-40" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if we have enough layers to proceed</span></span>
<span id="cb101-41"><a href="#cb101-41" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (terra<span class="sc">::</span><span class="fu">nlyr</span>(layers_for_sampling) <span class="sc">&lt;</span> <span class="dv">3</span>) {</span>
<span id="cb101-42"><a href="#cb101-42" aria-hidden="true" tabindex="-1"></a>     <span class="fu">cat</span>(<span class="st">"Error: Need at least host richness, fish richness, and one environmental layer for variance partitioning. Found:"</span>, terra<span class="sc">::</span><span class="fu">nlyr</span>(layers_for_sampling), <span class="st">"layers.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-43"><a href="#cb101-43" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb101-44"><a href="#cb101-44" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb101-45"><a href="#cb101-45" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Rename layers for clarity in the output dataframe</span></span>
<span id="cb101-46"><a href="#cb101-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(layers_for_sampling)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"HostRichness"</span></span>
<span id="cb101-47"><a href="#cb101-47" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(layers_for_sampling)[<span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">"FishRichness_EnvOnly"</span></span>
<span id="cb101-48"><a href="#cb101-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (config<span class="sc">$</span>use_pca_predictors <span class="sc">&amp;&amp;</span> terra<span class="sc">::</span><span class="fu">nlyr</span>(layers_for_sampling) <span class="sc">&gt;</span> <span class="dv">2</span>) {</span>
<span id="cb101-49"><a href="#cb101-49" aria-hidden="true" tabindex="-1"></a>      <span class="fu">names</span>(layers_for_sampling)[<span class="dv">3</span><span class="sc">:</span>(<span class="dv">2</span><span class="sc">+</span>config<span class="sc">$</span>n_pca_components)] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components)</span>
<span id="cb101-50"><a href="#cb101-50" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb101-51"><a href="#cb101-51" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb101-52"><a href="#cb101-52" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Stack for sampling has layers:"</span>, <span class="fu">paste</span>(<span class="fu">names</span>(layers_for_sampling), <span class="at">collapse=</span><span class="st">", "</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-53"><a href="#cb101-53" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb101-54"><a href="#cb101-54" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Sample random points</span></span>
<span id="cb101-55"><a href="#cb101-55" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Using dismo::randomPoints for consistency with ant paper if preferred, or terra::spatSample</span></span>
<span id="cb101-56"><a href="#cb101-56" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># for reproducibility</span></span>
<span id="cb101-57"><a href="#cb101-57" aria-hidden="true" tabindex="-1"></a>    <span class="co"># bg_points_sampled_varpart &lt;- dismo::randomPoints(raster::raster(layers_for_sampling[[1]]), 5000) # dismo needs raster object</span></span>
<span id="cb101-58"><a href="#cb101-58" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Note: dismo::randomPoints might struggle with large SpatRasters or specific extents.</span></span>
<span id="cb101-59"><a href="#cb101-59" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Using terra::spatSample might be more robust here with SpatRasters.</span></span>
<span id="cb101-60"><a href="#cb101-60" aria-hidden="true" tabindex="-1"></a>    <span class="co"># We need to sample from areas where ALL layers have data.</span></span>
<span id="cb101-61"><a href="#cb101-61" aria-hidden="true" tabindex="-1"></a>    <span class="co"># One way is to sample from the first layer, then extract from the stack.</span></span>
<span id="cb101-62"><a href="#cb101-62" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Or, create a mask where all layers are non-NA.</span></span>
<span id="cb101-63"><a href="#cb101-63" aria-hidden="true" tabindex="-1"></a>    mask_all_valid <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(layers_for_sampling)) <span class="sc">==</span> <span class="fu">nlyr</span>(layers_for_sampling)</span>
<span id="cb101-64"><a href="#cb101-64" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">global</span>(mask_all_valid, <span class="st">"sum"</span>, <span class="at">na.rm=</span>T)<span class="sc">$</span>sum <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb101-65"><a href="#cb101-65" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Error: No cells where all predictor layers for variance partitioning have valid data.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-66"><a href="#cb101-66" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb101-67"><a href="#cb101-67" aria-hidden="true" tabindex="-1"></a>      bg_points_sampled_varpart <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">spatSample</span>(mask_all_valid, <span class="dv">5000</span>, <span class="at">method =</span> <span class="st">"random"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>, <span class="at">xy=</span><span class="cn">TRUE</span>, <span class="at">warn=</span><span class="cn">FALSE</span>)</span>
<span id="cb101-68"><a href="#cb101-68" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Sampled"</span>, <span class="fu">nrow</span>(bg_points_sampled_varpart), <span class="st">"points for variance partitioning.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-69"><a href="#cb101-69" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb101-70"><a href="#cb101-70" aria-hidden="true" tabindex="-1"></a>      df_sampled_values <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">extract</span>(layers_for_sampling, bg_points_sampled_varpart[, <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>)], <span class="at">ID =</span> <span class="cn">FALSE</span>)</span>
<span id="cb101-71"><a href="#cb101-71" aria-hidden="true" tabindex="-1"></a>      df_sampled_values <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(df_sampled_values)</span>
<span id="cb101-72"><a href="#cb101-72" aria-hidden="true" tabindex="-1"></a>      df_sampled_values <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(df_sampled_values) <span class="co"># Remove any rows with NAs that might have slipped through</span></span>
<span id="cb101-73"><a href="#cb101-73" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb101-74"><a href="#cb101-74" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Using"</span>, <span class="fu">nrow</span>(df_sampled_values), <span class="st">"complete cases for variance partitioning.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-75"><a href="#cb101-75" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb101-76"><a href="#cb101-76" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="fu">nrow</span>(df_sampled_values) <span class="sc">&gt;</span> <span class="dv">10</span> <span class="sc">&amp;&amp;</span> <span class="st">"HostRichness"</span> <span class="sc">%in%</span> <span class="fu">names</span>(df_sampled_values) <span class="sc">&amp;&amp;</span> <span class="st">"FishRichness_EnvOnly"</span> <span class="sc">%in%</span> <span class="fu">names</span>(df_sampled_values) <span class="sc">&amp;&amp;</span> </span>
<span id="cb101-77"><a href="#cb101-77" aria-hidden="true" tabindex="-1"></a>          ( (<span class="sc">!</span>config<span class="sc">$</span>use_pca_predictors <span class="sc">&amp;&amp;</span> <span class="fu">ncol</span>(df_sampled_values) <span class="sc">&gt;</span> <span class="dv">2</span>) <span class="sc">||</span> (config<span class="sc">$</span>use_pca_predictors <span class="sc">&amp;&amp;</span> <span class="fu">all</span>(<span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components) <span class="sc">%in%</span> <span class="fu">names</span>(df_sampled_values))) )</span>
<span id="cb101-78"><a href="#cb101-78" aria-hidden="true" tabindex="-1"></a>         ) {</span>
<span id="cb101-79"><a href="#cb101-79" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb101-80"><a href="#cb101-80" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Prepare data for varpart</span></span>
<span id="cb101-81"><a href="#cb101-81" aria-hidden="true" tabindex="-1"></a>        fish_richness_response <span class="ot">&lt;-</span> df_sampled_values<span class="sc">$</span>FishRichness_EnvOnly</span>
<span id="cb101-82"><a href="#cb101-82" aria-hidden="true" tabindex="-1"></a>        host_richness_predictor <span class="ot">&lt;-</span> df_sampled_values<span class="sc">$</span>HostRichness</span>
<span id="cb101-83"><a href="#cb101-83" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb101-84"><a href="#cb101-84" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (config<span class="sc">$</span>use_pca_predictors) {</span>
<span id="cb101-85"><a href="#cb101-85" aria-hidden="true" tabindex="-1"></a>          env_predictors_df <span class="ot">&lt;-</span> df_sampled_values[, <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span>config<span class="sc">$</span>n_pca_components), drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb101-86"><a href="#cb101-86" aria-hidden="true" tabindex="-1"></a>        } <span class="cf">else</span> {</span>
<span id="cb101-87"><a href="#cb101-87" aria-hidden="true" tabindex="-1"></a>          <span class="co"># If using VIF, select the environmental columns (excluding HostRichness and FishRichness_EnvOnly)</span></span>
<span id="cb101-88"><a href="#cb101-88" aria-hidden="true" tabindex="-1"></a>          env_col_names <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(<span class="fu">names</span>(df_sampled_values), <span class="fu">c</span>(<span class="st">"HostRichness"</span>, <span class="st">"FishRichness_EnvOnly"</span>))</span>
<span id="cb101-89"><a href="#cb101-89" aria-hidden="true" tabindex="-1"></a>          <span class="cf">if</span>(<span class="fu">length</span>(env_col_names) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb101-90"><a href="#cb101-90" aria-hidden="true" tabindex="-1"></a>            env_predictors_df <span class="ot">&lt;-</span> df_sampled_values[, env_col_names, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb101-91"><a href="#cb101-91" aria-hidden="true" tabindex="-1"></a>          } <span class="cf">else</span> {</span>
<span id="cb101-92"><a href="#cb101-92" aria-hidden="true" tabindex="-1"></a>            env_predictors_df <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb101-93"><a href="#cb101-93" aria-hidden="true" tabindex="-1"></a>            <span class="fu">cat</span>(<span class="st">"Warning: No environmental predictor columns identified for VIF-based variance partitioning.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-94"><a href="#cb101-94" aria-hidden="true" tabindex="-1"></a>          }</span>
<span id="cb101-95"><a href="#cb101-95" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb101-96"><a href="#cb101-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-97"><a href="#cb101-97" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(env_predictors_df) <span class="sc">&amp;&amp;</span> <span class="fu">ncol</span>(env_predictors_df) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb101-98"><a href="#cb101-98" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Perform variance partitioning</span></span>
<span id="cb101-99"><a href="#cb101-99" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Response: FishRichness_EnvOnly</span></span>
<span id="cb101-100"><a href="#cb101-100" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Predictor Group 1: HostRichness</span></span>
<span id="cb101-101"><a href="#cb101-101" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Predictor Group 2: Environmental variables (PCA components or VIF-selected)</span></span>
<span id="cb101-102"><a href="#cb101-102" aria-hidden="true" tabindex="-1"></a>          varpart_result <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">varpart</span>(fish_richness_response, host_richness_predictor, env_predictors_df)</span>
<span id="cb101-103"><a href="#cb101-103" aria-hidden="true" tabindex="-1"></a>          </span>
<span id="cb101-104"><a href="#cb101-104" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Variance Partitioning Results ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-105"><a href="#cb101-105" aria-hidden="true" tabindex="-1"></a>          <span class="fu">print</span>(varpart_result)</span>
<span id="cb101-106"><a href="#cb101-106" aria-hidden="true" tabindex="-1"></a>          <span class="fu">plot</span>(varpart_result, <span class="at">main =</span> <span class="st">"Variance Partitioning: Anemonefish Richness (Env-Only Models)"</span>)</span>
<span id="cb101-107"><a href="#cb101-107" aria-hidden="true" tabindex="-1"></a>          </span>
<span id="cb101-108"><a href="#cb101-108" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Pearson correlation between host and fish richness</span></span>
<span id="cb101-109"><a href="#cb101-109" aria-hidden="true" tabindex="-1"></a>          cor_test_result <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(df_sampled_values<span class="sc">$</span>HostRichness, df_sampled_values<span class="sc">$</span>FishRichness_EnvOnly)</span>
<span id="cb101-110"><a href="#cb101-110" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Pearson Correlation between Host Richness and Fish (Env-Only) Richness:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-111"><a href="#cb101-111" aria-hidden="true" tabindex="-1"></a>          <span class="fu">print</span>(cor_test_result)</span>
<span id="cb101-112"><a href="#cb101-112" aria-hidden="true" tabindex="-1"></a>          </span>
<span id="cb101-113"><a href="#cb101-113" aria-hidden="true" tabindex="-1"></a>        } <span class="cf">else</span> {</span>
<span id="cb101-114"><a href="#cb101-114" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"Warning: Environmental predictor data frame is empty. Skipping variance partitioning.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-115"><a href="#cb101-115" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb101-116"><a href="#cb101-116" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb101-117"><a href="#cb101-117" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"Warning: Not enough data or required columns missing for variance partitioning.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb101-118"><a href="#cb101-118" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb101-119"><a href="#cb101-119" aria-hidden="true" tabindex="-1"></a>    } <span class="co"># end else for mask_all_valid check</span></span>
<span id="cb101-120"><a href="#cb101-120" aria-hidden="true" tabindex="-1"></a>  } <span class="co"># end else for nlyr check</span></span>
<span id="cb101-121"><a href="#cb101-121" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Preparing Data for Variance Partitioning ---
Stack for sampling has layers: HostRichness, FishRichness_EnvOnly, PC1, PC2, PC3, PC4 

|---------|---------|---------|---------|
=========================================
                                          

|---------|---------|---------|---------|
=========================================
                                          
Sampled 5000 points for variance partitioning.
Using 3502 complete cases for variance partitioning.

--- Variance Partitioning Results ---

Partition of variance in RDA 

Call: vegan::varpart(Y = fish_richness_response, X =
host_richness_predictor, env_predictors_df)

Explanatory tables:
X1:  host_richness_predictor
X2:  env_predictors_df 

No. of explanatory tables: 2 
Total variation (SS): 2051.9 
            Variance: 0.58608 
No. of observations: 3502 

Partition table:
                     Df R.squared Adj.R.squared Testable
[a+c] = X1            1   0.92575       0.92573     TRUE
[b+c] = X2            4   0.70436       0.70402     TRUE
[a+b+c] = X1+X2       5   0.93457       0.93448     TRUE
Individual fractions                                    
[a] = X1|X2           1                 0.23046     TRUE
[b] = X2|X1           4                 0.00875     TRUE
[c]                   0                 0.69528    FALSE
[d] = Residuals                         0.06552    FALSE
---
Use function 'rda' to test significance of fractions of interest</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/richness-sampling-and-varpart-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Pearson Correlation between Host Richness and Fish (Env-Only) Richness:

    Pearson's product-moment correlation

data:  df_sampled_values$HostRichness and df_sampled_values$FishRichness_EnvOnly
t = 208.9, df = 3500, p-value &lt; 2.2e-16
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 0.9596199 0.9645434
sample estimates:
      cor 
0.9621601 </code></pre>
</div>
</div>
</section>
<section id="richness-change-analyses" class="level2">
<h2 class="anchored" data-anchor-id="richness-change-analyses">Richness Change Analyses</h2>
<p>This section analyzes the projected changes in host sea anemone and anemonefish (environmental-only models) richness under future climate scenarios.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glmmTMB)</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(emmeans) <span class="co"># For pairwise comparisons</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)   <span class="co"># For pivot_longer</span></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure current richness maps are loaded and cropped from the first chunk</span></span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">is.null</span>(host_richness_sum_cropped) <span class="sc">||</span> <span class="fu">is.null</span>(fish_richness_sum_cropped)) {</span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: Cropped current richness maps not available. Skipping richness change analysis.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb104-10"><a href="#cb104-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Host Anemone Richness Change ---</span></span>
<span id="cb104-11"><a href="#cb104-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Analyzing Host Sea Anemone Richness Changes ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-12"><a href="#cb104-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(host_richness_future_list) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb104-13"><a href="#cb104-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Stack current and all future host richness maps</span></span>
<span id="cb104-14"><a href="#cb104-14" aria-hidden="true" tabindex="-1"></a>    all_host_rasters_for_analysis <span class="ot">&lt;-</span> <span class="fu">c</span>(host_richness_sum_cropped) <span class="co"># Start with current</span></span>
<span id="cb104-15"><a href="#cb104-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(all_host_rasters_for_analysis)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"current"</span> <span class="co"># Explicitly name current</span></span>
<span id="cb104-16"><a href="#cb104-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-17"><a href="#cb104-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(scen_name <span class="cf">in</span> <span class="fu">names</span>(host_richness_future_list)) {</span>
<span id="cb104-18"><a href="#cb104-18" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(host_richness_future_list[[scen_name]])) {</span>
<span id="cb104-19"><a href="#cb104-19" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Ensure CRS and extent match before adding to stack. Resample if needed.</span></span>
<span id="cb104-20"><a href="#cb104-20" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span>(<span class="sc">!</span>terra<span class="sc">::</span><span class="fu">compareGeom</span>(host_richness_sum_cropped, host_richness_future_list[[scen_name]], <span class="at">stopOnError=</span><span class="cn">FALSE</span>, <span class="at">res=</span><span class="cn">TRUE</span>, <span class="at">crs=</span><span class="cn">TRUE</span>)){</span>
<span id="cb104-21"><a href="#cb104-21" aria-hidden="true" tabindex="-1"></a>            <span class="fu">cat</span>(<span class="st">"  Resampling host richness for scenario:"</span>, scen_name, <span class="st">"to match current.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-22"><a href="#cb104-22" aria-hidden="true" tabindex="-1"></a>            host_richness_future_list[[scen_name]] <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">resample</span>(host_richness_future_list[[scen_name]], host_richness_sum_cropped, <span class="at">method=</span><span class="st">"bilinear"</span>)</span>
<span id="cb104-23"><a href="#cb104-23" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb104-24"><a href="#cb104-24" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Ensure names are consistent, e.g., just the scenario name</span></span>
<span id="cb104-25"><a href="#cb104-25" aria-hidden="true" tabindex="-1"></a>        <span class="fu">names</span>(host_richness_future_list[[scen_name]]) <span class="ot">&lt;-</span> scen_name</span>
<span id="cb104-26"><a href="#cb104-26" aria-hidden="true" tabindex="-1"></a>        all_host_rasters_for_analysis <span class="ot">&lt;-</span> <span class="fu">c</span>(all_host_rasters_for_analysis, host_richness_future_list[[scen_name]])</span>
<span id="cb104-27"><a href="#cb104-27" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb104-28"><a href="#cb104-28" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-29"><a href="#cb104-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Final check after potential resampling</span></span>
<span id="cb104-30"><a href="#cb104-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (terra<span class="sc">::</span><span class="fu">nlyr</span>(all_host_rasters_for_analysis) <span class="sc">&lt;</span> <span class="dv">2</span>) {</span>
<span id="cb104-31"><a href="#cb104-31" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"  Warning: Not enough host richness rasters (current + at least one future) to analyze changes.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-32"><a href="#cb104-32" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb104-33"><a href="#cb104-33" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"  Host richness stack for change analysis has layers:"</span>, <span class="fu">paste</span>(<span class="fu">names</span>(all_host_rasters_for_analysis), <span class="at">collapse=</span><span class="st">", "</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-34"><a href="#cb104-34" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb104-35"><a href="#cb104-35" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Sample points and extract values</span></span>
<span id="cb104-36"><a href="#cb104-36" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set.seed</span>(<span class="dv">456</span>)</span>
<span id="cb104-37"><a href="#cb104-37" aria-hidden="true" tabindex="-1"></a>        mask_all_valid_host <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(all_host_rasters_for_analysis)) <span class="sc">==</span> <span class="fu">nlyr</span>(all_host_rasters_for_analysis)</span>
<span id="cb104-38"><a href="#cb104-38" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span>(<span class="fu">global</span>(mask_all_valid_host, <span class="st">"sum"</span>, <span class="at">na.rm=</span>T)<span class="sc">$</span>sum <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb104-39"><a href="#cb104-39" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"  Error: No cells where all host richness layers have valid data for change analysis.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-40"><a href="#cb104-40" aria-hidden="true" tabindex="-1"></a>        } <span class="cf">else</span> {</span>
<span id="cb104-41"><a href="#cb104-41" aria-hidden="true" tabindex="-1"></a>          pts_host_change <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">spatSample</span>(mask_all_valid_host, <span class="dv">5000</span>, <span class="at">method =</span> <span class="st">"random"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>, <span class="at">xy=</span><span class="cn">TRUE</span>, <span class="at">warn=</span><span class="cn">FALSE</span>)</span>
<span id="cb104-42"><a href="#cb104-42" aria-hidden="true" tabindex="-1"></a>          df_host_change <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">extract</span>(all_host_rasters_for_analysis, pts_host_change[, <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>)], <span class="at">ID =</span> <span class="cn">FALSE</span>)</span>
<span id="cb104-43"><a href="#cb104-43" aria-hidden="true" tabindex="-1"></a>          df_host_change <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(df_host_change)</span>
<span id="cb104-44"><a href="#cb104-44" aria-hidden="true" tabindex="-1"></a>          df_host_change<span class="sc">$</span>pixel_id <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(df_host_change) <span class="co"># Add a pixel identifier</span></span>
<span id="cb104-45"><a href="#cb104-45" aria-hidden="true" tabindex="-1"></a>          </span>
<span id="cb104-46"><a href="#cb104-46" aria-hidden="true" tabindex="-1"></a>          df_host_change_long <span class="ot">&lt;-</span> df_host_change <span class="sc">%&gt;%</span></span>
<span id="cb104-47"><a href="#cb104-47" aria-hidden="true" tabindex="-1"></a>            <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>pixel_id, <span class="at">names_to =</span> <span class="st">"scenario"</span>, <span class="at">values_to =</span> <span class="st">"richness"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb104-48"><a href="#cb104-48" aria-hidden="true" tabindex="-1"></a>            <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(richness)) <span class="sc">%&gt;%</span></span>
<span id="cb104-49"><a href="#cb104-49" aria-hidden="true" tabindex="-1"></a>            <span class="fu">mutate</span>(<span class="at">scenario =</span> <span class="fu">factor</span>(scenario, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"current"</span>, future_scenarios_to_load))) <span class="co"># Ensure 'current' is reference</span></span>
<span id="cb104-50"><a href="#cb104-50" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb104-51"><a href="#cb104-51" aria-hidden="true" tabindex="-1"></a>          <span class="cf">if</span> (<span class="fu">nrow</span>(df_host_change_long) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb104-52"><a href="#cb104-52" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Fit GLMM</span></span>
<span id="cb104-53"><a href="#cb104-53" aria-hidden="true" tabindex="-1"></a>            lmm_host_richness <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb104-54"><a href="#cb104-54" aria-hidden="true" tabindex="-1"></a>              <span class="fu">glmmTMB</span>(richness <span class="sc">~</span> scenario <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>pixel_id), <span class="at">data =</span> df_host_change_long, <span class="at">family =</span> <span class="fu">gaussian</span>()) <span class="co"># Assuming richness can be modeled as Gaussian</span></span>
<span id="cb104-55"><a href="#cb104-55" aria-hidden="true" tabindex="-1"></a>            }, <span class="at">error =</span> <span class="cf">function</span>(e) {<span class="fu">cat</span>(<span class="st">"Error fitting GLMM for host richness:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span>})</span>
<span id="cb104-56"><a href="#cb104-56" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb104-57"><a href="#cb104-57" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(lmm_host_richness)) {</span>
<span id="cb104-58"><a href="#cb104-58" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Summary of GLMM for Host Anemone Richness Change:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-59"><a href="#cb104-59" aria-hidden="true" tabindex="-1"></a>              <span class="fu">print</span>(<span class="fu">summary</span>(lmm_host_richness))</span>
<span id="cb104-60"><a href="#cb104-60" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Anova for Host Anemone Richness Change:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-61"><a href="#cb104-61" aria-hidden="true" tabindex="-1"></a>              <span class="fu">print</span>(car<span class="sc">::</span><span class="fu">Anova</span>(lmm_host_richness))</span>
<span id="cb104-62"><a href="#cb104-62" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Estimated Marginal Means for Host Anemone Richness Change:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-63"><a href="#cb104-63" aria-hidden="true" tabindex="-1"></a>              <span class="fu">print</span>(<span class="fu">emmeans</span>(lmm_host_richness, pairwise <span class="sc">~</span> scenario))</span>
<span id="cb104-64"><a href="#cb104-64" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb104-65"><a href="#cb104-65" aria-hidden="true" tabindex="-1"></a>          } <span class="cf">else</span> { <span class="fu">cat</span>(<span class="st">"  No data left after pivoting for host richness LMM.</span><span class="sc">\n</span><span class="st">"</span>) }</span>
<span id="cb104-66"><a href="#cb104-66" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb104-67"><a href="#cb104-67" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-68"><a href="#cb104-68" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> { <span class="fu">cat</span>(<span class="st">"  No future host richness maps available for change analysis.</span><span class="sc">\n</span><span class="st">"</span>) }</span>
<span id="cb104-69"><a href="#cb104-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-70"><a href="#cb104-70" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Anemonefish (Env-Only) Richness Change ---</span></span>
<span id="cb104-71"><a href="#cb104-71" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Analyzing Anemonefish (Env-Only) Richness Changes ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-72"><a href="#cb104-72" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(fish_richness_future_list) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb104-73"><a href="#cb104-73" aria-hidden="true" tabindex="-1"></a>    all_fish_rasters_for_analysis <span class="ot">&lt;-</span> <span class="fu">c</span>(fish_richness_sum_cropped) <span class="co"># Start with current</span></span>
<span id="cb104-74"><a href="#cb104-74" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(all_fish_rasters_for_analysis)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"current"</span></span>
<span id="cb104-75"><a href="#cb104-75" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-76"><a href="#cb104-76" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(scen_name <span class="cf">in</span> <span class="fu">names</span>(fish_richness_future_list)) {</span>
<span id="cb104-77"><a href="#cb104-77" aria-hidden="true" tabindex="-1"></a>       <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(fish_richness_future_list[[scen_name]])) {</span>
<span id="cb104-78"><a href="#cb104-78" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span>(<span class="sc">!</span>terra<span class="sc">::</span><span class="fu">compareGeom</span>(fish_richness_sum_cropped, fish_richness_future_list[[scen_name]], <span class="at">stopOnError=</span><span class="cn">FALSE</span>, <span class="at">res=</span><span class="cn">TRUE</span>, <span class="at">crs=</span><span class="cn">TRUE</span>)){</span>
<span id="cb104-79"><a href="#cb104-79" aria-hidden="true" tabindex="-1"></a>            <span class="fu">cat</span>(<span class="st">"  Resampling fish richness for scenario:"</span>, scen_name, <span class="st">"to match current.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-80"><a href="#cb104-80" aria-hidden="true" tabindex="-1"></a>            fish_richness_future_list[[scen_name]] <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">resample</span>(fish_richness_future_list[[scen_name]], fish_richness_sum_cropped, <span class="at">method=</span><span class="st">"bilinear"</span>)</span>
<span id="cb104-81"><a href="#cb104-81" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb104-82"><a href="#cb104-82" aria-hidden="true" tabindex="-1"></a>        <span class="fu">names</span>(fish_richness_future_list[[scen_name]]) <span class="ot">&lt;-</span> scen_name</span>
<span id="cb104-83"><a href="#cb104-83" aria-hidden="true" tabindex="-1"></a>        all_fish_rasters_for_analysis <span class="ot">&lt;-</span> <span class="fu">c</span>(all_fish_rasters_for_analysis, fish_richness_future_list[[scen_name]])</span>
<span id="cb104-84"><a href="#cb104-84" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb104-85"><a href="#cb104-85" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-86"><a href="#cb104-86" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(terra<span class="sc">::</span><span class="fu">nlyr</span>(all_fish_rasters_for_analysis) <span class="sc">&lt;</span> <span class="dv">2</span>) {</span>
<span id="cb104-87"><a href="#cb104-87" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"  Warning: Not enough fish richness rasters (current + at least one future) to analyze changes.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-88"><a href="#cb104-88" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb104-89"><a href="#cb104-89" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"  Anemonefish richness stack for change analysis has layers:"</span>, <span class="fu">paste</span>(<span class="fu">names</span>(all_fish_rasters_for_analysis), <span class="at">collapse=</span><span class="st">", "</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-90"><a href="#cb104-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-91"><a href="#cb104-91" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set.seed</span>(<span class="dv">789</span>)</span>
<span id="cb104-92"><a href="#cb104-92" aria-hidden="true" tabindex="-1"></a>        mask_all_valid_fish <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(all_fish_rasters_for_analysis)) <span class="sc">==</span> <span class="fu">nlyr</span>(all_fish_rasters_for_analysis)</span>
<span id="cb104-93"><a href="#cb104-93" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span>(<span class="fu">global</span>(mask_all_valid_fish, <span class="st">"sum"</span>, <span class="at">na.rm=</span>T)<span class="sc">$</span>sum <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb104-94"><a href="#cb104-94" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"  Error: No cells where all fish richness layers have valid data for change analysis.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-95"><a href="#cb104-95" aria-hidden="true" tabindex="-1"></a>        } <span class="cf">else</span> {</span>
<span id="cb104-96"><a href="#cb104-96" aria-hidden="true" tabindex="-1"></a>          pts_fish_change <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">spatSample</span>(mask_all_valid_fish, <span class="dv">5000</span>, <span class="at">method =</span> <span class="st">"random"</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>, <span class="at">xy=</span><span class="cn">TRUE</span>, <span class="at">warn=</span><span class="cn">FALSE</span>)</span>
<span id="cb104-97"><a href="#cb104-97" aria-hidden="true" tabindex="-1"></a>          df_fish_change <span class="ot">&lt;-</span> terra<span class="sc">::</span><span class="fu">extract</span>(all_fish_rasters_for_analysis, pts_fish_change[, <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>)], <span class="at">ID =</span> <span class="cn">FALSE</span>)</span>
<span id="cb104-98"><a href="#cb104-98" aria-hidden="true" tabindex="-1"></a>          df_fish_change <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(df_fish_change)</span>
<span id="cb104-99"><a href="#cb104-99" aria-hidden="true" tabindex="-1"></a>          df_fish_change<span class="sc">$</span>pixel_id <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(df_fish_change)</span>
<span id="cb104-100"><a href="#cb104-100" aria-hidden="true" tabindex="-1"></a>          </span>
<span id="cb104-101"><a href="#cb104-101" aria-hidden="true" tabindex="-1"></a>          df_fish_change_long <span class="ot">&lt;-</span> df_fish_change <span class="sc">%&gt;%</span></span>
<span id="cb104-102"><a href="#cb104-102" aria-hidden="true" tabindex="-1"></a>            <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>pixel_id, <span class="at">names_to =</span> <span class="st">"scenario"</span>, <span class="at">values_to =</span> <span class="st">"richness"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb104-103"><a href="#cb104-103" aria-hidden="true" tabindex="-1"></a>            <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(richness)) <span class="sc">%&gt;%</span></span>
<span id="cb104-104"><a href="#cb104-104" aria-hidden="true" tabindex="-1"></a>            <span class="fu">mutate</span>(<span class="at">scenario =</span> <span class="fu">factor</span>(scenario, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"current"</span>, future_scenarios_to_load)))</span>
<span id="cb104-105"><a href="#cb104-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-106"><a href="#cb104-106" aria-hidden="true" tabindex="-1"></a>          <span class="cf">if</span> (<span class="fu">nrow</span>(df_fish_change_long) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb104-107"><a href="#cb104-107" aria-hidden="true" tabindex="-1"></a>            lmm_fish_richness <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb104-108"><a href="#cb104-108" aria-hidden="true" tabindex="-1"></a>              <span class="fu">glmmTMB</span>(richness <span class="sc">~</span> scenario <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>pixel_id), <span class="at">data =</span> df_fish_change_long, <span class="at">family =</span> <span class="fu">gaussian</span>())</span>
<span id="cb104-109"><a href="#cb104-109" aria-hidden="true" tabindex="-1"></a>            }, <span class="at">error =</span> <span class="cf">function</span>(e) {<span class="fu">cat</span>(<span class="st">"Error fitting GLMM for fish richness:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span>})</span>
<span id="cb104-110"><a href="#cb104-110" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb104-111"><a href="#cb104-111" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(lmm_fish_richness)) {</span>
<span id="cb104-112"><a href="#cb104-112" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Summary of GLMM for Anemonefish (Env-Only) Richness Change:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-113"><a href="#cb104-113" aria-hidden="true" tabindex="-1"></a>              <span class="fu">print</span>(<span class="fu">summary</span>(lmm_fish_richness))</span>
<span id="cb104-114"><a href="#cb104-114" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Anova for Anemonefish (Env-Only) Richness Change:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-115"><a href="#cb104-115" aria-hidden="true" tabindex="-1"></a>              <span class="fu">print</span>(car<span class="sc">::</span><span class="fu">Anova</span>(lmm_fish_richness))</span>
<span id="cb104-116"><a href="#cb104-116" aria-hidden="true" tabindex="-1"></a>              <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Estimated Marginal Means for Anemonefish (Env-Only) Richness Change:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-117"><a href="#cb104-117" aria-hidden="true" tabindex="-1"></a>              <span class="fu">print</span>(<span class="fu">emmeans</span>(lmm_fish_richness, pairwise <span class="sc">~</span> scenario))</span>
<span id="cb104-118"><a href="#cb104-118" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb104-119"><a href="#cb104-119" aria-hidden="true" tabindex="-1"></a>          } <span class="cf">else</span> { <span class="fu">cat</span>(<span class="st">"  No data left after pivoting for fish richness LMM.</span><span class="sc">\n</span><span class="st">"</span>) }</span>
<span id="cb104-120"><a href="#cb104-120" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb104-121"><a href="#cb104-121" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-122"><a href="#cb104-122" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> { <span class="fu">cat</span>(<span class="st">"  No future anemonefish (env-only) richness maps available for change analysis.</span><span class="sc">\n</span><span class="st">"</span>) }</span>
<span id="cb104-123"><a href="#cb104-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-124"><a href="#cb104-124" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Combined Plot (Optional, if both groups have data) ---</span></span>
<span id="cb104-125"><a href="#cb104-125" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">exists</span>(<span class="st">"df_host_change_long"</span>) <span class="sc">&amp;&amp;</span> <span class="fu">exists</span>(<span class="st">"df_fish_change_long"</span>) <span class="sc">&amp;&amp;</span> </span>
<span id="cb104-126"><a href="#cb104-126" aria-hidden="true" tabindex="-1"></a>      <span class="fu">nrow</span>(df_host_change_long) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(df_fish_change_long) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb104-127"><a href="#cb104-127" aria-hidden="true" tabindex="-1"></a>    df_host_change_long<span class="sc">$</span>guild <span class="ot">&lt;-</span> <span class="st">"Host_Anemones"</span></span>
<span id="cb104-128"><a href="#cb104-128" aria-hidden="true" tabindex="-1"></a>    df_fish_change_long<span class="sc">$</span>guild <span class="ot">&lt;-</span> <span class="st">"Anemonefish_EnvOnly"</span></span>
<span id="cb104-129"><a href="#cb104-129" aria-hidden="true" tabindex="-1"></a>    df_all_richness_change <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(df_host_change_long, df_fish_change_long)</span>
<span id="cb104-130"><a href="#cb104-130" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-131"><a href="#cb104-131" aria-hidden="true" tabindex="-1"></a>    plot_richness_boxplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_all_richness_change, <span class="fu">aes</span>(<span class="at">x =</span> guild, <span class="at">y =</span> richness, <span class="at">fill =</span> scenario)) <span class="sc">+</span></span>
<span id="cb104-132"><a href="#cb104-132" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb104-133"><a href="#cb104-133" aria-hidden="true" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Projected Richness Changes by Guild and Scenario"</span>,</span>
<span id="cb104-134"><a href="#cb104-134" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> <span class="st">"Guild"</span>, <span class="at">y =</span> <span class="st">"Predicted Richness"</span>) <span class="sc">+</span></span>
<span id="cb104-135"><a href="#cb104-135" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb104-136"><a href="#cb104-136" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span>
<span id="cb104-137"><a href="#cb104-137" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(plot_richness_boxplot)</span>
<span id="cb104-138"><a href="#cb104-138" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-139"><a href="#cb104-139" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Combined GLMM for interaction</span></span>
<span id="cb104-140"><a href="#cb104-140" aria-hidden="true" tabindex="-1"></a>    lmm_combined_richness <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb104-141"><a href="#cb104-141" aria-hidden="true" tabindex="-1"></a>      <span class="fu">glmmTMB</span>(richness <span class="sc">~</span> scenario <span class="sc">*</span> guild <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>pixel_id), <span class="at">data =</span> df_all_richness_change, <span class="at">family =</span> <span class="fu">gaussian</span>())</span>
<span id="cb104-142"><a href="#cb104-142" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e) {<span class="fu">cat</span>(<span class="st">"Error fitting combined GLMM for richness:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span>})</span>
<span id="cb104-143"><a href="#cb104-143" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(lmm_combined_richness)) {</span>
<span id="cb104-144"><a href="#cb104-144" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Summary of Combined GLMM for Richness (Scenario * Guild Interaction):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-145"><a href="#cb104-145" aria-hidden="true" tabindex="-1"></a>      <span class="fu">print</span>(<span class="fu">summary</span>(lmm_combined_richness))</span>
<span id="cb104-146"><a href="#cb104-146" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Anova for Combined GLMM Richness:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb104-147"><a href="#cb104-147" aria-hidden="true" tabindex="-1"></a>      <span class="fu">print</span>(car<span class="sc">::</span><span class="fu">Anova</span>(lmm_combined_richness))</span>
<span id="cb104-148"><a href="#cb104-148" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb104-149"><a href="#cb104-149" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb104-150"><a href="#cb104-150" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-151"><a href="#cb104-151" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Maps of Richness Loss/Gain for a specific future scenario (e.g., worst-case SSP5-8.5 2100) ---</span></span>
<span id="cb104-152"><a href="#cb104-152" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Choose one future scenario for difference maps</span></span>
<span id="cb104-153"><a href="#cb104-153" aria-hidden="true" tabindex="-1"></a>  diff_scenario_host <span class="ot">&lt;-</span> <span class="st">"ssp585_2100"</span> <span class="co"># Example</span></span>
<span id="cb104-154"><a href="#cb104-154" aria-hidden="true" tabindex="-1"></a>  diff_scenario_fish <span class="ot">&lt;-</span> <span class="st">"ssp585_2100"</span> <span class="co"># Example</span></span>
<span id="cb104-155"><a href="#cb104-155" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-156"><a href="#cb104-156" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (diff_scenario_host <span class="sc">%in%</span> <span class="fu">names</span>(host_richness_future_list) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(host_richness_future_list[[diff_scenario_host]])) {</span>
<span id="cb104-157"><a href="#cb104-157" aria-hidden="true" tabindex="-1"></a>    host_richness_loss <span class="ot">&lt;-</span> host_richness_future_list[[diff_scenario_host]] <span class="sc">-</span> host_richness_sum_cropped</span>
<span id="cb104-158"><a href="#cb104-158" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(host_richness_loss) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"HostRichnessChange_"</span>, diff_scenario_host)</span>
<span id="cb104-159"><a href="#cb104-159" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(host_richness_loss, <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"Host Anemone Richness Change:"</span>, diff_scenario_host, <span class="st">"vs Current"</span>))</span>
<span id="cb104-160"><a href="#cb104-160" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-161"><a href="#cb104-161" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ggplot version</span></span>
<span id="cb104-162"><a href="#cb104-162" aria-hidden="true" tabindex="-1"></a>    df_host_loss <span class="ot">=</span> <span class="fu">as.data.frame</span>(host_richness_loss, <span class="at">xy=</span><span class="cn">TRUE</span>)</span>
<span id="cb104-163"><a href="#cb104-163" aria-hidden="true" tabindex="-1"></a>    plot_host_loss_map <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_host_loss) <span class="sc">+</span></span>
<span id="cb104-164"><a href="#cb104-164" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">y=</span>y, <span class="at">x=</span>x, <span class="at">fill=</span>.data[[<span class="fu">names</span>(host_richness_loss)]])) <span class="sc">+</span> <span class="co"># Use .data pronoun</span></span>
<span id="cb104-165"><a href="#cb104-165" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"red"</span>, <span class="at">mid =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">"darkgreen"</span>, <span class="at">midpoint =</span> <span class="dv">0</span>, <span class="at">name=</span><span class="st">"Richness Change"</span>) <span class="sc">+</span></span>
<span id="cb104-166"><a href="#cb104-166" aria-hidden="true" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(<span class="st">"Host Anemone Richness Change:"</span>, diff_scenario_host, <span class="st">"vs Current"</span>), <span class="at">x=</span><span class="st">"Longitude"</span>, <span class="at">y=</span><span class="st">"Latitude"</span>) <span class="sc">+</span></span>
<span id="cb104-167"><a href="#cb104-167" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"aliceblue"</span>))</span>
<span id="cb104-168"><a href="#cb104-168" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(plot_host_loss_map)</span>
<span id="cb104-169"><a href="#cb104-169" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb104-170"><a href="#cb104-170" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb104-171"><a href="#cb104-171" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (diff_scenario_fish <span class="sc">%in%</span> <span class="fu">names</span>(fish_richness_future_list) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(fish_richness_future_list[[diff_scenario_fish]])) {</span>
<span id="cb104-172"><a href="#cb104-172" aria-hidden="true" tabindex="-1"></a>    fish_richness_loss <span class="ot">&lt;-</span> fish_richness_future_list[[diff_scenario_fish]] <span class="sc">-</span> fish_richness_sum_cropped</span>
<span id="cb104-173"><a href="#cb104-173" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(fish_richness_loss) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"FishRichnessChange_"</span>, diff_scenario_fish)</span>
<span id="cb104-174"><a href="#cb104-174" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(fish_richness_loss, <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"Anemonefish (Env-Only) Richness Change:"</span>, diff_scenario_fish, <span class="st">"vs Current"</span>))</span>
<span id="cb104-175"><a href="#cb104-175" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-176"><a href="#cb104-176" aria-hidden="true" tabindex="-1"></a>    df_fish_loss <span class="ot">=</span> <span class="fu">as.data.frame</span>(fish_richness_loss, <span class="at">xy=</span><span class="cn">TRUE</span>)</span>
<span id="cb104-177"><a href="#cb104-177" aria-hidden="true" tabindex="-1"></a>    plot_fish_loss_map <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_fish_loss) <span class="sc">+</span></span>
<span id="cb104-178"><a href="#cb104-178" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_raster</span>(<span class="fu">aes</span>(<span class="at">y=</span>y, <span class="at">x=</span>x, <span class="at">fill=</span>.data[[<span class="fu">names</span>(fish_richness_loss)]])) <span class="sc">+</span></span>
<span id="cb104-179"><a href="#cb104-179" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"red"</span>, <span class="at">mid =</span> <span class="st">"white"</span>, <span class="at">high =</span> <span class="st">"darkgreen"</span>, <span class="at">midpoint =</span> <span class="dv">0</span>, <span class="at">name=</span><span class="st">"Richness Change"</span>) <span class="sc">+</span></span>
<span id="cb104-180"><a href="#cb104-180" aria-hidden="true" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(<span class="st">"Anemonefish (Env-Only) Richness Change:"</span>, diff_scenario_fish, <span class="st">"vs Current"</span>), <span class="at">x=</span><span class="st">"Longitude"</span>, <span class="at">y=</span><span class="st">"Latitude"</span>) <span class="sc">+</span></span>
<span id="cb104-181"><a href="#cb104-181" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"aliceblue"</span>))</span>
<span id="cb104-182"><a href="#cb104-182" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(plot_fish_loss_map)</span>
<span id="cb104-183"><a href="#cb104-183" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb104-184"><a href="#cb104-184" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Analyzing Host Sea Anemone Richness Changes ---
  Host richness stack for change analysis has layers: current, ssp119_2050, ssp119_2100, ssp585_2050, ssp585_2100 

Summary of GLMM for Host Anemone Richness Change:
 Family: gaussian  ( identity )
Formula:          richness ~ scenario + (1 | pixel_id)
Data: df_host_change_long

     AIC      BIC   logLik deviance df.resid 
 -5263.3  -5208.9   2638.7  -5277.3    17713 

Random effects:

Conditional model:
 Groups   Name        Variance Std.Dev.
 pixel_id (Intercept) 1.35683  1.1648  
 Residual             0.01229  0.1109  
Number of obs: 17720, groups:  pixel_id, 3544

Dispersion estimate for gaussian family (sigma^2): 0.0123 

Conditional model:
                     Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)          0.364879   0.019655  18.564   &lt;2e-16 ***
scenariossp119_2050 -0.006567   0.002634  -2.493   0.0127 *  
scenariossp119_2100 -0.002391   0.002634  -0.908   0.3640    
scenariossp585_2050 -0.005977   0.002634  -2.269   0.0232 *  
scenariossp585_2100  0.024408   0.002634   9.267   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Anova for Host Anemone Richness Change:
Analysis of Deviance Table (Type II Wald chisquare tests)

Response: richness
          Chisq Df Pr(&gt;Chisq)    
scenario 190.95  4  &lt; 2.2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Estimated Marginal Means for Host Anemone Richness Change:
$emmeans
 scenario    emmean     SE    df lower.CL upper.CL
 current      0.365 0.0197 17713    0.326    0.403
 ssp119_2050  0.358 0.0197 17713    0.320    0.397
 ssp119_2100  0.362 0.0197 17713    0.324    0.401
 ssp585_2050  0.359 0.0197 17713    0.320    0.397
 ssp585_2100  0.389 0.0197 17713    0.351    0.428

Confidence level used: 0.95 

$contrasts
 contrast                  estimate      SE    df t.ratio p.value
 current - ssp119_2050      0.00657 0.00263 17713   2.493  0.0921
 current - ssp119_2100      0.00239 0.00263 17713   0.908  0.8940
 current - ssp585_2050      0.00598 0.00263 17713   2.269  0.1549
 current - ssp585_2100     -0.02441 0.00263 17713  -9.267  &lt;.0001
 ssp119_2050 - ssp119_2100 -0.00418 0.00263 17713  -1.586  0.5067
 ssp119_2050 - ssp585_2050 -0.00059 0.00263 17713  -0.224  0.9994
 ssp119_2050 - ssp585_2100 -0.03097 0.00263 17713 -11.760  &lt;.0001
 ssp119_2100 - ssp585_2050  0.00359 0.00263 17713   1.362  0.6524
 ssp119_2100 - ssp585_2100 -0.02680 0.00263 17713 -10.175  &lt;.0001
 ssp585_2050 - ssp585_2100 -0.03038 0.00263 17713 -11.536  &lt;.0001

P value adjustment: tukey method for comparing a family of 5 estimates 


--- Analyzing Anemonefish (Env-Only) Richness Changes ---
  Anemonefish richness stack for change analysis has layers: current, ssp119_2050, ssp119_2100, ssp585_2050, ssp585_2100 

Summary of GLMM for Anemonefish (Env-Only) Richness Change:
 Family: gaussian  ( identity )
Formula:          richness ~ scenario + (1 | pixel_id)
Data: df_fish_change_long

     AIC      BIC   logLik deviance df.resid 
-15054.3 -14999.8   7534.1 -15068.3    17603 

Random effects:

Conditional model:
 Groups   Name        Variance Std.Dev.
 pixel_id (Intercept) 0.458669 0.67725 
 Residual             0.008024 0.08958 
Number of obs: 17610, groups:  pixel_id, 3522

Dispersion estimate for gaussian family (sigma^2): 0.00802 

Conditional model:
                     Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)          0.245205   0.011511  21.301  &lt; 2e-16 ***
scenariossp119_2050 -0.013489   0.002135  -6.319 2.63e-10 ***
scenariossp119_2100 -0.010626   0.002135  -4.978 6.43e-07 ***
scenariossp585_2050 -0.013635   0.002135  -6.388 1.68e-10 ***
scenariossp585_2100 -0.001472   0.002135  -0.690     0.49    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Anova for Anemonefish (Env-Only) Richness Change:
Analysis of Deviance Table (Type II Wald chisquare tests)

Response: richness
          Chisq Df Pr(&gt;Chisq)    
scenario 76.935  4  7.764e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Estimated Marginal Means for Anemonefish (Env-Only) Richness Change:
$emmeans
 scenario    emmean     SE    df lower.CL upper.CL
 current      0.245 0.0115 17603    0.223    0.268
 ssp119_2050  0.232 0.0115 17603    0.209    0.254
 ssp119_2100  0.235 0.0115 17603    0.212    0.257
 ssp585_2050  0.232 0.0115 17603    0.209    0.254
 ssp585_2100  0.244 0.0115 17603    0.221    0.266

Confidence level used: 0.95 

$contrasts
 contrast                   estimate      SE    df t.ratio p.value
 current - ssp119_2050      0.013489 0.00213 17603   6.319  &lt;.0001
 current - ssp119_2100      0.010626 0.00213 17603   4.978  &lt;.0001
 current - ssp585_2050      0.013635 0.00213 17603   6.388  &lt;.0001
 current - ssp585_2100      0.001472 0.00213 17603   0.690  0.9588
 ssp119_2050 - ssp119_2100 -0.002864 0.00213 17603  -1.341  0.6652
 ssp119_2050 - ssp585_2050  0.000146 0.00213 17603   0.068  1.0000
 ssp119_2050 - ssp585_2100 -0.012017 0.00213 17603  -5.630  &lt;.0001
 ssp119_2100 - ssp585_2050  0.003009 0.00213 17603   1.410  0.6213
 ssp119_2100 - ssp585_2100 -0.009154 0.00213 17603  -4.288  0.0002
 ssp585_2050 - ssp585_2100 -0.012163 0.00213 17603  -5.698  &lt;.0001

P value adjustment: tukey method for comparing a family of 5 estimates </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/richness-change-analysis-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Summary of Combined GLMM for Richness (Scenario * Guild Interaction):
 Family: gaussian  ( identity )
Formula:          richness ~ scenario * guild + (1 | pixel_id)
Data: df_all_richness_change

     AIC      BIC   logLik deviance df.resid 
 77918.3  78020.0 -38947.2  77894.3    35318 

Random effects:

Conditional model:
 Groups   Name        Variance Std.Dev.
 pixel_id (Intercept) 0.5899   0.7681  
 Residual             0.3840   0.6197  
Number of obs: 35330, groups:  pixel_id, 4562

Dispersion estimate for gaussian family (sigma^2): 0.384 

Conditional model:
                                        Estimate Std. Error z value Pr(&gt;|z|)
(Intercept)                             0.239918   0.015602  15.378   &lt;2e-16
scenariossp119_2050                    -0.013498   0.014767  -0.914    0.361
scenariossp119_2100                    -0.010634   0.014767  -0.720    0.471
scenariossp585_2050                    -0.013640   0.014767  -0.924    0.356
scenariossp585_2100                    -0.001480   0.014767  -0.100    0.920
guildHost_Anemones                      0.125842   0.015247   8.253   &lt;2e-16
scenariossp119_2050:guildHost_Anemones  0.006937   0.020851   0.333    0.739
scenariossp119_2100:guildHost_Anemones  0.008248   0.020851   0.396    0.692
scenariossp585_2050:guildHost_Anemones  0.007667   0.020851   0.368    0.713
scenariossp585_2100:guildHost_Anemones  0.025893   0.020851   1.242    0.214
                                          
(Intercept)                            ***
scenariossp119_2050                       
scenariossp119_2100                       
scenariossp585_2050                       
scenariossp585_2100                       
guildHost_Anemones                     ***
scenariossp119_2050:guildHost_Anemones    
scenariossp119_2100:guildHost_Anemones    
scenariossp585_2050:guildHost_Anemones    
scenariossp585_2100:guildHost_Anemones    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Anova for Combined GLMM Richness:
Analysis of Deviance Table (Type II Wald chisquare tests)

Response: richness
                  Chisq Df Pr(&gt;Chisq)    
scenario         6.0190  4     0.1977    
guild          313.8914  1     &lt;2e-16 ***
scenario:guild   1.7029  4     0.7902    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/richness-change-analysis-2.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/richness-change-analysis-3.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/richness-change-analysis-4.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/richness-change-analysis-5.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="anemonefish-phylogeny" class="level2">
<h2 class="anchored" data-anchor-id="anemonefish-phylogeny">Anemonefish Phylogeny</h2>
<p>This section loads a phylogenetic tree that includes the anemonefish species under study. It then prunes the tree to only these species and standardizes tip labels for consistency in subsequent analyses.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phytools) </span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(picante)  </span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr) <span class="co"># Added for data frame manipulation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 1. Define Anemonefish Species in Your Study ---</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Using the scientific names directly from your provided list</span></span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a><span class="co"># We will convert them to Genus_species format for tree matching.</span></span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>anemonefish_species_in_study_raw <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Amphiprion clarkii"</span>, <span class="st">"Amphiprion frenatus"</span>, <span class="st">"Amphiprion ocellaris"</span>, </span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Amphiprion perideraion"</span>, <span class="st">"Amphiprion polymnus"</span>, <span class="st">"Amphiprion sandaracinos"</span></span>
<span id="cb108-7"><a href="#cb108-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add ALL your other anemonefish species from your full list here if they are not covered by df_fish_comparison yet</span></span>
<span id="cb108-8"><a href="#cb108-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb108-9"><a href="#cb108-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-10"><a href="#cb108-10" aria-hidden="true" tabindex="-1"></a><span class="co"># If df_fish_comparison is already loaded and processed, use it as the primary source</span></span>
<span id="cb108-11"><a href="#cb108-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">exists</span>(<span class="st">"df_fish_comparison"</span>) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(df_fish_comparison) <span class="sc">&amp;&amp;</span> <span class="st">"species"</span> <span class="sc">%in%</span> <span class="fu">names</span>(df_fish_comparison) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(df_fish_comparison) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb108-12"><a href="#cb108-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Assuming df_fish_comparison$species is already in "Genus_species" format from the previous chunk</span></span>
<span id="cb108-13"><a href="#cb108-13" aria-hidden="true" tabindex="-1"></a>  anemonefish_species_in_study_tree_format <span class="ot">&lt;-</span> <span class="fu">unique</span>(df_fish_comparison<span class="sc">$</span>species)</span>
<span id="cb108-14"><a href="#cb108-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Using anemonefish species list from 'df_fish_comparison'.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb108-15"><a href="#cb108-15" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb108-16"><a href="#cb108-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: 'df_fish_comparison' not fully available. Using manually defined list and converting to Genus_species format for tree matching.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb108-17"><a href="#cb108-17" aria-hidden="true" tabindex="-1"></a>  anemonefish_species_in_study_tree_format <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">" "</span>, <span class="st">"_"</span>, anemonefish_species_in_study_raw)</span>
<span id="cb108-18"><a href="#cb108-18" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Using anemonefish species list from 'df_fish_comparison'.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Anemonefish species for tree matching (Genus_species format):"</span>, <span class="fu">paste</span>(anemonefish_species_in_study_tree_format, <span class="at">collapse=</span><span class="st">", "</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anemonefish species for tree matching (Genus_species format): Amphiprion_akindynos, Amphiprion_allardi, Amphiprion_bicinctus, Amphiprion_chrysopterus, Amphiprion_clarkii, Amphiprion_melanopus, Amphiprion_ocellaris, Amphiprion_percula, Amphiprion_perideraion, Amphiprion_polymnus, Amphiprion_sandaracinos, Premnas_biaculeatus </code></pre>
</div>
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(anemonefish_species_in_study_tree_format) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"No anemonefish species defined for phylogeny processing."</span>)</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 2. Load Anemonefish Phylogenetic Tree ---</span></span>
<span id="cb112-7"><a href="#cb112-7" aria-hidden="true" tabindex="-1"></a><span class="co"># !!! --- YOU MUST REPLACE THIS WITH THE ACTUAL PATH TO *YOUR* ANEMONEFISH TREE FILE --- !!!</span></span>
<span id="cb112-8"><a href="#cb112-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Example: "anemonefish_from_fishtree.nwk" or "some_pomacentridae_study.nex"</span></span>
<span id="cb112-9"><a href="#cb112-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Place this file in data/phylogeny/</span></span>
<span id="cb112-10"><a href="#cb112-10" aria-hidden="true" tabindex="-1"></a>anemonefish_tree_file_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(config<span class="sc">$</span>data_dir, <span class="st">"phylogeny"</span>, <span class="st">"anemonefish_phylogeny_OTL.nwk"</span>) </span>
<span id="cb112-11"><a href="#cb112-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-12"><a href="#cb112-12" aria-hidden="true" tabindex="-1"></a>pruned_anemonefish_tree <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="co"># Initialize</span></span>
<span id="cb112-13"><a href="#cb112-13" aria-hidden="true" tabindex="-1"></a>fish_label_map_df <span class="ot">&lt;-</span> <span class="cn">NULL</span>       <span class="co"># Initialize</span></span>
<span id="cb112-14"><a href="#cb112-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-15"><a href="#cb112-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">file.exists</span>(anemonefish_tree_file_path)) {</span>
<span id="cb112-16"><a href="#cb112-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Loading anemonefish tree from:"</span>, anemonefish_tree_file_path, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-17"><a href="#cb112-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb112-18"><a href="#cb112-18" aria-hidden="true" tabindex="-1"></a>  full_anemonefish_tree <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb112-19"><a href="#cb112-19" aria-hidden="true" tabindex="-1"></a>    file_ext_lower <span class="ot">&lt;-</span> <span class="fu">tolower</span>(tools<span class="sc">::</span><span class="fu">file_ext</span>(anemonefish_tree_file_path))</span>
<span id="cb112-20"><a href="#cb112-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (file_ext_lower <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"nex"</span>, <span class="st">"nxs"</span>)) {</span>
<span id="cb112-21"><a href="#cb112-21" aria-hidden="true" tabindex="-1"></a>      ape<span class="sc">::</span><span class="fu">read.nexus</span>(anemonefish_tree_file_path)</span>
<span id="cb112-22"><a href="#cb112-22" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> <span class="cf">if</span> (file_ext_lower <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"nwk"</span>, <span class="st">"newick"</span>, <span class="st">"tre"</span>, <span class="st">"tree"</span>)) {</span>
<span id="cb112-23"><a href="#cb112-23" aria-hidden="true" tabindex="-1"></a>      ape<span class="sc">::</span><span class="fu">read.tree</span>(anemonefish_tree_file_path)</span>
<span id="cb112-24"><a href="#cb112-24" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb112-25"><a href="#cb112-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Warning: Unknown tree file extension for"</span>, <span class="fu">basename</span>(anemonefish_tree_file_path), <span class="st">". Attempting read.tree().</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-26"><a href="#cb112-26" aria-hidden="true" tabindex="-1"></a>      ape<span class="sc">::</span><span class="fu">read.tree</span>(anemonefish_tree_file_path) <span class="co"># Default attempt</span></span>
<span id="cb112-27"><a href="#cb112-27" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb112-28"><a href="#cb112-28" aria-hidden="true" tabindex="-1"></a>  }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb112-29"><a href="#cb112-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Error loading anemonefish tree:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span></span>
<span id="cb112-30"><a href="#cb112-30" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb112-31"><a href="#cb112-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb112-32"><a href="#cb112-32" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(full_anemonefish_tree) <span class="sc">&amp;&amp;</span> <span class="fu">inherits</span>(full_anemonefish_tree, <span class="st">"phylo"</span>)) {</span>
<span id="cb112-33"><a href="#cb112-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Full anemonefish tree loaded. Original tips:"</span>, <span class="fu">length</span>(full_anemonefish_tree<span class="sc">$</span>tip.label), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-34"><a href="#cb112-34" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plot(full_anemonefish_tree, cex = 0.5, main = "Full Loaded Anemonefish Tree (Initial)", type="fan", no.margin=TRUE)</span></span>
<span id="cb112-35"><a href="#cb112-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-36"><a href="#cb112-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># --- 2a. Standardize Tip Labels of the Loaded Tree (CRUCIAL) ---</span></span>
<span id="cb112-37"><a href="#cb112-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># This step is highly dependent on the format of YOUR tree file.</span></span>
<span id="cb112-38"><a href="#cb112-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Common operations:</span></span>
<span id="cb112-39"><a href="#cb112-39" aria-hidden="true" tabindex="-1"></a>    <span class="co">#   - Replace spaces with underscores: full_anemonefish_tree$tip.label &lt;- gsub(" ", "_", full_anemonefish_tree$tip.label)</span></span>
<span id="cb112-40"><a href="#cb112-40" aria-hidden="true" tabindex="-1"></a>    <span class="co">#   - Remove strain/accession numbers if present, e.g., using str_extract or sub.</span></span>
<span id="cb112-41"><a href="#cb112-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># For example, if tips are "Genus species strainXYZ", you might do:</span></span>
<span id="cb112-42"><a href="#cb112-42" aria-hidden="true" tabindex="-1"></a>    <span class="co">#   temp_labels &lt;- gsub(" ", "_", full_anemonefish_tree$tip.label) # First, underscores</span></span>
<span id="cb112-43"><a href="#cb112-43" aria-hidden="true" tabindex="-1"></a>    <span class="co">#   full_anemonefish_tree$tip.label &lt;- str_extract(temp_labels, "^[A-Za-z]+_[a-z]+") # Keep only Genus_species part</span></span>
<span id="cb112-44"><a href="#cb112-44" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Ensure the result is Genus_species</span></span>
<span id="cb112-45"><a href="#cb112-45" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb112-46"><a href="#cb112-46" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Example: (assuming your tree already uses Genus_species or you've cleaned it above)</span></span>
<span id="cb112-47"><a href="#cb112-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># full_anemonefish_tree$tip.label &lt;- gsub(" ", "_", full_anemonefish_tree$tip.label) # Basic cleaning</span></span>
<span id="cb112-48"><a href="#cb112-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-49"><a href="#cb112-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"First 10 tip labels from loaded tree (after potential cleaning):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-50"><a href="#cb112-50" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">head</span>(full_anemonefish_tree<span class="sc">$</span>tip.label, <span class="dv">10</span>))</span>
<span id="cb112-51"><a href="#cb112-51" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Anemonefish species from your study list (for matching):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-52"><a href="#cb112-52" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">head</span>(anemonefish_species_in_study_tree_format, <span class="dv">10</span>))</span>
<span id="cb112-53"><a href="#cb112-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-54"><a href="#cb112-54" aria-hidden="true" tabindex="-1"></a>    <span class="co"># --- 3. Prune Tree to Studied Species ---</span></span>
<span id="cb112-55"><a href="#cb112-55" aria-hidden="true" tabindex="-1"></a>    species_to_keep_in_tree <span class="ot">&lt;-</span> <span class="fu">intersect</span>(full_anemonefish_tree<span class="sc">$</span>tip.label, anemonefish_species_in_study_tree_format)</span>
<span id="cb112-56"><a href="#cb112-56" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb112-57"><a href="#cb112-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(species_to_keep_in_tree) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb112-58"><a href="#cb112-58" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"CRITICAL WARNING: No matching species found between your study list and the tree's tip labels after cleaning.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-59"><a href="#cb112-59" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"Please check:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-60"><a href="#cb112-60" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"1. The `anemonefish_species_in_study_tree_format` list (should be Genus_species).</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-61"><a href="#cb112-61" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"2. The tip labels in your tree file (`"</span>, anemonefish_tree_file_path, <span class="st">"`) and the cleaning step above.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-62"><a href="#cb112-62" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">length</span>(species_to_keep_in_tree) <span class="sc">&lt;</span> <span class="dv">2</span>) {</span>
<span id="cb112-63"><a href="#cb112-63" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"Warning: Fewer than 2 matching anemonefish species found. Cannot create a meaningful pruned tree.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-64"><a href="#cb112-64" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"Matching species found in tree:"</span>, <span class="fu">paste</span>(species_to_keep_in_tree, <span class="at">collapse=</span><span class="st">", "</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-65"><a href="#cb112-65" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb112-66"><a href="#cb112-66" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Found"</span>, <span class="fu">length</span>(species_to_keep_in_tree), <span class="st">"matching anemonefish species to keep in the tree.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-67"><a href="#cb112-67" aria-hidden="true" tabindex="-1"></a>      species_to_drop_from_tree <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(full_anemonefish_tree<span class="sc">$</span>tip.label, species_to_keep_in_tree)</span>
<span id="cb112-68"><a href="#cb112-68" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb112-69"><a href="#cb112-69" aria-hidden="true" tabindex="-1"></a>      pruned_anemonefish_tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">drop.tip</span>(full_anemonefish_tree, species_to_drop_from_tree)</span>
<span id="cb112-70"><a href="#cb112-70" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Pruned anemonefish tree to"</span>, <span class="fu">length</span>(pruned_anemonefish_tree<span class="sc">$</span>tip.label), <span class="st">"species in study.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-71"><a href="#cb112-71" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb112-72"><a href="#cb112-72" aria-hidden="true" tabindex="-1"></a>      <span class="co"># --- 4. Standardize Tip Labels to Short Codes ---</span></span>
<span id="cb112-73"><a href="#cb112-73" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(pruned_anemonefish_tree) <span class="sc">&amp;&amp;</span> <span class="fu">length</span>(pruned_anemonefish_tree<span class="sc">$</span>tip.label) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb112-74"><a href="#cb112-74" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb112-75"><a href="#cb112-75" aria-hidden="true" tabindex="-1"></a>        fish_label_map_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb112-76"><a href="#cb112-76" aria-hidden="true" tabindex="-1"></a>          <span class="at">Original_Genus_Species =</span> pruned_anemonefish_tree<span class="sc">$</span>tip.label, <span class="co"># These are now Genus_species</span></span>
<span id="cb112-77"><a href="#cb112-77" aria-hidden="true" tabindex="-1"></a>          <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb112-78"><a href="#cb112-78" aria-hidden="true" tabindex="-1"></a>        ) <span class="sc">%&gt;%</span></span>
<span id="cb112-79"><a href="#cb112-79" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(</span>
<span id="cb112-80"><a href="#cb112-80" aria-hidden="true" tabindex="-1"></a>          <span class="at">genus_part =</span> <span class="fu">word</span>(Original_Genus_Species, <span class="dv">1</span>, <span class="at">sep =</span> <span class="st">"_"</span>),</span>
<span id="cb112-81"><a href="#cb112-81" aria-hidden="true" tabindex="-1"></a>          <span class="at">species_part =</span> <span class="fu">word</span>(Original_Genus_Species, <span class="dv">2</span>, <span class="at">sep =</span> <span class="st">"_"</span>),</span>
<span id="cb112-82"><a href="#cb112-82" aria-hidden="true" tabindex="-1"></a>          <span class="co"># Create a more robust short code: First 4 of Genus + First 3 of species</span></span>
<span id="cb112-83"><a href="#cb112-83" aria-hidden="true" tabindex="-1"></a>          <span class="at">short_code_temp =</span> <span class="fu">paste0</span>(</span>
<span id="cb112-84"><a href="#cb112-84" aria-hidden="true" tabindex="-1"></a>            <span class="fu">toupper</span>(<span class="fu">str_sub</span>(genus_part, <span class="dv">1</span>, <span class="fu">min</span>(<span class="dv">4</span>, <span class="fu">nchar</span>(genus_part), <span class="at">na.rm =</span> <span class="cn">TRUE</span>))), </span>
<span id="cb112-85"><a href="#cb112-85" aria-hidden="true" tabindex="-1"></a>            <span class="fu">toupper</span>(<span class="fu">str_sub</span>(species_part, <span class="dv">1</span>, <span class="fu">min</span>(<span class="dv">3</span>, <span class="fu">nchar</span>(species_part), <span class="at">na.rm =</span> <span class="cn">TRUE</span>)))</span>
<span id="cb112-86"><a href="#cb112-86" aria-hidden="true" tabindex="-1"></a>          )</span>
<span id="cb112-87"><a href="#cb112-87" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb112-88"><a href="#cb112-88" aria-hidden="true" tabindex="-1"></a>        fish_label_map_df<span class="sc">$</span>short_code <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(fish_label_map_df<span class="sc">$</span>short_code_temp, <span class="at">sep=</span><span class="st">""</span>)</span>
<span id="cb112-89"><a href="#cb112-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-90"><a href="#cb112-90" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Anemonefish Label Mapping (Genus_species -&gt; Short Code):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-91"><a href="#cb112-91" aria-hidden="true" tabindex="-1"></a>        <span class="fu">print</span>(fish_label_map_df[, <span class="fu">c</span>(<span class="st">"Original_Genus_Species"</span>, <span class="st">"short_code"</span>)])</span>
<span id="cb112-92"><a href="#cb112-92" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb112-93"><a href="#cb112-93" aria-hidden="true" tabindex="-1"></a>        pruned_anemonefish_tree<span class="sc">$</span>tip.label <span class="ot">&lt;-</span> fish_label_map_df<span class="sc">$</span>short_code</span>
<span id="cb112-94"><a href="#cb112-94" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Standardized tip labels for pruned anemonefish tree using short codes.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-95"><a href="#cb112-95" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb112-96"><a href="#cb112-96" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plot</span>(pruned_anemonefish_tree, <span class="at">cex =</span> <span class="fl">0.8</span>, <span class="at">main =</span> <span class="st">"Pruned Anemonefish Tree (Short Labels)"</span>)</span>
<span id="cb112-97"><a href="#cb112-97" aria-hidden="true" tabindex="-1"></a>        <span class="fu">axisPhylo</span>() <span class="co"># Add a scale bar</span></span>
<span id="cb112-98"><a href="#cb112-98" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb112-99"><a href="#cb112-99" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Optional: Check if tree is ultrametric</span></span>
<span id="cb112-100"><a href="#cb112-100" aria-hidden="true" tabindex="-1"></a>        <span class="co"># if (length(pruned_anemonefish_tree$edge.length) &gt; 0 &amp;&amp; !anyNA(pruned_anemonefish_tree$edge.length)) {</span></span>
<span id="cb112-101"><a href="#cb112-101" aria-hidden="true" tabindex="-1"></a>        <span class="co">#   is_ultrametric_fish &lt;- is.ultrametric(pruned_anemonefish_tree)</span></span>
<span id="cb112-102"><a href="#cb112-102" aria-hidden="true" tabindex="-1"></a>        <span class="co">#   cat("Is pruned anemonefish tree ultrametric?", is_ultrametric_fish, "\n")</span></span>
<span id="cb112-103"><a href="#cb112-103" aria-hidden="true" tabindex="-1"></a>        <span class="co"># } else {</span></span>
<span id="cb112-104"><a href="#cb112-104" aria-hidden="true" tabindex="-1"></a>        <span class="co">#   cat("Cannot check ultrametric status: tree may lack branch lengths or have NAs.\n")</span></span>
<span id="cb112-105"><a href="#cb112-105" aria-hidden="true" tabindex="-1"></a>        <span class="co"># }</span></span>
<span id="cb112-106"><a href="#cb112-106" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb112-107"><a href="#cb112-107" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"Pruned anemonefish tree is NULL or has no tip labels after pruning attempt.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-108"><a href="#cb112-108" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb112-109"><a href="#cb112-109" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb112-110"><a href="#cb112-110" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb112-111"><a href="#cb112-111" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Anemonefish tree could not be loaded or is not a 'phylo' object.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-112"><a href="#cb112-112" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb112-113"><a href="#cb112-113" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb112-114"><a href="#cb112-114" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Anemonefish phylogeny file not found at:"</span>, anemonefish_tree_file_path, <span class="st">"</span><span class="sc">\n</span><span class="st">Please provide a valid tree file and update the path.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb112-115"><a href="#cb112-115" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading anemonefish tree from: /Users/chrisrauch/anemonefish_post_analysis/data/phylogeny/anemonefish_phylogeny_OTL.nwk 
Full anemonefish tree loaded. Original tips: 11 
First 10 tip labels from loaded tree (after potential cleaning):
 [1] "Amphiprion_melanopus"    "Amphiprion_akindynos"   
 [3] "Amphiprion_allardi"      "Amphiprion_bicinctus"   
 [5] "Amphiprion_chrysopterus" "Amphiprion_perideraion" 
 [7] "Amphiprion_sandaracinos" "Amphiprion_clarkii"     
 [9] "Premnas_biaculeatus"     "Amphiprion_ocellaris"   
Anemonefish species from your study list (for matching):
 [1] "Amphiprion_akindynos"    "Amphiprion_allardi"     
 [3] "Amphiprion_bicinctus"    "Amphiprion_chrysopterus"
 [5] "Amphiprion_clarkii"      "Amphiprion_melanopus"   
 [7] "Amphiprion_ocellaris"    "Amphiprion_percula"     
 [9] "Amphiprion_perideraion"  "Amphiprion_polymnus"    
Found 11 matching anemonefish species to keep in the tree.
Pruned anemonefish tree to 11 species in study.

Anemonefish Label Mapping (Genus_species -&gt; Short Code):
    Original_Genus_Species short_code
1     Amphiprion_melanopus    AMPHMEL
2     Amphiprion_akindynos    AMPHAKI
3       Amphiprion_allardi    AMPHALL
4     Amphiprion_bicinctus    AMPHBIC
5  Amphiprion_chrysopterus    AMPHCHR
6   Amphiprion_perideraion    AMPHPER
7  Amphiprion_sandaracinos    AMPHSAN
8       Amphiprion_clarkii    AMPHCLA
9      Premnas_biaculeatus    PREMBIA
10    Amphiprion_ocellaris    AMPHOCE
11      Amphiprion_percula   AMPHPER1

Standardized tip labels for pruned anemonefish tree using short codes.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/process-anemonefish-phylogeny-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The object 'pruned_anemonefish_tree' (if created) and 'fish_label_map_df' </span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a><span class="co"># are now available for subsequent phylogenetic analyses.</span></span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a><span class="co"># If 'pruned_anemonefish_tree' is NULL, subsequent phylogenetic analyses should be skipped.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="schoeners-d-niche-overlap-calculations" class="level2">
<h2 class="anchored" data-anchor-id="schoeners-d-niche-overlap-calculations">Schoenerâ€™s D Niche Overlap Calculations</h2>
<p>This section calculates Schoenerâ€™s D niche overlap index between each host sea anemone and its potential anemonefish mutualists (using anemonefish environmental-only models). This is done for the current period and projected future scenarios to assess potential changes in spatial co-occurrence.</p>
</section>
<section id="schoeners-d-niche-overlap-calculations-1" class="level2">
<h2 class="anchored" data-anchor-id="schoeners-d-niche-overlap-calculations-1">Schoenerâ€™s D Niche Overlap Calculations</h2>
<p>This section calculates Schoenerâ€™s D niche overlap index between each host sea anemone and its potential anemonefish mutualists (using anemonefish environmental-only models). This is done for the current period and projected future scenarios to assess potential changes in spatial co-occurrence.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ENMeval) <span class="co"># For calc.niche.overlap</span></span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)   <span class="co"># For pivot_longer, pivot_wider</span></span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(corrplot)</span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr) <span class="co"># For str_replace_all</span></span>
<span id="cb115-7"><a href="#cb115-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-8"><a href="#cb115-8" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Verify Prerequisites ---</span></span>
<span id="cb115-9"><a href="#cb115-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"host_pred_stack_cropped"</span>) <span class="sc">||</span> <span class="fu">is.null</span>(host_pred_stack_cropped)) {</span>
<span id="cb115-10"><a href="#cb115-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"`host_pred_stack_cropped` not found. Run previous chunks."</span>)</span>
<span id="cb115-11"><a href="#cb115-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb115-12"><a href="#cb115-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"fish_pred_stack_cropped"</span>) <span class="sc">||</span> <span class="fu">is.null</span>(fish_pred_stack_cropped)) {</span>
<span id="cb115-13"><a href="#cb115-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"`fish_pred_stack_cropped` (env-only models) not found. Run previous chunks."</span>)</span>
<span id="cb115-14"><a href="#cb115-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb115-15"><a href="#cb115-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"host_pred_stacks_future_individual"</span>) <span class="sc">||</span> <span class="sc">!</span><span class="fu">is.list</span>(host_pred_stacks_future_individual)) {</span>
<span id="cb115-16"><a href="#cb115-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"`host_pred_stacks_future_individual` not found or not a list. Run previous chunks."</span>)</span>
<span id="cb115-17"><a href="#cb115-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb115-18"><a href="#cb115-18" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"fish_pred_stacks_future_individual"</span>) <span class="sc">||</span> <span class="sc">!</span><span class="fu">is.list</span>(fish_pred_stacks_future_individual)) {</span>
<span id="cb115-19"><a href="#cb115-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"`fish_pred_stacks_future_individual` (env-only models) not found or not a list. Run previous chunks."</span>)</span>
<span id="cb115-20"><a href="#cb115-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb115-21"><a href="#cb115-21" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"future_scenarios_to_load"</span>)) {</span>
<span id="cb115-22"><a href="#cb115-22" aria-hidden="true" tabindex="-1"></a>  future_scenarios_to_load <span class="ot">&lt;-</span> config<span class="sc">$</span>env_scenarios[config<span class="sc">$</span>env_scenarios <span class="sc">!=</span> <span class="st">"current"</span>]</span>
<span id="cb115-23"><a href="#cb115-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(future_scenarios_to_load) <span class="sc">==</span> <span class="dv">0</span>) <span class="fu">stop</span>(<span class="st">"No future scenarios defined."</span>)</span>
<span id="cb115-24"><a href="#cb115-24" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb115-25"><a href="#cb115-25" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"anemone_species_df"</span>) <span class="sc">||</span> <span class="sc">!</span><span class="fu">exists</span>(<span class="st">"anemonefish_species_df"</span>)) {</span>
<span id="cb115-26"><a href="#cb115-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"Species data frames (anemone_species_df, anemonefish_species_df) not found."</span>)</span>
<span id="cb115-27"><a href="#cb115-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb115-28"><a href="#cb115-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-29"><a href="#cb115-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-30"><a href="#cb115-30" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 1. Calculate Current Overlap ---</span></span>
<span id="cb115-31"><a href="#cb115-31" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"--- Calculating Current Niche Overlap (Schoener's D) ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Calculating Current Niche Overlap (Schoener's D) ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure layer names are consistent and usable (Genus_species format)</span></span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>names_hosts_current <span class="ot">&lt;-</span> <span class="fu">names</span>(host_pred_stack_cropped)</span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>names_fish_current <span class="ot">&lt;-</span> <span class="fu">names</span>(fish_pred_stack_cropped) <span class="co"># These are from env-only models</span></span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-5"><a href="#cb117-5" aria-hidden="true" tabindex="-1"></a>all_preds_current_for_overlap <span class="ot">&lt;-</span> <span class="fu">c</span>(host_pred_stack_cropped, fish_pred_stack_cropped)</span>
<span id="cb117-6"><a href="#cb117-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure all layer names are unique before stacking for calc.niche.overlap</span></span>
<span id="cb117-7"><a href="#cb117-7" aria-hidden="true" tabindex="-1"></a><span class="co"># This shouldn't be an issue if host and fish names are distinct.</span></span>
<span id="cb117-8"><a href="#cb117-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">any</span>(<span class="fu">duplicated</span>(<span class="fu">names</span>(all_preds_current_for_overlap)))){</span>
<span id="cb117-9"><a href="#cb117-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">warning</span>(<span class="st">"Duplicate layer names found before calculating current overlap. This may cause issues."</span>)</span>
<span id="cb117-10"><a href="#cb117-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb117-11"><a href="#cb117-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-12"><a href="#cb117-12" aria-hidden="true" tabindex="-1"></a>overlap_current_matrix_d <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb117-13"><a href="#cb117-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (terra<span class="sc">::</span><span class="fu">nlyr</span>(all_preds_current_for_overlap) <span class="sc">&gt;=</span> <span class="dv">2</span>) {</span>
<span id="cb117-14"><a href="#cb117-14" aria-hidden="true" tabindex="-1"></a>  overlap_current_matrix_d <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb117-15"><a href="#cb117-15" aria-hidden="true" tabindex="-1"></a>    ENMeval<span class="sc">::</span><span class="fu">calc.niche.overlap</span>(all_preds_current_for_overlap, <span class="st">"D"</span>)</span>
<span id="cb117-16"><a href="#cb117-16" aria-hidden="true" tabindex="-1"></a>  }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb117-17"><a href="#cb117-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Error calculating current niche overlap:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span></span>
<span id="cb117-18"><a href="#cb117-18" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb117-19"><a href="#cb117-19" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb117-20"><a href="#cb117-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Not enough layers in the combined current stack to calculate overlap.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb117-21"><a href="#cb117-21" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Error calculating current niche overlap: unable to find an inherited method for function 'nlayers' for signature 'x = "SpatRaster"' </code></pre>
</div>
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="co"># if (is.null(overlap_current_matrix_d)) {</span></span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   stop("Failed to calculate current niche overlap matrix. Cannot proceed.")</span></span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a><span class="co"># cat("Current niche overlap matrix (Schoener's D) calculated.\n")</span></span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a><span class="co"># # --- 2. Calculate Future Overlaps (Loop through scenarios) ---</span></span>
<span id="cb119-8"><a href="#cb119-8" aria-hidden="true" tabindex="-1"></a><span class="co"># overlap_future_matrices_d &lt;- list()</span></span>
<span id="cb119-9"><a href="#cb119-9" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-10"><a href="#cb119-10" aria-hidden="true" tabindex="-1"></a><span class="co"># for (scenario_name_fut in future_scenarios_to_load) {</span></span>
<span id="cb119-11"><a href="#cb119-11" aria-hidden="true" tabindex="-1"></a><span class="co">#   cat("\n--- Calculating Future Niche Overlap for Scenario:", scenario_name_fut, "---\n")</span></span>
<span id="cb119-12"><a href="#cb119-12" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-13"><a href="#cb119-13" aria-hidden="true" tabindex="-1"></a><span class="co">#   current_future_host_stack_indiv &lt;- host_pred_stacks_future_individual[[scenario_name_fut]]</span></span>
<span id="cb119-14"><a href="#cb119-14" aria-hidden="true" tabindex="-1"></a><span class="co">#   current_future_fish_stack_indiv &lt;- fish_pred_stacks_future_individual[[scenario_name_fut]] # Env-only models</span></span>
<span id="cb119-15"><a href="#cb119-15" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-16"><a href="#cb119-16" aria-hidden="true" tabindex="-1"></a><span class="co">#   if (is.null(current_future_host_stack_indiv) || terra::nlyr(current_future_host_stack_indiv) == 0 ||</span></span>
<span id="cb119-17"><a href="#cb119-17" aria-hidden="true" tabindex="-1"></a><span class="co">#       is.null(current_future_fish_stack_indiv) || terra::nlyr(current_future_fish_stack_indiv) == 0) {</span></span>
<span id="cb119-18"><a href="#cb119-18" aria-hidden="true" tabindex="-1"></a><span class="co">#     cat("  Skipping scenario", scenario_name_fut, "- missing future host or fish prediction stacks.\n")</span></span>
<span id="cb119-19"><a href="#cb119-19" aria-hidden="true" tabindex="-1"></a><span class="co">#     overlap_future_matrices_d[[scenario_name_fut]] &lt;- NULL</span></span>
<span id="cb119-20"><a href="#cb119-20" aria-hidden="true" tabindex="-1"></a><span class="co">#     next</span></span>
<span id="cb119-21"><a href="#cb119-21" aria-hidden="true" tabindex="-1"></a><span class="co">#   }</span></span>
<span id="cb119-22"><a href="#cb119-22" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-23"><a href="#cb119-23" aria-hidden="true" tabindex="-1"></a><span class="co">#   # Important: Ensure layer names in future stacks match the order and content of current stacks</span></span>
<span id="cb119-24"><a href="#cb119-24" aria-hidden="true" tabindex="-1"></a><span class="co">#   # This assumes the lists of species (and thus layers) are consistent.</span></span>
<span id="cb119-25"><a href="#cb119-25" aria-hidden="true" tabindex="-1"></a><span class="co">#   # The loading chunk should have named them correctly as Genus_species.</span></span>
<span id="cb119-26"><a href="#cb119-26" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-27"><a href="#cb119-27" aria-hidden="true" tabindex="-1"></a><span class="co">#   all_preds_future_for_overlap &lt;- c(current_future_host_stack_indiv, current_future_fish_stack_indiv)</span></span>
<span id="cb119-28"><a href="#cb119-28" aria-hidden="true" tabindex="-1"></a><span class="co">#   if (any(duplicated(names(all_preds_future_for_overlap)))){</span></span>
<span id="cb119-29"><a href="#cb119-29" aria-hidden="true" tabindex="-1"></a><span class="co">#     warning("Duplicate layer names found before calculating future overlap for ", scenario_name_fut, ". This may cause issues.")</span></span>
<span id="cb119-30"><a href="#cb119-30" aria-hidden="true" tabindex="-1"></a><span class="co">#   }</span></span>
<span id="cb119-31"><a href="#cb119-31" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-32"><a href="#cb119-32" aria-hidden="true" tabindex="-1"></a><span class="co">#   if (terra::nlyr(all_preds_future_for_overlap) &gt;= 2) {</span></span>
<span id="cb119-33"><a href="#cb119-33" aria-hidden="true" tabindex="-1"></a><span class="co">#     overlap_future_matrices_d[[scenario_name_fut]] &lt;- tryCatch({</span></span>
<span id="cb119-34"><a href="#cb119-34" aria-hidden="true" tabindex="-1"></a><span class="co">#       # Align future stack geometry with current stack geometry for calcNicheOverlap robustness</span></span>
<span id="cb119-35"><a href="#cb119-35" aria-hidden="true" tabindex="-1"></a><span class="co">#       if(!terra::compareGeom(all_preds_current_for_overlap, all_preds_future_for_overlap, stopOnError=FALSE, res=TRUE, crs=TRUE, ext=TRUE)){</span></span>
<span id="cb119-36"><a href="#cb119-36" aria-hidden="true" tabindex="-1"></a><span class="co">#           cat("   Resampling future combined stack for scenario", scenario_name_fut, "to match current geometry.\n")</span></span>
<span id="cb119-37"><a href="#cb119-37" aria-hidden="true" tabindex="-1"></a><span class="co">#           all_preds_future_for_overlap &lt;- terra::resample(all_preds_future_for_overlap, all_preds_current_for_overlap, method="bilinear")</span></span>
<span id="cb119-38"><a href="#cb119-38" aria-hidden="true" tabindex="-1"></a><span class="co">#       }</span></span>
<span id="cb119-39"><a href="#cb119-39" aria-hidden="true" tabindex="-1"></a><span class="co">#       ENMeval::calc.niche.overlap(all_preds_future_for_overlap, "D")</span></span>
<span id="cb119-40"><a href="#cb119-40" aria-hidden="true" tabindex="-1"></a><span class="co">#     }, error = function(e) {</span></span>
<span id="cb119-41"><a href="#cb119-41" aria-hidden="true" tabindex="-1"></a><span class="co">#       cat("  Error calculating future niche overlap for", scenario_name_fut, ":", e$message, "\n"); NULL</span></span>
<span id="cb119-42"><a href="#cb119-42" aria-hidden="true" tabindex="-1"></a><span class="co">#     })</span></span>
<span id="cb119-43"><a href="#cb119-43" aria-hidden="true" tabindex="-1"></a><span class="co">#     if (!is.null(overlap_future_matrices_d[[scenario_name_fut]])) {</span></span>
<span id="cb119-44"><a href="#cb119-44" aria-hidden="true" tabindex="-1"></a><span class="co">#       cat("  Future niche overlap matrix (Schoener's D) calculated for", scenario_name_fut, "\n")</span></span>
<span id="cb119-45"><a href="#cb119-45" aria-hidden="true" tabindex="-1"></a><span class="co">#     }</span></span>
<span id="cb119-46"><a href="#cb119-46" aria-hidden="true" tabindex="-1"></a><span class="co">#   } else {</span></span>
<span id="cb119-47"><a href="#cb119-47" aria-hidden="true" tabindex="-1"></a><span class="co">#     cat("  Not enough layers in combined future stack for", scenario_name_fut, "to calculate overlap.\n")</span></span>
<span id="cb119-48"><a href="#cb119-48" aria-hidden="true" tabindex="-1"></a><span class="co">#     overlap_future_matrices_d[[scenario_name_fut]] &lt;- NULL</span></span>
<span id="cb119-49"><a href="#cb119-49" aria-hidden="true" tabindex="-1"></a><span class="co">#   }</span></span>
<span id="cb119-50"><a href="#cb119-50" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb119-51"><a href="#cb119-51" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-52"><a href="#cb119-52" aria-hidden="true" tabindex="-1"></a><span class="co"># # --- 3. Process and Compare Overlap Matrices ---</span></span>
<span id="cb119-53"><a href="#cb119-53" aria-hidden="true" tabindex="-1"></a><span class="co"># # Ensure names_hosts_current and names_fish_current are Genus_species format</span></span>
<span id="cb119-54"><a href="#cb119-54" aria-hidden="true" tabindex="-1"></a><span class="co"># valid_host_rows &lt;- intersect(names_hosts_current, rownames(overlap_current_matrix_d))</span></span>
<span id="cb119-55"><a href="#cb119-55" aria-hidden="true" tabindex="-1"></a><span class="co"># valid_fish_cols &lt;- intersect(names_fish_current, colnames(overlap_current_matrix_d))</span></span>
<span id="cb119-56"><a href="#cb119-56" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-57"><a href="#cb119-57" aria-hidden="true" tabindex="-1"></a><span class="co"># if (length(valid_host_rows) == 0 || length(valid_fish_cols) == 0) {</span></span>
<span id="cb119-58"><a href="#cb119-58" aria-hidden="true" tabindex="-1"></a><span class="co">#   stop("No valid host or fish names found to subset the current overlap matrix. Check layer naming of input stacks.")</span></span>
<span id="cb119-59"><a href="#cb119-59" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb119-60"><a href="#cb119-60" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-61"><a href="#cb119-61" aria-hidden="true" tabindex="-1"></a><span class="co"># # Extract the host-fish interaction part of the matrix</span></span>
<span id="cb119-62"><a href="#cb119-62" aria-hidden="true" tabindex="-1"></a><span class="co"># current_host_fish_overlap_raw &lt;- overlap_current_matrix_d[valid_host_rows, valid_fish_cols, drop = FALSE]</span></span>
<span id="cb119-63"><a href="#cb119-63" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-64"><a href="#cb119-64" aria-hidden="true" tabindex="-1"></a><span class="co"># # Convert to long format</span></span>
<span id="cb119-65"><a href="#cb119-65" aria-hidden="true" tabindex="-1"></a><span class="co"># current_host_fish_overlap_long &lt;- current_host_fish_overlap_raw %&gt;%</span></span>
<span id="cb119-66"><a href="#cb119-66" aria-hidden="true" tabindex="-1"></a><span class="co">#   as.data.frame() %&gt;%</span></span>
<span id="cb119-67"><a href="#cb119-67" aria-hidden="true" tabindex="-1"></a><span class="co">#   tibble::rownames_to_column("host_species") %&gt;%</span></span>
<span id="cb119-68"><a href="#cb119-68" aria-hidden="true" tabindex="-1"></a><span class="co">#   pivot_longer(cols = -host_species, names_to = "fish_species", values_to = "overlap_current")</span></span>
<span id="cb119-69"><a href="#cb119-69" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-70"><a href="#cb119-70" aria-hidden="true" tabindex="-1"></a><span class="co"># cat("\nMean current overlap (Schoener's D) between all host-fish pairs:", round(mean(current_host_fish_overlap_long$overlap_current, na.rm = TRUE), 3), "\n")</span></span>
<span id="cb119-71"><a href="#cb119-71" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-72"><a href="#cb119-72" aria-hidden="true" tabindex="-1"></a><span class="co"># # --- Choose scenarios for detailed comparison (e.g., SSP1-1.9 2050 and SSP5-8.5 2100) ---</span></span>
<span id="cb119-73"><a href="#cb119-73" aria-hidden="true" tabindex="-1"></a><span class="co"># # These names must match the keys in overlap_future_matrices_d and config$env_scenarios</span></span>
<span id="cb119-74"><a href="#cb119-74" aria-hidden="true" tabindex="-1"></a><span class="co"># scenarios_for_detailed_comparison &lt;- intersect(c("ssp119_2050", "ssp585_2100"), names(overlap_future_matrices_d))</span></span>
<span id="cb119-75"><a href="#cb119-75" aria-hidden="true" tabindex="-1"></a><span class="co"># if(length(scenarios_for_detailed_comparison) == 0) {</span></span>
<span id="cb119-76"><a href="#cb119-76" aria-hidden="true" tabindex="-1"></a><span class="co">#     cat("Warning: None of the specified detailed comparison scenarios (ssp119_2050, ssp585_2100) have overlap data. Plotting will be skipped for them.\n")</span></span>
<span id="cb119-77"><a href="#cb119-77" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb119-78"><a href="#cb119-78" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-79"><a href="#cb119-79" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-80"><a href="#cb119-80" aria-hidden="true" tabindex="-1"></a><span class="co"># # --- Loop through chosen future scenarios for t-tests and plots ---</span></span>
<span id="cb119-81"><a href="#cb119-81" aria-hidden="true" tabindex="-1"></a><span class="co"># for (chosen_future_scenario in scenarios_for_detailed_comparison) {</span></span>
<span id="cb119-82"><a href="#cb119-82" aria-hidden="true" tabindex="-1"></a><span class="co">#   cat("\n--- Detailed Comparison for Future Scenario:", chosen_future_scenario, "---\n")</span></span>
<span id="cb119-83"><a href="#cb119-83" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-84"><a href="#cb119-84" aria-hidden="true" tabindex="-1"></a><span class="co">#   overlap_future_chosen_matrix &lt;- overlap_future_matrices_d[[chosen_future_scenario]]</span></span>
<span id="cb119-85"><a href="#cb119-85" aria-hidden="true" tabindex="-1"></a><span class="co">#   if (is.null(overlap_future_chosen_matrix)) {</span></span>
<span id="cb119-86"><a href="#cb119-86" aria-hidden="true" tabindex="-1"></a><span class="co">#     cat("  No future overlap matrix available for", chosen_future_scenario, ". Skipping detailed comparison.\n")</span></span>
<span id="cb119-87"><a href="#cb119-87" aria-hidden="true" tabindex="-1"></a><span class="co">#     next</span></span>
<span id="cb119-88"><a href="#cb119-88" aria-hidden="true" tabindex="-1"></a><span class="co">#   }</span></span>
<span id="cb119-89"><a href="#cb119-89" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-90"><a href="#cb119-90" aria-hidden="true" tabindex="-1"></a><span class="co">#   future_host_fish_overlap_raw_chosen &lt;- overlap_future_chosen_matrix[valid_host_rows, valid_fish_cols, drop = FALSE]</span></span>
<span id="cb119-91"><a href="#cb119-91" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-92"><a href="#cb119-92" aria-hidden="true" tabindex="-1"></a><span class="co">#   future_host_fish_overlap_long_chosen &lt;- future_host_fish_overlap_raw_chosen %&gt;%</span></span>
<span id="cb119-93"><a href="#cb119-93" aria-hidden="true" tabindex="-1"></a><span class="co">#     as.data.frame() %&gt;%</span></span>
<span id="cb119-94"><a href="#cb119-94" aria-hidden="true" tabindex="-1"></a><span class="co">#     tibble::rownames_to_column("host_species") %&gt;%</span></span>
<span id="cb119-95"><a href="#cb119-95" aria-hidden="true" tabindex="-1"></a><span class="co">#     pivot_longer(cols = -host_species, names_to = "fish_species", values_to = "overlap_future")</span></span>
<span id="cb119-96"><a href="#cb119-96" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-97"><a href="#cb119-97" aria-hidden="true" tabindex="-1"></a><span class="co">#   cat("Mean future overlap (Schoener's D) for", chosen_future_scenario, ":", round(mean(future_host_fish_overlap_long_chosen$overlap_future, na.rm = TRUE), 3), "\n")</span></span>
<span id="cb119-98"><a href="#cb119-98" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-99"><a href="#cb119-99" aria-hidden="true" tabindex="-1"></a><span class="co">#   comparison_df &lt;- current_host_fish_overlap_long %&gt;%</span></span>
<span id="cb119-100"><a href="#cb119-100" aria-hidden="true" tabindex="-1"></a><span class="co">#     left_join(future_host_fish_overlap_long_chosen, by = c("host_species", "fish_species")) %&gt;%</span></span>
<span id="cb119-101"><a href="#cb119-101" aria-hidden="true" tabindex="-1"></a><span class="co">#     filter(!is.na(overlap_current), !is.na(overlap_future))</span></span>
<span id="cb119-102"><a href="#cb119-102" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-103"><a href="#cb119-103" aria-hidden="true" tabindex="-1"></a><span class="co">#   if (nrow(comparison_df) &gt; 1) {</span></span>
<span id="cb119-104"><a href="#cb119-104" aria-hidden="true" tabindex="-1"></a><span class="co">#     ttest_result &lt;- t.test(comparison_df$overlap_current, comparison_df$overlap_future, paired = TRUE)</span></span>
<span id="cb119-105"><a href="#cb119-105" aria-hidden="true" tabindex="-1"></a><span class="co">#     cat("\nPaired t-test: Current vs Future (", chosen_future_scenario, ") Overlap:\n")</span></span>
<span id="cb119-106"><a href="#cb119-106" aria-hidden="true" tabindex="-1"></a><span class="co">#     print(ttest_result)</span></span>
<span id="cb119-107"><a href="#cb119-107" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-108"><a href="#cb119-108" aria-hidden="true" tabindex="-1"></a><span class="co">#     # Scatter Plot</span></span>
<span id="cb119-109"><a href="#cb119-109" aria-hidden="true" tabindex="-1"></a><span class="co">#     plot_scatter &lt;- ggplot(comparison_df, aes(x = overlap_current, y = overlap_future)) +</span></span>
<span id="cb119-110"><a href="#cb119-110" aria-hidden="true" tabindex="-1"></a><span class="co">#       geom_point(alpha = 0.6, color = "darkblue") +</span></span>
<span id="cb119-111"><a href="#cb119-111" aria-hidden="true" tabindex="-1"></a><span class="co">#       geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "red") +</span></span>
<span id="cb119-112"><a href="#cb119-112" aria-hidden="true" tabindex="-1"></a><span class="co">#       xlim(0, 1) + ylim(0, 1) +</span></span>
<span id="cb119-113"><a href="#cb119-113" aria-hidden="true" tabindex="-1"></a><span class="co">#       labs(</span></span>
<span id="cb119-114"><a href="#cb119-114" aria-hidden="true" tabindex="-1"></a><span class="co">#         title = paste("Niche Overlap Change:", chosen_future_scenario, "vs Current"),</span></span>
<span id="cb119-115"><a href="#cb119-115" aria-hidden="true" tabindex="-1"></a><span class="co">#         x = "Schoener's D (Current Conditions)",</span></span>
<span id="cb119-116"><a href="#cb119-116" aria-hidden="true" tabindex="-1"></a><span class="co">#         y = paste("Schoener's D (Future -", chosen_future_scenario, ")")</span></span>
<span id="cb119-117"><a href="#cb119-117" aria-hidden="true" tabindex="-1"></a><span class="co">#       ) +</span></span>
<span id="cb119-118"><a href="#cb119-118" aria-hidden="true" tabindex="-1"></a><span class="co">#       theme_bw(base_size = 12) +</span></span>
<span id="cb119-119"><a href="#cb119-119" aria-hidden="true" tabindex="-1"></a><span class="co">#       coord_fixed()</span></span>
<span id="cb119-120"><a href="#cb119-120" aria-hidden="true" tabindex="-1"></a><span class="co">#     print(plot_scatter)</span></span>
<span id="cb119-121"><a href="#cb119-121" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-122"><a href="#cb119-122" aria-hidden="true" tabindex="-1"></a><span class="co">#     # Proportional Change Matrix Plot</span></span>
<span id="cb119-123"><a href="#cb119-123" aria-hidden="true" tabindex="-1"></a><span class="co">#     # Prepare display names (full scientific names with spaces)</span></span>
<span id="cb119-124"><a href="#cb119-124" aria-hidden="true" tabindex="-1"></a><span class="co">#     host_display_names_map &lt;- anemone_species_df %&gt;%</span></span>
<span id="cb119-125"><a href="#cb119-125" aria-hidden="true" tabindex="-1"></a><span class="co">#       select(scientificName) %&gt;%</span></span>
<span id="cb119-126"><a href="#cb119-126" aria-hidden="true" tabindex="-1"></a><span class="co">#       mutate(sanitized_name = gsub(" ", "_", scientificName)) %&gt;%</span></span>
<span id="cb119-127"><a href="#cb119-127" aria-hidden="true" tabindex="-1"></a><span class="co">#       distinct(sanitized_name, .keep_all = TRUE) %&gt;%</span></span>
<span id="cb119-128"><a href="#cb119-128" aria-hidden="true" tabindex="-1"></a><span class="co">#       tibble::deframe()</span></span>
<span id="cb119-129"><a href="#cb119-129" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-130"><a href="#cb119-130" aria-hidden="true" tabindex="-1"></a><span class="co">#     fish_display_names_map &lt;- anemonefish_species_df %&gt;%</span></span>
<span id="cb119-131"><a href="#cb119-131" aria-hidden="true" tabindex="-1"></a><span class="co">#       select(scientificName) %&gt;%</span></span>
<span id="cb119-132"><a href="#cb119-132" aria-hidden="true" tabindex="-1"></a><span class="co">#       mutate(sanitized_name = gsub(" ", "_", scientificName)) %&gt;%</span></span>
<span id="cb119-133"><a href="#cb119-133" aria-hidden="true" tabindex="-1"></a><span class="co">#       distinct(sanitized_name, .keep_all = TRUE) %&gt;%</span></span>
<span id="cb119-134"><a href="#cb119-134" aria-hidden="true" tabindex="-1"></a><span class="co">#       tibble::deframe()</span></span>
<span id="cb119-135"><a href="#cb119-135" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-136"><a href="#cb119-136" aria-hidden="true" tabindex="-1"></a><span class="co">#     prop_change_df &lt;- comparison_df %&gt;%</span></span>
<span id="cb119-137"><a href="#cb119-137" aria-hidden="true" tabindex="-1"></a><span class="co">#       mutate(</span></span>
<span id="cb119-138"><a href="#cb119-138" aria-hidden="true" tabindex="-1"></a><span class="co">#         proportional_change = case_when(</span></span>
<span id="cb119-139"><a href="#cb119-139" aria-hidden="true" tabindex="-1"></a><span class="co">#           overlap_current == 0 &amp; overlap_future == 0 ~ 0, # Both zero, no change</span></span>
<span id="cb119-140"><a href="#cb119-140" aria-hidden="true" tabindex="-1"></a><span class="co">#           overlap_current == 0 &amp; overlap_future &gt; 0  ~ 1, # New overlap, max positive change</span></span>
<span id="cb119-141"><a href="#cb119-141" aria-hidden="true" tabindex="-1"></a><span class="co">#           overlap_current &gt; 0  &amp; overlap_future == 0 ~ -1, # Lost overlap, max negative change</span></span>
<span id="cb119-142"><a href="#cb119-142" aria-hidden="true" tabindex="-1"></a><span class="co">#           overlap_current &gt; 0                         ~ (overlap_future - overlap_current) / overlap_current,</span></span>
<span id="cb119-143"><a href="#cb119-143" aria-hidden="true" tabindex="-1"></a><span class="co">#           TRUE                                        ~ NA_real_ # Should not happen if filtered NAs</span></span>
<span id="cb119-144"><a href="#cb119-144" aria-hidden="true" tabindex="-1"></a><span class="co">#         )</span></span>
<span id="cb119-145"><a href="#cb119-145" aria-hidden="true" tabindex="-1"></a><span class="co">#       ) %&gt;%</span></span>
<span id="cb119-146"><a href="#cb119-146" aria-hidden="true" tabindex="-1"></a><span class="co">#       mutate(proportional_change_capped = pmax(-1, pmin(1, proportional_change))) %&gt;%</span></span>
<span id="cb119-147"><a href="#cb119-147" aria-hidden="true" tabindex="-1"></a><span class="co">#       select(host_species, fish_species, proportional_change_capped)</span></span>
<span id="cb119-148"><a href="#cb119-148" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-149"><a href="#cb119-149" aria-hidden="true" tabindex="-1"></a><span class="co">#     if(nrow(prop_change_df) &gt; 0) {</span></span>
<span id="cb119-150"><a href="#cb119-150" aria-hidden="true" tabindex="-1"></a><span class="co">#         prop_change_matrix &lt;- prop_change_df %&gt;%</span></span>
<span id="cb119-151"><a href="#cb119-151" aria-hidden="true" tabindex="-1"></a><span class="co">#           pivot_wider(names_from = fish_species, values_from = proportional_change_capped) %&gt;%</span></span>
<span id="cb119-152"><a href="#cb119-152" aria-hidden="true" tabindex="-1"></a><span class="co">#           tibble::column_to_rownames("host_species") %&gt;%</span></span>
<span id="cb119-153"><a href="#cb119-153" aria-hidden="true" tabindex="-1"></a><span class="co">#           as.matrix()</span></span>
<span id="cb119-154"><a href="#cb119-154" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-155"><a href="#cb119-155" aria-hidden="true" tabindex="-1"></a><span class="co">#         # Use display names if mapping is successful</span></span>
<span id="cb119-156"><a href="#cb119-156" aria-hidden="true" tabindex="-1"></a><span class="co">#         rownames_display &lt;- host_display_names_map[rownames(prop_change_matrix)]</span></span>
<span id="cb119-157"><a href="#cb119-157" aria-hidden="true" tabindex="-1"></a><span class="co">#         colnames_display &lt;- fish_display_names_map[colnames(prop_change_matrix)]</span></span>
<span id="cb119-158"><a href="#cb119-158" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-159"><a href="#cb119-159" aria-hidden="true" tabindex="-1"></a><span class="co">#         # Fallback if any name not found in map</span></span>
<span id="cb119-160"><a href="#cb119-160" aria-hidden="true" tabindex="-1"></a><span class="co">#         rownames(prop_change_matrix) &lt;- ifelse(is.na(rownames_display), rownames(prop_change_matrix), rownames_display)</span></span>
<span id="cb119-161"><a href="#cb119-161" aria-hidden="true" tabindex="-1"></a><span class="co">#         colnames(prop_change_matrix) &lt;- ifelse(is.na(colnames_display), colnames(prop_change_matrix), colnames_display)</span></span>
<span id="cb119-162"><a href="#cb119-162" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-163"><a href="#cb119-163" aria-hidden="true" tabindex="-1"></a><span class="co">#         if(nrow(prop_change_matrix) &gt; 0 &amp;&amp; ncol(prop_change_matrix) &gt; 0) {</span></span>
<span id="cb119-164"><a href="#cb119-164" aria-hidden="true" tabindex="-1"></a><span class="co">#           corrplot::corrplot(prop_change_matrix,</span></span>
<span id="cb119-165"><a href="#cb119-165" aria-hidden="true" tabindex="-1"></a><span class="co">#                            method = "circle", is.corr = FALSE, tl.srt = 60, tl.col = 'black',</span></span>
<span id="cb119-166"><a href="#cb119-166" aria-hidden="true" tabindex="-1"></a><span class="co">#                            tl.cex = 0.6, cl.cex = 0.7,</span></span>
<span id="cb119-167"><a href="#cb119-167" aria-hidden="true" tabindex="-1"></a><span class="co">#                            col.lim = c(-1, 1),</span></span>
<span id="cb119-168"><a href="#cb119-168" aria-hidden="true" tabindex="-1"></a><span class="co">#                            col = colorRampPalette(c("firebrick3", "white", "dodgerblue3"))(200),</span></span>
<span id="cb119-169"><a href="#cb119-169" aria-hidden="true" tabindex="-1"></a><span class="co">#                            title = paste("Proportional Change in Host-Fish Overlap\n(", chosen_future_scenario, " vs Current)"),</span></span>
<span id="cb119-170"><a href="#cb119-170" aria-hidden="true" tabindex="-1"></a><span class="co">#                            mar = c(0,0,2,0)) # Adjusted margins for title</span></span>
<span id="cb119-171"><a href="#cb119-171" aria-hidden="true" tabindex="-1"></a><span class="co">#         } else {cat("  Not enough data to create proportional change matrix plot for", chosen_future_scenario, "\n")}</span></span>
<span id="cb119-172"><a href="#cb119-172" aria-hidden="true" tabindex="-1"></a><span class="co">#     } else {cat("  No data for proportional change calculation for", chosen_future_scenario, "\n")}</span></span>
<span id="cb119-173"><a href="#cb119-173" aria-hidden="true" tabindex="-1"></a><span class="co">#   } else { cat("  Not enough paired overlap data for t-test or plotting for", chosen_future_scenario, "\n") }</span></span>
<span id="cb119-174"><a href="#cb119-174" aria-hidden="true" tabindex="-1"></a><span class="co"># } # End loop chosen_future_scenario</span></span>
<span id="cb119-175"><a href="#cb119-175" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb119-176"><a href="#cb119-176" aria-hidden="true" tabindex="-1"></a><span class="co"># cat("\n--- Schoener's D calculations and comparisons finished. ---\n")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="phylogenetic-signal-of-anemonefish-niche-overlap" class="level2">
<h2 class="anchored" data-anchor-id="phylogenetic-signal-of-anemonefish-niche-overlap">Phylogenetic Signal of Anemonefish Niche Overlap</h2>
<p>This section tests for phylogenetic signal in the niche overlap values among anemonefish species (using environmental-only models). This is done for the current period and a selected future scenario to see if the relationship between phylogeny and overlap changes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(picante) </span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)     </span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">is.null</span>(pruned_anemonefish_tree)) {</span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: `pruned_anemonefish_tree` not available. Skipping phylogenetic signal of fish-fish overlap.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-6"><a href="#cb120-6" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">is.null</span>(overlap_current_matrix_d) <span class="sc">||</span> </span>
<span id="cb120-7"><a href="#cb120-7" aria-hidden="true" tabindex="-1"></a>           <span class="fu">length</span>(<span class="fu">intersect</span>(pruned_anemonefish_tree<span class="sc">$</span>tip.label, fish_label_map_df<span class="sc">$</span>short_code[fish_label_map_df<span class="sc">$</span>Original_Genus_Species <span class="sc">%in%</span> names_fish_current])) <span class="sc">&lt;</span> <span class="dv">2</span>) { </span>
<span id="cb120-8"><a href="#cb120-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: Current overlap matrix or mapping to tree tips insufficient for anemonefish. Skipping phylogenetic signal of fish-fish overlap.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-9"><a href="#cb120-9" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb120-10"><a href="#cb120-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"--- Calculating Phylogenetic Signal of Anemonefish-Anemonefish Niche Overlap ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-11"><a href="#cb120-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb120-12"><a href="#cb120-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- 1. Create phylogenetic distance matrix for anemonefish ---</span></span>
<span id="cb120-13"><a href="#cb120-13" aria-hidden="true" tabindex="-1"></a>  fish_phy_dist_matrix <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb120-14"><a href="#cb120-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(pruned_anemonefish_tree<span class="sc">$</span>edge.length) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">anyNA</span>(pruned_anemonefish_tree<span class="sc">$</span>edge.length)) {</span>
<span id="cb120-15"><a href="#cb120-15" aria-hidden="true" tabindex="-1"></a>      fish_phy_dist_matrix <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>(ape<span class="sc">::</span><span class="fu">cophenetic.phylo</span>(pruned_anemonefish_tree), <span class="at">error =</span> <span class="cf">function</span>(e){</span>
<span id="cb120-16"><a href="#cb120-16" aria-hidden="true" tabindex="-1"></a>          <span class="fu">cat</span>(<span class="st">"Error creating phylogenetic distance matrix:"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span></span>
<span id="cb120-17"><a href="#cb120-17" aria-hidden="true" tabindex="-1"></a>      })</span>
<span id="cb120-18"><a href="#cb120-18" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb120-19"><a href="#cb120-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Warning: Pruned anemonefish tree lacks branch lengths or has NAs; cannot calculate cophenetic distances for Mantel test.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-20"><a href="#cb120-20" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb120-21"><a href="#cb120-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-22"><a href="#cb120-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure fish_label_map_df is available (should be from phylogeny chunk)</span></span>
<span id="cb120-23"><a href="#cb120-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"fish_label_map_df"</span>) <span class="sc">||</span> <span class="fu">is.null</span>(fish_label_map_df)) {</span>
<span id="cb120-24"><a href="#cb120-24" aria-hidden="true" tabindex="-1"></a>      <span class="fu">stop</span>(<span class="st">"`fish_label_map_df` (mapping short codes to Genus_species) is missing."</span>)</span>
<span id="cb120-25"><a href="#cb120-25" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb120-26"><a href="#cb120-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-27"><a href="#cb120-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- Function to perform Mantel test for a given overlap matrix ---</span></span>
<span id="cb120-28"><a href="#cb120-28" aria-hidden="true" tabindex="-1"></a>  perform_overlap_mantel_test <span class="ot">&lt;-</span> <span class="cf">function</span>(overlap_matrix, matrix_label, phy_dist_matrix, tree_tips_short_codes, label_map_df) {</span>
<span id="cb120-29"><a href="#cb120-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Processing Mantel Test for:"</span>, matrix_label, <span class="st">"---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-30"><a href="#cb120-30" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-31"><a href="#cb120-31" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">is.null</span>(overlap_matrix) <span class="sc">||</span> <span class="fu">is.null</span>(phy_dist_matrix)) {</span>
<span id="cb120-32"><a href="#cb120-32" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Skipping Mantel test for"</span>, matrix_label, <span class="st">"- missing overlap or phylogenetic distance matrix.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-33"><a href="#cb120-33" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb120-34"><a href="#cb120-34" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb120-35"><a href="#cb120-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-36"><a href="#cb120-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Map short codes on tree back to Genus_species to subset the full overlap matrix</span></span>
<span id="cb120-37"><a href="#cb120-37" aria-hidden="true" tabindex="-1"></a>    genus_species_for_tree_tips_local <span class="ot">&lt;-</span> label_map_df<span class="sc">$</span>Original_Genus_Species[<span class="fu">match</span>(tree_tips_short_codes, label_map_df<span class="sc">$</span>short_code)]</span>
<span id="cb120-38"><a href="#cb120-38" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-39"><a href="#cb120-39" aria-hidden="true" tabindex="-1"></a>    valid_fish_for_overlap_subset <span class="ot">&lt;-</span> <span class="fu">intersect</span>(genus_species_for_tree_tips_local, <span class="fu">colnames</span>(overlap_matrix))</span>
<span id="cb120-40"><a href="#cb120-40" aria-hidden="true" tabindex="-1"></a>    valid_fish_for_overlap_subset <span class="ot">&lt;-</span> <span class="fu">intersect</span>(valid_fish_for_overlap_subset, <span class="fu">rownames</span>(overlap_matrix))</span>
<span id="cb120-41"><a href="#cb120-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-42"><a href="#cb120-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(valid_fish_for_overlap_subset) <span class="sc">&lt;</span> <span class="dv">2</span>) {</span>
<span id="cb120-43"><a href="#cb120-43" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Less than 2 anemonefish species with both tree data and overlap data for"</span>, matrix_label, <span class="st">". Skipping Mantel test.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-44"><a href="#cb120-44" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb120-45"><a href="#cb120-45" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb120-46"><a href="#cb120-46" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-47"><a href="#cb120-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Subset the full overlap matrix to only include these fish species</span></span>
<span id="cb120-48"><a href="#cb120-48" aria-hidden="true" tabindex="-1"></a>    fish_fish_overlap_submatrix_local <span class="ot">&lt;-</span> overlap_matrix[valid_fish_for_overlap_subset, valid_fish_for_overlap_subset, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb120-49"><a href="#cb120-49" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-50"><a href="#cb120-50" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Convert overlap (similarity) to dissimilarity (1 - D)</span></span>
<span id="cb120-51"><a href="#cb120-51" aria-hidden="true" tabindex="-1"></a>    fish_fish_dissimilarity_matrix_local <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">-</span> fish_fish_overlap_submatrix_local</span>
<span id="cb120-52"><a href="#cb120-52" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-53"><a href="#cb120-53" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Rename rows/cols of dissimilarity matrix to short codes for matching with phy_dist_matrix</span></span>
<span id="cb120-54"><a href="#cb120-54" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(fish_fish_dissimilarity_matrix_local) <span class="ot">&lt;-</span> label_map_df<span class="sc">$</span>short_code[<span class="fu">match</span>(<span class="fu">rownames</span>(fish_fish_dissimilarity_matrix_local), label_map_df<span class="sc">$</span>Original_Genus_Species)]</span>
<span id="cb120-55"><a href="#cb120-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(fish_fish_dissimilarity_matrix_local) <span class="ot">&lt;-</span> label_map_df<span class="sc">$</span>short_code[<span class="fu">match</span>(<span class="fu">colnames</span>(fish_fish_dissimilarity_matrix_local), label_map_df<span class="sc">$</span>Original_Genus_Species)]</span>
<span id="cb120-56"><a href="#cb120-56" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-57"><a href="#cb120-57" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Ensure phy_dist_matrix also uses short codes (it should from previous chunk)</span></span>
<span id="cb120-58"><a href="#cb120-58" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Match order of phy_dist_matrix and dissimilarity_matrix</span></span>
<span id="cb120-59"><a href="#cb120-59" aria-hidden="true" tabindex="-1"></a>    common_short_codes_for_mantel <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">rownames</span>(phy_dist_matrix), <span class="fu">rownames</span>(fish_fish_dissimilarity_matrix_local))</span>
<span id="cb120-60"><a href="#cb120-60" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-61"><a href="#cb120-61" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(common_short_codes_for_mantel) <span class="sc">&lt;</span> <span class="dv">2</span>) {</span>
<span id="cb120-62"><a href="#cb120-62" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Not enough common species with short codes between phy_dist and overlap_dist for"</span>, matrix_label, <span class="st">". Skipping Mantel.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-63"><a href="#cb120-63" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(<span class="cn">NULL</span>)</span>
<span id="cb120-64"><a href="#cb120-64" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb120-65"><a href="#cb120-65" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-66"><a href="#cb120-66" aria-hidden="true" tabindex="-1"></a>    phy_dist_final <span class="ot">&lt;-</span> phy_dist_matrix[common_short_codes_for_mantel, common_short_codes_for_mantel]</span>
<span id="cb120-67"><a href="#cb120-67" aria-hidden="true" tabindex="-1"></a>    dissim_final <span class="ot">&lt;-</span> fish_fish_dissimilarity_matrix_local[common_short_codes_for_mantel, common_short_codes_for_mantel]</span>
<span id="cb120-68"><a href="#cb120-68" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-69"><a href="#cb120-69" aria-hidden="true" tabindex="-1"></a>    mantel_result <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb120-70"><a href="#cb120-70" aria-hidden="true" tabindex="-1"></a>      vegan<span class="sc">::</span><span class="fu">mantel</span>(<span class="fu">as.dist</span>(phy_dist_final), <span class="fu">as.dist</span>(dissim_final), <span class="at">na.rm =</span> <span class="cn">TRUE</span>, <span class="at">permutations =</span> <span class="dv">999</span>)</span>
<span id="cb120-71"><a href="#cb120-71" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e){<span class="fu">cat</span>(<span class="st">"  Error during Mantel test for"</span>, matrix_label, <span class="st">":"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span>})</span>
<span id="cb120-72"><a href="#cb120-72" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-73"><a href="#cb120-73" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(mantel_result)) {</span>
<span id="cb120-74"><a href="#cb120-74" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Mantel Test Results for"</span>, matrix_label, <span class="st">":</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-75"><a href="#cb120-75" aria-hidden="true" tabindex="-1"></a>      <span class="fu">print</span>(mantel_result)</span>
<span id="cb120-76"><a href="#cb120-76" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb120-77"><a href="#cb120-77" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(mantel_result)</span>
<span id="cb120-78"><a href="#cb120-78" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb120-79"><a href="#cb120-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-80"><a href="#cb120-80" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- 2. Current Anemonefish-Anemonefish Overlap Mantel Test ---</span></span>
<span id="cb120-81"><a href="#cb120-81" aria-hidden="true" tabindex="-1"></a>  mantel_current_fish_overlap <span class="ot">&lt;-</span> <span class="fu">perform_overlap_mantel_test</span>(</span>
<span id="cb120-82"><a href="#cb120-82" aria-hidden="true" tabindex="-1"></a>    <span class="at">overlap_matrix =</span> overlap_current_matrix_d, </span>
<span id="cb120-83"><a href="#cb120-83" aria-hidden="true" tabindex="-1"></a>    <span class="at">matrix_label =</span> <span class="st">"Current Anemonefish-Anemonefish Overlap"</span>,</span>
<span id="cb120-84"><a href="#cb120-84" aria-hidden="true" tabindex="-1"></a>    <span class="at">phy_dist_matrix =</span> fish_phy_dist_matrix,</span>
<span id="cb120-85"><a href="#cb120-85" aria-hidden="true" tabindex="-1"></a>    <span class="at">tree_tips_short_codes =</span> pruned_anemonefish_tree<span class="sc">$</span>tip.label,</span>
<span id="cb120-86"><a href="#cb120-86" aria-hidden="true" tabindex="-1"></a>    <span class="at">label_map_df =</span> fish_label_map_df</span>
<span id="cb120-87"><a href="#cb120-87" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb120-88"><a href="#cb120-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-89"><a href="#cb120-89" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --- 3. Future Anemonefish-Anemonefish Overlap Mantel Test (for a selected scenario) ---</span></span>
<span id="cb120-90"><a href="#cb120-90" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Choose ONE future scenario to test here, e.g., the "worst-case" or most interesting one.</span></span>
<span id="cb120-91"><a href="#cb120-91" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure this chosen_future_scenario_for_mantel was processed in the Schoener's D section.</span></span>
<span id="cb120-92"><a href="#cb120-92" aria-hidden="true" tabindex="-1"></a>  chosen_future_scenario_for_mantel <span class="ot">&lt;-</span> <span class="st">"ssp585_2100"</span> <span class="co"># Or "ssp119_2050", etc.</span></span>
<span id="cb120-93"><a href="#cb120-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-94"><a href="#cb120-94" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (chosen_future_scenario_for_mantel <span class="sc">%in%</span> <span class="fu">names</span>(overlap_future_matrices_d) <span class="sc">&amp;&amp;</span> </span>
<span id="cb120-95"><a href="#cb120-95" aria-hidden="true" tabindex="-1"></a>      <span class="sc">!</span><span class="fu">is.null</span>(overlap_future_matrices_d[[chosen_future_scenario_for_mantel]])) {</span>
<span id="cb120-96"><a href="#cb120-96" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-97"><a href="#cb120-97" aria-hidden="true" tabindex="-1"></a>    overlap_future_chosen_for_mantel <span class="ot">&lt;-</span> overlap_future_matrices_d[[chosen_future_scenario_for_mantel]]</span>
<span id="cb120-98"><a href="#cb120-98" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb120-99"><a href="#cb120-99" aria-hidden="true" tabindex="-1"></a>    mantel_future_fish_overlap <span class="ot">&lt;-</span> <span class="fu">perform_overlap_mantel_test</span>(</span>
<span id="cb120-100"><a href="#cb120-100" aria-hidden="true" tabindex="-1"></a>      <span class="at">overlap_matrix =</span> overlap_future_chosen_for_mantel,</span>
<span id="cb120-101"><a href="#cb120-101" aria-hidden="true" tabindex="-1"></a>      <span class="at">matrix_label =</span> <span class="fu">paste</span>(<span class="st">"Future ("</span>, chosen_future_scenario_for_mantel, <span class="st">") Anemonefish-Anemonefish Overlap"</span>),</span>
<span id="cb120-102"><a href="#cb120-102" aria-hidden="true" tabindex="-1"></a>      <span class="at">phy_dist_matrix =</span> fish_phy_dist_matrix,</span>
<span id="cb120-103"><a href="#cb120-103" aria-hidden="true" tabindex="-1"></a>      <span class="at">tree_tips_short_codes =</span> pruned_anemonefish_tree<span class="sc">$</span>tip.label,</span>
<span id="cb120-104"><a href="#cb120-104" aria-hidden="true" tabindex="-1"></a>      <span class="at">label_map_df =</span> fish_label_map_df</span>
<span id="cb120-105"><a href="#cb120-105" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb120-106"><a href="#cb120-106" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb120-107"><a href="#cb120-107" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Overlap data for future scenario '"</span>, chosen_future_scenario_for_mantel, <span class="st">"' not available for Mantel test.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb120-108"><a href="#cb120-108" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb120-109"><a href="#cb120-109" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Warning: Current overlap matrix or mapping to tree tips insufficient for anemonefish. Skipping phylogenetic signal of fish-fish overlap.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Phylogenetic signal of overlap analysis finished. ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Phylogenetic signal of overlap analysis finished. ---</code></pre>
</div>
</div>
<section id="levins-b-niche-breadth-current-predictions" class="level3">
<h3 class="anchored" data-anchor-id="levins-b-niche-breadth-current-predictions">Levinâ€™s B Niche Breadth (Current Predictions)</h3>
<p>This section calculates Levinâ€™s B (<span class="math inline">\(B_2\)</span>) niche breadth metric for each host sea anemone species and each anemonefish species (using their environmental-only model predictions for the current period). This metric quantifies the uniformity of a speciesâ€™ predicted suitability across its range. We then compare the mean niche breadth between the two guilds.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ENMTools) <span class="co"># For raster.breadth</span></span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensure the current individual prediction stacks are available and cropped</span></span>
<span id="cb124-5"><a href="#cb124-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">is.null</span>(host_pred_stack_cropped) <span class="sc">||</span> <span class="sc">!</span><span class="fu">inherits</span>(host_pred_stack_cropped, <span class="st">"SpatRaster"</span>)) {</span>
<span id="cb124-6"><a href="#cb124-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"`host_pred_stack_cropped` is not available or not a SpatRaster. Run previous chunks."</span>)</span>
<span id="cb124-7"><a href="#cb124-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb124-8"><a href="#cb124-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">is.null</span>(fish_pred_stack_cropped) <span class="sc">||</span> <span class="sc">!</span><span class="fu">inherits</span>(fish_pred_stack_cropped, <span class="st">"SpatRaster"</span>)) {</span>
<span id="cb124-9"><a href="#cb124-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"`fish_pred_stack_cropped` (for env-only fish models) is not available or not a SpatRaster. Run previous chunks."</span>)</span>
<span id="cb124-10"><a href="#cb124-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb124-11"><a href="#cb124-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-12"><a href="#cb124-12" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 1. Calculate Niche Breadth for Host Sea Anemones ---</span></span>
<span id="cb124-13"><a href="#cb124-13" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"--- Calculating Levin's B Niche Breadth for Host Sea Anemones (Current) ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Calculating Levin's B Niche Breadth for Host Sea Anemones (Current) ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>host_niche_breadth_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">species_name_sanitized =</span> <span class="fu">character</span>(<span class="dv">0</span>), </span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">levins_B1 =</span> <span class="fu">numeric</span>(<span class="dv">0</span>), </span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">levins_B2 =</span> <span class="fu">numeric</span>(<span class="dv">0</span>),</span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb126-6"><a href="#cb126-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb126-7"><a href="#cb126-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb126-8"><a href="#cb126-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (terra<span class="sc">::</span><span class="fu">nlyr</span>(host_pred_stack_cropped) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb126-9"><a href="#cb126-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>terra<span class="sc">::</span><span class="fu">nlyr</span>(host_pred_stack_cropped)) {</span>
<span id="cb126-10"><a href="#cb126-10" aria-hidden="true" tabindex="-1"></a>    species_raster_host <span class="ot">&lt;-</span> host_pred_stack_cropped[[i]]</span>
<span id="cb126-11"><a href="#cb126-11" aria-hidden="true" tabindex="-1"></a>    sp_name_host <span class="ot">&lt;-</span> <span class="fu">names</span>(species_raster_host)</span>
<span id="cb126-12"><a href="#cb126-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Calculating breadth for host:"</span>, sp_name_host, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb126-13"><a href="#cb126-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(species_raster_host, <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"Suitability - Host:"</span>, sp_name_host)) <span class="co"># Optional plot</span></span>
<span id="cb126-14"><a href="#cb126-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb126-15"><a href="#cb126-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># raster.breadth expects a SpatRaster, which species_raster_host should be</span></span>
<span id="cb126-16"><a href="#cb126-16" aria-hidden="true" tabindex="-1"></a>    breadth_metrics_host <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb126-17"><a href="#cb126-17" aria-hidden="true" tabindex="-1"></a>      ENMTools<span class="sc">::</span><span class="fu">raster.breadth</span>(species_raster_host)</span>
<span id="cb126-18"><a href="#cb126-18" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb126-19"><a href="#cb126-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Could not calculate breadth for host"</span>, sp_name_host, <span class="st">":"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb126-20"><a href="#cb126-20" aria-hidden="true" tabindex="-1"></a>      <span class="cn">NULL</span></span>
<span id="cb126-21"><a href="#cb126-21" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb126-22"><a href="#cb126-22" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb126-23"><a href="#cb126-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(breadth_metrics_host)) {</span>
<span id="cb126-24"><a href="#cb126-24" aria-hidden="true" tabindex="-1"></a>      host_niche_breadth_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(host_niche_breadth_df, </span>
<span id="cb126-25"><a href="#cb126-25" aria-hidden="true" tabindex="-1"></a>                                     <span class="fu">data.frame</span>(<span class="at">species_name_sanitized =</span> sp_name_host, </span>
<span id="cb126-26"><a href="#cb126-26" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">levins_B1 =</span> breadth_metrics_host<span class="sc">$</span>B1, </span>
<span id="cb126-27"><a href="#cb126-27" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">levins_B2 =</span> breadth_metrics_host<span class="sc">$</span>B2))</span>
<span id="cb126-28"><a href="#cb126-28" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb126-29"><a href="#cb126-29" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb126-30"><a href="#cb126-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Finished calculating niche breadth for hosts.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb126-31"><a href="#cb126-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(host_niche_breadth_df)</span>
<span id="cb126-32"><a href="#cb126-32" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb126-33"><a href="#cb126-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"No host anemone prediction layers to calculate breadth for.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb126-34"><a href="#cb126-34" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Entacmaea_quadricolor </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Radianthus_crispa </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-2.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Radianthus_magnifica </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-3.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Stichodactyla_gigantea </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-4.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Cryptodendrum_adhaesivum </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-5.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Macrodactyla_doreensis </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-6.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Stichodactyla_mertensii </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-7.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Stichodactyla_haddoni </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-8.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for host: Radianthus_malu </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-9.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Finished calculating niche breadth for hosts.
    species_name_sanitized levins_B1  levins_B2
1    Entacmaea_quadricolor 0.8478375 0.07389608
2        Radianthus_crispa 0.8566194 0.08627583
3     Radianthus_magnifica 0.8425530 0.07226086
4   Stichodactyla_gigantea 0.8444508 0.06880579
5 Cryptodendrum_adhaesivum 0.8619131 0.09297080
6   Macrodactyla_doreensis 0.8641089 0.09668081
7  Stichodactyla_mertensii 0.8466158 0.07373661
8    Stichodactyla_haddoni 0.8486856 0.07867407
9          Radianthus_malu 0.8624786 0.08777046</code></pre>
</div>
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 2. Calculate Niche Breadth for Anemonefish (Environmental-Only Models) ---</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Calculating Levin's B Niche Breadth for Anemonefish (Env-Only, Current) ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Calculating Levin's B Niche Breadth for Anemonefish (Env-Only, Current) ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>fish_niche_breadth_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">species_name_sanitized =</span> <span class="fu">character</span>(<span class="dv">0</span>), </span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">levins_B1 =</span> <span class="fu">numeric</span>(<span class="dv">0</span>), </span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">levins_B2 =</span> <span class="fu">numeric</span>(<span class="dv">0</span>),</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (terra<span class="sc">::</span><span class="fu">nlyr</span>(fish_pred_stack_cropped) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb139-9"><a href="#cb139-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>terra<span class="sc">::</span><span class="fu">nlyr</span>(fish_pred_stack_cropped)) {</span>
<span id="cb139-10"><a href="#cb139-10" aria-hidden="true" tabindex="-1"></a>    species_raster_fish <span class="ot">&lt;-</span> fish_pred_stack_cropped[[i]]</span>
<span id="cb139-11"><a href="#cb139-11" aria-hidden="true" tabindex="-1"></a>    sp_name_fish <span class="ot">&lt;-</span> <span class="fu">names</span>(species_raster_fish)</span>
<span id="cb139-12"><a href="#cb139-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Calculating breadth for anemonefish:"</span>, sp_name_fish, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb139-13"><a href="#cb139-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(species_raster_fish, <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"Suitability - Fish (Env-Only):"</span>, sp_name_fish)) <span class="co"># Optional plot</span></span>
<span id="cb139-14"><a href="#cb139-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-15"><a href="#cb139-15" aria-hidden="true" tabindex="-1"></a>    breadth_metrics_fish <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb139-16"><a href="#cb139-16" aria-hidden="true" tabindex="-1"></a>      ENMTools<span class="sc">::</span><span class="fu">raster.breadth</span>(species_raster_fish)</span>
<span id="cb139-17"><a href="#cb139-17" aria-hidden="true" tabindex="-1"></a>    }, <span class="at">error =</span> <span class="cf">function</span>(e) {</span>
<span id="cb139-18"><a href="#cb139-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Could not calculate breadth for fish"</span>, sp_name_fish, <span class="st">":"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb139-19"><a href="#cb139-19" aria-hidden="true" tabindex="-1"></a>      <span class="cn">NULL</span></span>
<span id="cb139-20"><a href="#cb139-20" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb139-21"><a href="#cb139-21" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb139-22"><a href="#cb139-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(breadth_metrics_fish)) {</span>
<span id="cb139-23"><a href="#cb139-23" aria-hidden="true" tabindex="-1"></a>      fish_niche_breadth_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(fish_niche_breadth_df, </span>
<span id="cb139-24"><a href="#cb139-24" aria-hidden="true" tabindex="-1"></a>                                     <span class="fu">data.frame</span>(<span class="at">species_name_sanitized =</span> sp_name_fish, </span>
<span id="cb139-25"><a href="#cb139-25" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">levins_B1 =</span> breadth_metrics_fish<span class="sc">$</span>B1, </span>
<span id="cb139-26"><a href="#cb139-26" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">levins_B2 =</span> breadth_metrics_fish<span class="sc">$</span>B2))</span>
<span id="cb139-27"><a href="#cb139-27" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb139-28"><a href="#cb139-28" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb139-29"><a href="#cb139-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Finished calculating niche breadth for anemonefish (env-only).</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb139-30"><a href="#cb139-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(fish_niche_breadth_df)</span>
<span id="cb139-31"><a href="#cb139-31" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb139-32"><a href="#cb139-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"No anemonefish (env-only) prediction layers to calculate breadth for.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb139-33"><a href="#cb139-33" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for anemonefish: Amphiprion_clarkii </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-10.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for anemonefish: Amphiprion_ocellaris </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-11.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for anemonefish: Amphiprion_perideraion </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-12.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for anemonefish: Amphiprion_polymnus </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-13.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating breadth for anemonefish: Amphiprion_sandaracinos </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/levin-b-niche-breadth-14.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Finished calculating niche breadth for anemonefish (env-only).
   species_name_sanitized levins_B1  levins_B2
1      Amphiprion_clarkii 0.8593767 0.09264915
2    Amphiprion_ocellaris 0.8465541 0.08104621
3  Amphiprion_perideraion 0.8439280 0.06941610
4     Amphiprion_polymnus 0.8772922 0.11679409
5 Amphiprion_sandaracinos 0.8598594 0.09208923</code></pre>
</div>
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 3. Compare Mean Niche Breadth (Levin's B2) ---</span></span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">nrow</span>(host_niche_breadth_df) <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(fish_niche_breadth_df) <span class="sc">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Comparing Mean Niche Breadth (Levin's B2) between Guilds ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Perform t-test if both groups have data for B2</span></span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="st">"levins_B2"</span> <span class="sc">%in%</span> <span class="fu">names</span>(host_niche_breadth_df) <span class="sc">&amp;&amp;</span> <span class="st">"levins_B2"</span> <span class="sc">%in%</span> <span class="fu">names</span>(fish_niche_breadth_df)) {</span>
<span id="cb146-7"><a href="#cb146-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb146-8"><a href="#cb146-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Remove rows with NA in levins_B2 for a fair comparison</span></span>
<span id="cb146-9"><a href="#cb146-9" aria-hidden="true" tabindex="-1"></a>    valid_hosts_b2 <span class="ot">&lt;-</span> host_niche_breadth_df<span class="sc">$</span>levins_B2[<span class="sc">!</span><span class="fu">is.na</span>(host_niche_breadth_df<span class="sc">$</span>levins_B2)]</span>
<span id="cb146-10"><a href="#cb146-10" aria-hidden="true" tabindex="-1"></a>    valid_fish_b2 <span class="ot">&lt;-</span> fish_niche_breadth_df<span class="sc">$</span>levins_B2[<span class="sc">!</span><span class="fu">is.na</span>(fish_niche_breadth_df<span class="sc">$</span>levins_B2)]</span>
<span id="cb146-11"><a href="#cb146-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb146-12"><a href="#cb146-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(valid_hosts_b2) <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">&amp;&amp;</span> <span class="fu">length</span>(valid_fish_b2) <span class="sc">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb146-13"><a href="#cb146-13" aria-hidden="true" tabindex="-1"></a>      ttest_breadth_b2 <span class="ot">&lt;-</span> <span class="fu">t.test</span>(valid_hosts_b2, valid_fish_b2)</span>
<span id="cb146-14"><a href="#cb146-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Welch Two Sample t-test for Levin's B2 (Hosts vs. Fish Env-Only):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb146-15"><a href="#cb146-15" aria-hidden="true" tabindex="-1"></a>      <span class="fu">print</span>(ttest_breadth_b2)</span>
<span id="cb146-16"><a href="#cb146-16" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb146-17"><a href="#cb146-17" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"Not enough valid B2 values for t-test after removing NAs.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb146-18"><a href="#cb146-18" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb146-19"><a href="#cb146-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb146-20"><a href="#cb146-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Standard Deviation of Levin's B2:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb146-21"><a href="#cb146-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Hosts:"</span>, <span class="fu">sd</span>(valid_hosts_b2, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb146-22"><a href="#cb146-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Anemonefish (Env-Only):"</span>, <span class="fu">sd</span>(valid_fish_b2, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb146-23"><a href="#cb146-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb146-24"><a href="#cb146-24" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb146-25"><a href="#cb146-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"Levin's B2 column not found in one or both breadth data frames.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb146-26"><a href="#cb146-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb146-27"><a href="#cb146-27" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb146-28"><a href="#cb146-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Not enough data to compare niche breadths between hosts and anemonefish.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb146-29"><a href="#cb146-29" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Comparing Mean Niche Breadth (Levin's B2) between Guilds ---
Welch Two Sample t-test for Levin's B2 (Hosts vs. Fish Env-Only):

    Welch Two Sample t-test

data:  valid_hosts_b2 and valid_fish_b2
t = -1.0755, df = 5.4761, p-value = 0.3272
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -0.03052215  0.01218453
sample estimates:
 mean of x  mean of y 
0.08123015 0.09039896 


Standard Deviation of Levin's B2:
  Hosts: 0.009975523 
  Anemonefish (Env-Only): 0.01755262 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The data frames 'host_niche_breadth_df' and 'fish_niche_breadth_df' are now available.</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a><span class="co"># They contain species names and their Levin's B1 and B2 values.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="correlation-between-niche-breadth-and-suitability-shifts" class="level2">
<h2 class="anchored" data-anchor-id="correlation-between-niche-breadth-and-suitability-shifts">Correlation between Niche Breadth and Suitability Shifts</h2>
<p>This section investigates whether a speciesâ€™ current niche breadth (Levinâ€™s <span class="math inline">\(B_2\)</span>) is correlated with its projected mean change in environmental suitability under future climate scenarios.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr) <span class="co"># For pivot_wider if needed</span></span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb149-5"><a href="#cb149-5" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Ensure prerequisite data is available ---</span></span>
<span id="cb149-6"><a href="#cb149-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"host_niche_breadth_df"</span>) <span class="sc">||</span> <span class="sc">!</span><span class="fu">exists</span>(<span class="st">"fish_niche_breadth_df"</span>)) {</span>
<span id="cb149-7"><a href="#cb149-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"Niche breadth data frames ('host_niche_breadth_df', 'fish_niche_breadth_df') not found. Run previous chunk."</span>)</span>
<span id="cb149-8"><a href="#cb149-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb149-9"><a href="#cb149-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"suitability_shifts_list"</span>) <span class="sc">||</span> <span class="sc">!</span><span class="fu">is.list</span>(suitability_shifts_list) <span class="sc">||</span> <span class="fu">length</span>(suitability_shifts_list) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb149-10"><a href="#cb149-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"`suitability_shifts_list` not found or empty. Run 'Individual Species Suitability Shifts' chunk."</span>)</span>
<span id="cb149-11"><a href="#cb149-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb149-12"><a href="#cb149-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"future_scenarios_to_load"</span>) <span class="sc">||</span> <span class="fu">length</span>(future_scenarios_to_load) <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb149-13"><a href="#cb149-13" aria-hidden="true" tabindex="-1"></a>  future_scenarios_to_load <span class="ot">&lt;-</span> config<span class="sc">$</span>env_scenarios[config<span class="sc">$</span>env_scenarios <span class="sc">!=</span> <span class="st">"current"</span>]</span>
<span id="cb149-14"><a href="#cb149-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(future_scenarios_to_load) <span class="sc">==</span> <span class="dv">0</span>) <span class="fu">stop</span>(<span class="st">"No future scenarios defined."</span>)</span>
<span id="cb149-15"><a href="#cb149-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb149-16"><a href="#cb149-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb149-17"><a href="#cb149-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb149-18"><a href="#cb149-18" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 1. Prepare Host Anemone Data (Breadth and Shifts) ---</span></span>
<span id="cb149-19"><a href="#cb149-19" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"--- Processing Host Anemone Breadth vs. Suitability Shift ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Processing Host Anemone Breadth vs. Suitability Shift ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">nrow</span>(host_niche_breadth_df) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Start with breadth data</span></span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>  host_breadth_shifts_df <span class="ot">&lt;-</span> host_niche_breadth_df <span class="sc">%&gt;%</span></span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(species_name_sanitized, <span class="at">breadth_B2 =</span> levins_B2)</span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add shift data for each future scenario</span></span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (scenario_name <span class="cf">in</span> future_scenarios_to_load) {</span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>    shift_data_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"hosts_"</span>, scenario_name)</span>
<span id="cb151-9"><a href="#cb151-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (shift_data_name <span class="sc">%in%</span> <span class="fu">names</span>(suitability_shifts_list) <span class="sc">&amp;&amp;</span> </span>
<span id="cb151-10"><a href="#cb151-10" aria-hidden="true" tabindex="-1"></a>        <span class="sc">!</span><span class="fu">is.null</span>(suitability_shifts_list[[shift_data_name]]) <span class="sc">&amp;&amp;</span></span>
<span id="cb151-11"><a href="#cb151-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(suitability_shifts_list[[shift_data_name]]) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb151-12"><a href="#cb151-12" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb151-13"><a href="#cb151-13" aria-hidden="true" tabindex="-1"></a>      temp_shift_df <span class="ot">&lt;-</span> suitability_shifts_list[[shift_data_name]] <span class="sc">%&gt;%</span></span>
<span id="cb151-14"><a href="#cb151-14" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> <span class="co"># Ensure it's a data frame</span></span>
<span id="cb151-15"><a href="#cb151-15" aria-hidden="true" tabindex="-1"></a>        tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"species_name_sanitized"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb151-16"><a href="#cb151-16" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(species_name_sanitized, <span class="at">mean_suitability_shift =</span> mean)</span>
<span id="cb151-17"><a href="#cb151-17" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb151-18"><a href="#cb151-18" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Dynamically create column name for the shift in this scenario</span></span>
<span id="cb151-19"><a href="#cb151-19" aria-hidden="true" tabindex="-1"></a>      shift_col_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"shift_"</span>, <span class="fu">gsub</span>(<span class="st">"ssp"</span>, <span class="st">""</span>, scenario_name)) <span class="co"># e.g., shift_119_2050</span></span>
<span id="cb151-20"><a href="#cb151-20" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb151-21"><a href="#cb151-21" aria-hidden="true" tabindex="-1"></a>      host_breadth_shifts_df <span class="ot">&lt;-</span> host_breadth_shifts_df <span class="sc">%&gt;%</span></span>
<span id="cb151-22"><a href="#cb151-22" aria-hidden="true" tabindex="-1"></a>        <span class="fu">left_join</span>(temp_shift_df <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="sc">!!</span><span class="at">shift_col_name :=</span> mean_suitability_shift), </span>
<span id="cb151-23"><a href="#cb151-23" aria-hidden="true" tabindex="-1"></a>                  <span class="at">by =</span> <span class="st">"species_name_sanitized"</span>)</span>
<span id="cb151-24"><a href="#cb151-24" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb151-25"><a href="#cb151-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Shift data for hosts in scenario '"</span>, scenario_name, <span class="st">"' not found or empty.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb151-26"><a href="#cb151-26" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb151-27"><a href="#cb151-27" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb151-28"><a href="#cb151-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb151-29"><a href="#cb151-29" aria-hidden="true" tabindex="-1"></a>  host_breadth_shifts_df <span class="ot">&lt;-</span> host_breadth_shifts_df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(breadth_B2))</span>
<span id="cb151-30"><a href="#cb151-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Host anemone data prepared for breadth vs. shift analysis:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb151-31"><a href="#cb151-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(host_breadth_shifts_df))</span>
<span id="cb151-32"><a href="#cb151-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-33"><a href="#cb151-33" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Perform correlations for each future scenario for hosts</span></span>
<span id="cb151-34"><a href="#cb151-34" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (scenario_name <span class="cf">in</span> future_scenarios_to_load) {</span>
<span id="cb151-35"><a href="#cb151-35" aria-hidden="true" tabindex="-1"></a>    shift_col_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"shift_"</span>, <span class="fu">gsub</span>(<span class="st">"ssp"</span>, <span class="st">""</span>, scenario_name))</span>
<span id="cb151-36"><a href="#cb151-36" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (shift_col_name <span class="sc">%in%</span> <span class="fu">names</span>(host_breadth_shifts_df)) {</span>
<span id="cb151-37"><a href="#cb151-37" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Ensure there are enough non-NA pairs for correlation</span></span>
<span id="cb151-38"><a href="#cb151-38" aria-hidden="true" tabindex="-1"></a>      temp_df_cor_host <span class="ot">&lt;-</span> host_breadth_shifts_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(breadth_B2, <span class="sc">!!</span><span class="fu">sym</span>(shift_col_name)) <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()</span>
<span id="cb151-39"><a href="#cb151-39" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="fu">nrow</span>(temp_df_cor_host) <span class="sc">&gt;</span> <span class="dv">2</span>) {</span>
<span id="cb151-40"><a href="#cb151-40" aria-hidden="true" tabindex="-1"></a>        cor_test_host <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(temp_df_cor_host<span class="sc">$</span>breadth_B2, temp_df_cor_host[[shift_col_name]], <span class="at">method =</span> <span class="st">"pearson"</span>)</span>
<span id="cb151-41"><a href="#cb151-41" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Correlation (Host Anemones) - Breadth vs. Shift ("</span>, scenario_name, <span class="st">"):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb151-42"><a href="#cb151-42" aria-hidden="true" tabindex="-1"></a>        <span class="fu">print</span>(cor_test_host)</span>
<span id="cb151-43"><a href="#cb151-43" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb151-44"><a href="#cb151-44" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Not enough data points to calculate correlation for hosts in scenario:"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb151-45"><a href="#cb151-45" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb151-46"><a href="#cb151-46" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb151-47"><a href="#cb151-47" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb151-48"><a href="#cb151-48" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb151-49"><a href="#cb151-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"No host niche breadth data available.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb151-50"><a href="#cb151-50" aria-hidden="true" tabindex="-1"></a>  host_breadth_shifts_df <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb151-51"><a href="#cb151-51" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Host anemone data prepared for breadth vs. shift analysis:
    species_name_sanitized breadth_B2 shift_119_2050 shift_119_2100
1    Entacmaea_quadricolor 0.07389608   -0.002565663  -0.0019356642
2        Radianthus_crispa 0.08627583   -0.004331801  -0.0041563170
3     Radianthus_magnifica 0.07226086   -0.000731002  -0.0001000172
4   Stichodactyla_gigantea 0.06880579    0.001831661   0.0025015008
5 Cryptodendrum_adhaesivum 0.09297080   -0.003191857  -0.0026667014
6   Macrodactyla_doreensis 0.09668081   -0.002246241  -0.0014202995
  shift_585_2050 shift_585_2100
1  -0.0027426973    0.002040631
2  -0.0039582078   -0.001187740
3  -0.0009430532    0.001665024
4   0.0015206267    0.003224462
5  -0.0033539312   -0.005392539
6  -0.0027050459   -0.006258454

Correlation (Host Anemones) - Breadth vs. Shift ( ssp119_2050 ):

    Pearson's product-moment correlation

data:  temp_df_cor_host$breadth_B2 and temp_df_cor_host[[shift_col_name]]
t = -0.87913, df = 7, p-value = 0.4085
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.8098537  0.4411841
sample estimates:
      cor 
-0.315329 


Correlation (Host Anemones) - Breadth vs. Shift ( ssp119_2100 ):

    Pearson's product-moment correlation

data:  temp_df_cor_host$breadth_B2 and temp_df_cor_host[[shift_col_name]]
t = -0.86244, df = 7, p-value = 0.417
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.8077813  0.4459978
sample estimates:
       cor 
-0.3099218 


Correlation (Host Anemones) - Breadth vs. Shift ( ssp585_2050 ):

    Pearson's product-moment correlation

data:  temp_df_cor_host$breadth_B2 and temp_df_cor_host[[shift_col_name]]
t = -0.77316, df = 7, p-value = 0.4647
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.7962833  0.4714489
sample estimates:
       cor 
-0.2804959 


Correlation (Host Anemones) - Breadth vs. Shift ( ssp585_2100 ):

    Pearson's product-moment correlation

data:  temp_df_cor_host$breadth_B2 and temp_df_cor_host[[shift_col_name]]
t = -1.3697, df = 7, p-value = 0.2131
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.8609827  0.2942139
sample estimates:
       cor 
-0.4597378 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 2. Prepare Anemonefish (Env-Only) Data (Breadth and Shifts) ---</span></span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Processing Anemonefish (Env-Only) Breadth vs. Suitability Shift ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Processing Anemonefish (Env-Only) Breadth vs. Suitability Shift ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">nrow</span>(fish_niche_breadth_df) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a>  fish_breadth_shifts_df <span class="ot">&lt;-</span> fish_niche_breadth_df <span class="sc">%&gt;%</span></span>
<span id="cb155-3"><a href="#cb155-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(species_name_sanitized, <span class="at">breadth_B2 =</span> levins_B2)</span>
<span id="cb155-4"><a href="#cb155-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb155-5"><a href="#cb155-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (scenario_name <span class="cf">in</span> future_scenarios_to_load) {</span>
<span id="cb155-6"><a href="#cb155-6" aria-hidden="true" tabindex="-1"></a>    shift_data_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"fish_env_only_"</span>, scenario_name)</span>
<span id="cb155-7"><a href="#cb155-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (shift_data_name <span class="sc">%in%</span> <span class="fu">names</span>(suitability_shifts_list) <span class="sc">&amp;&amp;</span> </span>
<span id="cb155-8"><a href="#cb155-8" aria-hidden="true" tabindex="-1"></a>        <span class="sc">!</span><span class="fu">is.null</span>(suitability_shifts_list[[shift_data_name]]) <span class="sc">&amp;&amp;</span></span>
<span id="cb155-9"><a href="#cb155-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(suitability_shifts_list[[shift_data_name]]) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb155-10"><a href="#cb155-10" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb155-11"><a href="#cb155-11" aria-hidden="true" tabindex="-1"></a>      temp_shift_df_fish <span class="ot">&lt;-</span> suitability_shifts_list[[shift_data_name]] <span class="sc">%&gt;%</span></span>
<span id="cb155-12"><a href="#cb155-12" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb155-13"><a href="#cb155-13" aria-hidden="true" tabindex="-1"></a>        tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"species_name_sanitized"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb155-14"><a href="#cb155-14" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(species_name_sanitized, <span class="at">mean_suitability_shift =</span> mean)</span>
<span id="cb155-15"><a href="#cb155-15" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb155-16"><a href="#cb155-16" aria-hidden="true" tabindex="-1"></a>      shift_col_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"shift_"</span>, <span class="fu">gsub</span>(<span class="st">"ssp"</span>, <span class="st">""</span>, scenario_name))</span>
<span id="cb155-17"><a href="#cb155-17" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb155-18"><a href="#cb155-18" aria-hidden="true" tabindex="-1"></a>      fish_breadth_shifts_df <span class="ot">&lt;-</span> fish_breadth_shifts_df <span class="sc">%&gt;%</span></span>
<span id="cb155-19"><a href="#cb155-19" aria-hidden="true" tabindex="-1"></a>        <span class="fu">left_join</span>(temp_shift_df_fish <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="sc">!!</span><span class="at">shift_col_name :=</span> mean_suitability_shift), </span>
<span id="cb155-20"><a href="#cb155-20" aria-hidden="true" tabindex="-1"></a>                  <span class="at">by =</span> <span class="st">"species_name_sanitized"</span>)</span>
<span id="cb155-21"><a href="#cb155-21" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb155-22"><a href="#cb155-22" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"  Warning: Shift data for anemonefish (env-only) in scenario '"</span>, scenario_name, <span class="st">"' not found or empty.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb155-23"><a href="#cb155-23" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb155-24"><a href="#cb155-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb155-25"><a href="#cb155-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb155-26"><a href="#cb155-26" aria-hidden="true" tabindex="-1"></a>  fish_breadth_shifts_df <span class="ot">&lt;-</span> fish_breadth_shifts_df <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(breadth_B2))</span>
<span id="cb155-27"><a href="#cb155-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Anemonefish (env-only) data prepared for breadth vs. shift analysis:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb155-28"><a href="#cb155-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(fish_breadth_shifts_df))</span>
<span id="cb155-29"><a href="#cb155-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb155-30"><a href="#cb155-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Perform correlations for each future scenario for fish</span></span>
<span id="cb155-31"><a href="#cb155-31" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (scenario_name <span class="cf">in</span> future_scenarios_to_load) {</span>
<span id="cb155-32"><a href="#cb155-32" aria-hidden="true" tabindex="-1"></a>    shift_col_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"shift_"</span>, <span class="fu">gsub</span>(<span class="st">"ssp"</span>, <span class="st">""</span>, scenario_name))</span>
<span id="cb155-33"><a href="#cb155-33" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (shift_col_name <span class="sc">%in%</span> <span class="fu">names</span>(fish_breadth_shifts_df)) {</span>
<span id="cb155-34"><a href="#cb155-34" aria-hidden="true" tabindex="-1"></a>      temp_df_cor_fish <span class="ot">&lt;-</span> fish_breadth_shifts_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(breadth_B2, <span class="sc">!!</span><span class="fu">sym</span>(shift_col_name)) <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()</span>
<span id="cb155-35"><a href="#cb155-35" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="fu">nrow</span>(temp_df_cor_fish) <span class="sc">&gt;</span> <span class="dv">2</span>){</span>
<span id="cb155-36"><a href="#cb155-36" aria-hidden="true" tabindex="-1"></a>        cor_test_fish <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(temp_df_cor_fish<span class="sc">$</span>breadth_B2, temp_df_cor_fish[[shift_col_name]], <span class="at">method =</span> <span class="st">"pearson"</span>)</span>
<span id="cb155-37"><a href="#cb155-37" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Correlation (Anemonefish Env-Only) - Breadth vs. Shift ("</span>, scenario_name, <span class="st">"):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb155-38"><a href="#cb155-38" aria-hidden="true" tabindex="-1"></a>        <span class="fu">print</span>(cor_test_fish)</span>
<span id="cb155-39"><a href="#cb155-39" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb155-40"><a href="#cb155-40" aria-hidden="true" tabindex="-1"></a>         <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Not enough data points to calculate correlation for fish in scenario:"</span>, scenario_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb155-41"><a href="#cb155-41" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb155-42"><a href="#cb155-42" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb155-43"><a href="#cb155-43" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb155-44"><a href="#cb155-44" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb155-45"><a href="#cb155-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"No anemonefish (env-only) niche breadth data available.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb155-46"><a href="#cb155-46" aria-hidden="true" tabindex="-1"></a>  fish_breadth_shifts_df <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb155-47"><a href="#cb155-47" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anemonefish (env-only) data prepared for breadth vs. shift analysis:
   species_name_sanitized breadth_B2 shift_119_2050 shift_119_2100
1      Amphiprion_clarkii 0.09264915  -0.0021256043  -0.0017270964
2    Amphiprion_ocellaris 0.08104621   0.0008215227   0.0016161519
3  Amphiprion_perideraion 0.06941610  -0.0058780502  -0.0053946167
4     Amphiprion_polymnus 0.11679409  -0.0036867805  -0.0031296265
5 Amphiprion_sandaracinos 0.09208923  -0.0014726434  -0.0008689387
  shift_585_2050 shift_585_2100
1  -0.0015740219   0.0057458900
2   0.0004199413   0.0019575836
3  -0.0060101212   0.0007593765
4  -0.0034570267  -0.0039715203
5  -0.0015338029  -0.0029538589

Correlation (Anemonefish Env-Only) - Breadth vs. Shift ( ssp119_2050 ):

    Pearson's product-moment correlation

data:  temp_df_cor_fish$breadth_B2 and temp_df_cor_fish[[shift_col_name]]
t = 0.10936, df = 3, p-value = 0.9198
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.8674809  0.8954946
sample estimates:
       cor 
0.06301095 


Correlation (Anemonefish Env-Only) - Breadth vs. Shift ( ssp119_2100 ):

    Pearson's product-moment correlation

data:  temp_df_cor_fish$breadth_B2 and temp_df_cor_fish[[shift_col_name]]
t = 0.095891, df = 3, p-value = 0.9297
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.8693884  0.8939467
sample estimates:
       cor 
0.05527802 


Correlation (Anemonefish Env-Only) - Breadth vs. Shift ( ssp585_2050 ):

    Pearson's product-moment correlation

data:  temp_df_cor_fish$breadth_B2 and temp_df_cor_fish[[shift_col_name]]
t = 0.25495, df = 3, p-value = 0.8152
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.8452365  0.9108643
sample estimates:
      cor 
0.1456289 


Correlation (Anemonefish Env-Only) - Breadth vs. Shift ( ssp585_2100 ):

    Pearson's product-moment correlation

data:  temp_df_cor_fish$breadth_B2 and temp_df_cor_fish[[shift_col_name]]
t = -0.93881, df = 3, p-value = 0.4171
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.9566113  0.7000625
sample estimates:
       cor 
-0.4765266 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- 3. Prepare data for combined plotting (similar to ant paper's all_bre) ---</span></span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This creates a long-format data frame for easier ggplotting with facets or colors</span></span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a>all_breadth_shifts_long_df <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-5"><a href="#cb157-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(host_breadth_shifts_df) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(host_breadth_shifts_df) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb157-6"><a href="#cb157-6" aria-hidden="true" tabindex="-1"></a>  host_plot_df <span class="ot">&lt;-</span> host_breadth_shifts_df <span class="sc">%&gt;%</span> </span>
<span id="cb157-7"><a href="#cb157-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">starts_with</span>(<span class="st">"shift_"</span>), <span class="at">names_to =</span> <span class="st">"scenario_shift_code"</span>, <span class="at">values_to =</span> <span class="st">"suitability_shift"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb157-8"><a href="#cb157-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">guild =</span> <span class="st">"Host_Anemones"</span>)</span>
<span id="cb157-9"><a href="#cb157-9" aria-hidden="true" tabindex="-1"></a>  all_breadth_shifts_long_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(all_breadth_shifts_long_df, host_plot_df)</span>
<span id="cb157-10"><a href="#cb157-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb157-11"><a href="#cb157-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(fish_breadth_shifts_df) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(fish_breadth_shifts_df) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb157-12"><a href="#cb157-12" aria-hidden="true" tabindex="-1"></a>  fish_plot_df <span class="ot">&lt;-</span> fish_breadth_shifts_df <span class="sc">%&gt;%</span></span>
<span id="cb157-13"><a href="#cb157-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">starts_with</span>(<span class="st">"shift_"</span>), <span class="at">names_to =</span> <span class="st">"scenario_shift_code"</span>, <span class="at">values_to =</span> <span class="st">"suitability_shift"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb157-14"><a href="#cb157-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">guild =</span> <span class="st">"Anemonefish_EnvOnly"</span>)</span>
<span id="cb157-15"><a href="#cb157-15" aria-hidden="true" tabindex="-1"></a>  all_breadth_shifts_long_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(all_breadth_shifts_long_df, fish_plot_df)</span>
<span id="cb157-16"><a href="#cb157-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb157-17"><a href="#cb157-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb157-18"><a href="#cb157-18" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(all_breadth_shifts_long_df) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(all_breadth_shifts_long_df) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb157-19"><a href="#cb157-19" aria-hidden="true" tabindex="-1"></a>  all_breadth_shifts_long_df <span class="ot">&lt;-</span> all_breadth_shifts_long_df <span class="sc">%&gt;%</span></span>
<span id="cb157-20"><a href="#cb157-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(breadth_B2), <span class="sc">!</span><span class="fu">is.na</span>(suitability_shift)) <span class="sc">%&gt;%</span></span>
<span id="cb157-21"><a href="#cb157-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Recreate a more descriptive scenario name for plotting</span></span>
<span id="cb157-22"><a href="#cb157-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">scenario_display =</span> <span class="fu">str_replace_all</span>(scenario_shift_code, </span>
<span id="cb157-23"><a href="#cb157-23" aria-hidden="true" tabindex="-1"></a>                                             <span class="fu">c</span>(<span class="st">"shift_"</span> <span class="ot">=</span> <span class="st">"SSP"</span>, <span class="st">"_"</span> <span class="ot">=</span> <span class="st">"-"</span>, <span class="st">"dec50"</span> <span class="ot">=</span> <span class="st">"(2050)"</span>, <span class="st">"dec100"</span> <span class="ot">=</span> <span class="st">"(2100)"</span>))) </span>
<span id="cb157-24"><a href="#cb157-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb157-25"><a href="#cb157-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Combined long format data for plotting breadth vs. suitability shift:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb157-26"><a href="#cb157-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(all_breadth_shifts_long_df))</span>
<span id="cb157-27"><a href="#cb157-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb157-28"><a href="#cb157-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Plotting (example for all scenarios faceted by scenario and guild)</span></span>
<span id="cb157-29"><a href="#cb157-29" aria-hidden="true" tabindex="-1"></a>  plot_breadth_vs_shift <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(all_breadth_shifts_long_df, <span class="fu">aes</span>(<span class="at">x =</span> suitability_shift, <span class="at">y =</span> breadth_B2)) <span class="sc">+</span></span>
<span id="cb157-30"><a href="#cb157-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb157-31"><a href="#cb157-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">TRUE</span>, <span class="fu">aes</span>(<span class="at">color =</span> guild), <span class="at">linetype=</span><span class="st">"dashed"</span>) <span class="sc">+</span> <span class="co"># Add se and color by guild</span></span>
<span id="cb157-32"><a href="#cb157-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_grid</span>(guild <span class="sc">~</span> scenario_display, <span class="at">scales =</span> <span class="st">"free_x"</span>) <span class="sc">+</span> <span class="co"># Facet by guild and scenario</span></span>
<span id="cb157-33"><a href="#cb157-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb157-34"><a href="#cb157-34" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"Niche Breadth (Levin's B2) vs. Mean Suitability Shift"</span>,</span>
<span id="cb157-35"><a href="#cb157-35" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"Mean Change in Suitability (Future - Current)"</span>,</span>
<span id="cb157-36"><a href="#cb157-36" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Levin's B2 (Niche Breadth)"</span></span>
<span id="cb157-37"><a href="#cb157-37" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb157-38"><a href="#cb157-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb157-39"><a href="#cb157-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">8</span>),</span>
<span id="cb157-40"><a href="#cb157-40" aria-hidden="true" tabindex="-1"></a>          <span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb157-41"><a href="#cb157-41" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb157-42"><a href="#cb157-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(plot_breadth_vs_shift)</span>
<span id="cb157-43"><a href="#cb157-43" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb157-44"><a href="#cb157-44" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb157-45"><a href="#cb157-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Not enough combined data to plot breadth vs. suitability shifts.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb157-46"><a href="#cb157-46" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Combined long format data for plotting breadth vs. suitability shift:
# A tibble: 6 Ã— 6
  species_name_sanitized breadth_B2 scenario_shift_code suitability_shift guild 
  &lt;chr&gt;                       &lt;dbl&gt; &lt;chr&gt;                           &lt;dbl&gt; &lt;chr&gt; 
1 Entacmaea_quadricolor      0.0739 shift_119_2050               -0.00257 Host_â€¦
2 Entacmaea_quadricolor      0.0739 shift_119_2100               -0.00194 Host_â€¦
3 Entacmaea_quadricolor      0.0739 shift_585_2050               -0.00274 Host_â€¦
4 Entacmaea_quadricolor      0.0739 shift_585_2100                0.00204 Host_â€¦
5 Radianthus_crispa          0.0863 shift_119_2050               -0.00433 Host_â€¦
6 Radianthus_crispa          0.0863 shift_119_2100               -0.00416 Host_â€¦
# â„¹ 1 more variable: scenario_display &lt;chr&gt;</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/breadth-suitability-correlation-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The data frames `host_breadth_shifts_df`, `fish_breadth_shifts_df`, </span></span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a><span class="co"># and `all_breadth_shifts_long_df` are now available.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="figure-niche-breadth-vs.-suitability-shift" class="level3">
<h3 class="anchored" data-anchor-id="figure-niche-breadth-vs.-suitability-shift">Figure: Niche Breadth vs.&nbsp;Suitability Shift</h3>
<p>This figure visualizes the relationship between current niche breadth (Levinâ€™s <span class="math inline">\(B_2\)</span>) and the projected mean change in suitability under a specific future scenario for both host sea anemones and anemonefish (environmental-only models). This is analogous to Figure 5 in the reference desert ant paper.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot) <span class="co"># For plot_grid</span></span>
<span id="cb160-3"><a href="#cb160-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-4"><a href="#cb160-4" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Choose ONE future scenario for this figure ---</span></span>
<span id="cb160-5"><a href="#cb160-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This should match one of the columns created in host_breadth_shifts_df and fish_breadth_shifts_df</span></span>
<span id="cb160-6"><a href="#cb160-6" aria-hidden="true" tabindex="-1"></a><span class="co"># e.g., "shift_585_2100" (derived from "ssp585_2100")</span></span>
<span id="cb160-7"><a href="#cb160-7" aria-hidden="true" tabindex="-1"></a>chosen_scenario_for_figure <span class="ot">&lt;-</span> <span class="st">"shift_585_2100"</span> <span class="co"># CHANGE AS NEEDED</span></span>
<span id="cb160-8"><a href="#cb160-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-9"><a href="#cb160-9" aria-hidden="true" tabindex="-1"></a>plot_a_host <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb160-10"><a href="#cb160-10" aria-hidden="true" tabindex="-1"></a>plot_b_fish <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb160-11"><a href="#cb160-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb160-12"><a href="#cb160-12" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Host Anemones Plot ---</span></span>
<span id="cb160-13"><a href="#cb160-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(host_breadth_shifts_df) <span class="sc">&amp;&amp;</span> chosen_scenario_for_figure <span class="sc">%in%</span> <span class="fu">names</span>(host_breadth_shifts_df) <span class="sc">&amp;&amp;</span> <span class="st">"breadth_B2"</span> <span class="sc">%in%</span> <span class="fu">names</span>(host_breadth_shifts_df)) {</span>
<span id="cb160-14"><a href="#cb160-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb160-15"><a href="#cb160-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check correlation for the dashed line logic (significant vs. non-significant)</span></span>
<span id="cb160-16"><a href="#cb160-16" aria-hidden="true" tabindex="-1"></a>  cor_data_host <span class="ot">&lt;-</span> host_breadth_shifts_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(breadth_B2, <span class="sc">!!</span><span class="fu">sym</span>(chosen_scenario_for_figure)) <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()</span>
<span id="cb160-17"><a href="#cb160-17" aria-hidden="true" tabindex="-1"></a>  host_cor_test <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb160-18"><a href="#cb160-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">nrow</span>(cor_data_host) <span class="sc">&gt;</span> <span class="dv">2</span>) {</span>
<span id="cb160-19"><a href="#cb160-19" aria-hidden="true" tabindex="-1"></a>    host_cor_test <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(cor_data_host<span class="sc">$</span>breadth_B2, cor_data_host[[chosen_scenario_for_figure]])</span>
<span id="cb160-20"><a href="#cb160-20" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb160-21"><a href="#cb160-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb160-22"><a href="#cb160-22" aria-hidden="true" tabindex="-1"></a>  plot_a_host <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(host_breadth_shifts_df, <span class="fu">aes</span>(<span class="at">x =</span> .data[[chosen_scenario_for_figure]], <span class="at">y =</span> breadth_B2)) <span class="sc">+</span></span>
<span id="cb160-23"><a href="#cb160-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"dodgerblue3"</span>) <span class="sc">+</span></span>
<span id="cb160-24"><a href="#cb160-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, </span>
<span id="cb160-25"><a href="#cb160-25" aria-hidden="true" tabindex="-1"></a>                <span class="at">linetype =</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(host_cor_test) <span class="sc">&amp;&amp;</span> host_cor_test<span class="sc">$</span>p.value <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="st">"solid"</span> <span class="cf">else</span> <span class="st">"dashed"</span>, </span>
<span id="cb160-26"><a href="#cb160-26" aria-hidden="true" tabindex="-1"></a>                <span class="at">se =</span> <span class="cn">TRUE</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span> <span class="co"># Show SE, solid if p &lt; 0.05</span></span>
<span id="cb160-27"><a href="#cb160-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb160-28"><a href="#cb160-28" aria-hidden="true" tabindex="-1"></a>      <span class="co"># title = "Host Anemones",</span></span>
<span id="cb160-29"><a href="#cb160-29" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="fu">paste</span>(<span class="st">"Mean Suitability Shift ("</span>, <span class="fu">gsub</span>(<span class="st">"shift_"</span>, <span class="st">"SSP"</span>, chosen_scenario_for_figure), <span class="st">")"</span>),</span>
<span id="cb160-30"><a href="#cb160-30" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Environmental Breadth (Levin's B2)"</span></span>
<span id="cb160-31"><a href="#cb160-31" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb160-32"><a href="#cb160-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb160-33"><a href="#cb160-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(), <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>())</span>
<span id="cb160-34"><a href="#cb160-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb160-35"><a href="#cb160-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Host plot generated for"</span>, chosen_scenario_for_figure, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb160-36"><a href="#cb160-36" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(host_cor_test)) <span class="fu">print</span>(host_cor_test)</span>
<span id="cb160-37"><a href="#cb160-37" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb160-38"><a href="#cb160-38" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb160-39"><a href="#cb160-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Skipping host breadth vs. shift plot: data for"</span>, chosen_scenario_for_figure, <span class="st">"or breadth_B2 missing.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb160-40"><a href="#cb160-40" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Host plot generated for shift_585_2100 

    Pearson's product-moment correlation

data:  cor_data_host$breadth_B2 and cor_data_host[[chosen_scenario_for_figure]]
t = -1.3697, df = 7, p-value = 0.2131
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.8609827  0.2942139
sample estimates:
       cor 
-0.4597378 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Anemonefish (Env-Only) Plot ---</span></span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(fish_breadth_shifts_df) <span class="sc">&amp;&amp;</span> chosen_scenario_for_figure <span class="sc">%in%</span> <span class="fu">names</span>(fish_breadth_shifts_df) <span class="sc">&amp;&amp;</span> <span class="st">"breadth_B2"</span> <span class="sc">%in%</span> <span class="fu">names</span>(fish_breadth_shifts_df)) {</span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb162-4"><a href="#cb162-4" aria-hidden="true" tabindex="-1"></a>  cor_data_fish <span class="ot">&lt;-</span> fish_breadth_shifts_df <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(breadth_B2, <span class="sc">!!</span><span class="fu">sym</span>(chosen_scenario_for_figure)) <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()</span>
<span id="cb162-5"><a href="#cb162-5" aria-hidden="true" tabindex="-1"></a>  fish_cor_test <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb162-6"><a href="#cb162-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">nrow</span>(cor_data_fish) <span class="sc">&gt;</span> <span class="dv">2</span>) {</span>
<span id="cb162-7"><a href="#cb162-7" aria-hidden="true" tabindex="-1"></a>    fish_cor_test <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(cor_data_fish<span class="sc">$</span>breadth_B2, cor_data_fish[[chosen_scenario_for_figure]])</span>
<span id="cb162-8"><a href="#cb162-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb162-9"><a href="#cb162-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb162-10"><a href="#cb162-10" aria-hidden="true" tabindex="-1"></a>  plot_b_fish <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(fish_breadth_shifts_df, <span class="fu">aes</span>(<span class="at">x =</span> .data[[chosen_scenario_for_figure]], <span class="at">y =</span> breadth_B2)) <span class="sc">+</span></span>
<span id="cb162-11"><a href="#cb162-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"firebrick3"</span>) <span class="sc">+</span></span>
<span id="cb162-12"><a href="#cb162-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, </span>
<span id="cb162-13"><a href="#cb162-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">linetype =</span> <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(fish_cor_test) <span class="sc">&amp;&amp;</span> fish_cor_test<span class="sc">$</span>p.value <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="st">"solid"</span> <span class="cf">else</span> <span class="st">"dashed"</span>, </span>
<span id="cb162-14"><a href="#cb162-14" aria-hidden="true" tabindex="-1"></a>                <span class="at">se =</span> <span class="cn">TRUE</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb162-15"><a href="#cb162-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb162-16"><a href="#cb162-16" aria-hidden="true" tabindex="-1"></a>      <span class="co"># title = "Anemonefish (Env-Only)",</span></span>
<span id="cb162-17"><a href="#cb162-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="fu">paste</span>(<span class="st">"Mean Suitability Shift ("</span>, <span class="fu">gsub</span>(<span class="st">"shift_"</span>, <span class="st">"SSP"</span>, chosen_scenario_for_figure), <span class="st">")"</span>),</span>
<span id="cb162-18"><a href="#cb162-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Environmental Breadth (Levin's B2)"</span></span>
<span id="cb162-19"><a href="#cb162-19" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb162-20"><a href="#cb162-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">11</span>) <span class="sc">+</span></span>
<span id="cb162-21"><a href="#cb162-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(), <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>())</span>
<span id="cb162-22"><a href="#cb162-22" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb162-23"><a href="#cb162-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Anemonefish plot generated for"</span>, chosen_scenario_for_figure, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb162-24"><a href="#cb162-24" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(fish_cor_test)) <span class="fu">print</span>(fish_cor_test)</span>
<span id="cb162-25"><a href="#cb162-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb162-26"><a href="#cb162-26" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb162-27"><a href="#cb162-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Skipping anemonefish breadth vs. shift plot: data for"</span>, chosen_scenario_for_figure, <span class="st">"or breadth_B2 missing.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb162-28"><a href="#cb162-28" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anemonefish plot generated for shift_585_2100 

    Pearson's product-moment correlation

data:  cor_data_fish$breadth_B2 and cor_data_fish[[chosen_scenario_for_figure]]
t = -0.93881, df = 3, p-value = 0.4171
alternative hypothesis: true correlation is not equal to 0
95 percent confidence interval:
 -0.9566113  0.7000625
sample estimates:
       cor 
-0.4765266 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Combine Plots ---</span></span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(plot_a_host) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(plot_b_fish)) {</span>
<span id="cb164-3"><a href="#cb164-3" aria-hidden="true" tabindex="-1"></a>  combined_breadth_plot <span class="ot">&lt;-</span> cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(plot_a_host, plot_b_fish, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb164-4"><a href="#cb164-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(combined_breadth_plot)</span>
<span id="cb164-5"><a href="#cb164-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Combined breadth vs. suitability shift plot generated.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb164-6"><a href="#cb164-6" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb164-7"><a href="#cb164-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Could not generate combined breadth plot as one or both individual plots are missing.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb164-8"><a href="#cb164-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/breadth-suitability-figure-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Combined breadth vs. suitability shift plot generated.</code></pre>
</div>
</div>
</section>
</section>
<section id="phylogenetic-signal-of-niche-breadth-and-suitability-shifts-anemonefish" class="level2">
<h2 class="anchored" data-anchor-id="phylogenetic-signal-of-niche-breadth-and-suitability-shifts-anemonefish">Phylogenetic Signal of Niche Breadth and Suitability Shifts (Anemonefish)</h2>
<p>This section tests for phylogenetic signal (Blombergâ€™s K) in niche breadth (Levinâ€™s B2) and in the mean suitability shifts for anemonefish species under future scenarios.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(phytools) # For phylosig, multiPhylosignal</span></span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a><span class="co"># library(picante)  # For phylosignal (Blomberg's K)</span></span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a><span class="co"># library(dplyr)</span></span>
<span id="cb166-4"><a href="#cb166-4" aria-hidden="true" tabindex="-1"></a><span class="co"># library(geiger)</span></span>
<span id="cb166-5"><a href="#cb166-5" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb166-6"><a href="#cb166-6" aria-hidden="true" tabindex="-1"></a><span class="co"># if (is.null(pruned_anemonefish_tree)) {</span></span>
<span id="cb166-7"><a href="#cb166-7" aria-hidden="true" tabindex="-1"></a><span class="co">#   cat("Warning: `pruned_anemonefish_tree` not available. Skipping phylogenetic signal of traits for anemonefish.\n")</span></span>
<span id="cb166-8"><a href="#cb166-8" aria-hidden="true" tabindex="-1"></a><span class="co"># } else if (is.null(fish_breadth_shifts_df) || nrow(fish_breadth_shifts_df) == 0) {</span></span>
<span id="cb166-9"><a href="#cb166-9" aria-hidden="true" tabindex="-1"></a><span class="co">#   cat("Warning: `fish_breadth_shifts_df` not available or empty. Skipping phylogenetic signal of traits.\n")</span></span>
<span id="cb166-10"><a href="#cb166-10" aria-hidden="true" tabindex="-1"></a><span class="co"># } else if (!exists("fish_label_map_df") || is.null(fish_label_map_df)) {</span></span>
<span id="cb166-11"><a href="#cb166-11" aria-hidden="true" tabindex="-1"></a><span class="co">#   cat("Warning: `fish_label_map_df` (mapping Original_Genus_Species to short_code) is missing. Skipping phylogenetic signal.\n")</span></span>
<span id="cb166-12"><a href="#cb166-12" aria-hidden="true" tabindex="-1"></a><span class="co"># } else {</span></span>
<span id="cb166-13"><a href="#cb166-13" aria-hidden="true" tabindex="-1"></a><span class="co">#   cat("--- Testing Phylogenetic Signal for Anemonefish Traits ---\n")</span></span>
<span id="cb166-14"><a href="#cb166-14" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb166-15"><a href="#cb166-15" aria-hidden="true" tabindex="-1"></a><span class="co">#   # Ensure tree is strictly bifurcating if needed by some phytools functions</span></span>
<span id="cb166-16"><a href="#cb166-16" aria-hidden="true" tabindex="-1"></a><span class="co">#   # (multi2di might resolve polytomies randomly, consider implications)</span></span>
<span id="cb166-17"><a href="#cb166-17" aria-hidden="true" tabindex="-1"></a><span class="co">#   if(!is.binary(pruned_anemonefish_tree) &amp;&amp; length(pruned_anemonefish_tree$tip.label) &gt; 2) {</span></span>
<span id="cb166-18"><a href="#cb166-18" aria-hidden="true" tabindex="-1"></a><span class="co">#       pruned_anemonefish_tree_binary &lt;- multi2di(pruned_anemonefish_tree, random = TRUE) # Random resolution</span></span>
<span id="cb166-19"><a href="#cb166-19" aria-hidden="true" tabindex="-1"></a><span class="co">#       cat("  Resolved polytomies in anemonefish tree for phylosignal analysis.\n")</span></span>
<span id="cb166-20"><a href="#cb166-20" aria-hidden="true" tabindex="-1"></a><span class="co">#   } else {</span></span>
<span id="cb166-21"><a href="#cb166-21" aria-hidden="true" tabindex="-1"></a><span class="co">#       pruned_anemonefish_tree_binary &lt;- pruned_anemonefish_tree</span></span>
<span id="cb166-22"><a href="#cb166-22" aria-hidden="true" tabindex="-1"></a><span class="co">#   }</span></span>
<span id="cb166-23"><a href="#cb166-23" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb166-24"><a href="#cb166-24" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb166-25"><a href="#cb166-25" aria-hidden="true" tabindex="-1"></a><span class="co">#   # --- 1. Phylogenetic Signal in Niche Breadth (Levin's B2) ---</span></span>
<span id="cb166-26"><a href="#cb166-26" aria-hidden="true" tabindex="-1"></a><span class="co">#   cat("\n  Testing phylogenetic signal for Niche Breadth (Levin's B2)...\n")</span></span>
<span id="cb166-27"><a href="#cb166-27" aria-hidden="true" tabindex="-1"></a><span class="co">#   </span></span>
<span id="cb166-28"><a href="#cb166-28" aria-hidden="true" tabindex="-1"></a><span class="co">#   trait_data_breadth &lt;- fish_breadth_shifts_df %&gt;%</span></span>
<span id="cb166-29"><a href="#cb166-29" aria-hidden="true" tabindex="-1"></a><span class="co">#     inner_join(fish_label_map_df, by = c("species_name_sanitized" = "Original_Genus_Species")) %&gt;%</span></span>
<span id="cb166-30"><a href="#cb166-30" aria-hidden="true" tabindex="-1"></a><span class="co">#     filter(short_code %in% pruned_anemonefish_tree_binary$tip.label) %&gt;% # Keep only species in the tree</span></span>
<span id="cb166-31"><a href="#cb166-31" aria-hidden="true" tabindex="-1"></a><span class="co">#     dplyr::select(short_code, breadth_B2) %&gt;%</span></span>
<span id="cb166-32"><a href="#cb166-32" aria-hidden="true" tabindex="-1"></a><span class="co">#     filter(!is.na(breadth_B2)) %&gt;%</span></span>
<span id="cb166-33"><a href="#cb166-33" aria-hidden="true" tabindex="-1"></a><span class="co">#     distinct(short_code, .keep_all = TRUE) # Ensure unique species</span></span>
<span id="cb166-34"><a href="#cb166-34" aria-hidden="true" tabindex="-1"></a><span class="co">#     </span></span>
<span id="cb166-35"><a href="#cb166-35" aria-hidden="true" tabindex="-1"></a><span class="co">#   if (nrow(trait_data_breadth) &gt; 2) {</span></span>
<span id="cb166-36"><a href="#cb166-36" aria-hidden="true" tabindex="-1"></a><span class="co">#     breadth_values &lt;- trait_data_breadth$breadth_B2</span></span>
<span id="cb166-37"><a href="#cb166-37" aria-hidden="true" tabindex="-1"></a><span class="co">#     names(breadth_values) &lt;- trait_data_breadth$short_code</span></span>
<span id="cb166-38"><a href="#cb166-38" aria-hidden="true" tabindex="-1"></a><span class="co">#     </span></span>
<span id="cb166-39"><a href="#cb166-39" aria-hidden="true" tabindex="-1"></a><span class="co">#     # Match tree and data</span></span>
<span id="cb166-40"><a href="#cb166-40" aria-hidden="true" tabindex="-1"></a><span class="co">#     matched_data_breadth &lt;- treedata(pruned_anemonefish_tree_binary, breadth_values, sort=TRUE, warnings=FALSE)</span></span>
<span id="cb166-41"><a href="#cb166-41" aria-hidden="true" tabindex="-1"></a><span class="co">#     </span></span>
<span id="cb166-42"><a href="#cb166-42" aria-hidden="true" tabindex="-1"></a><span class="co">#     if (length(matched_data_breadth$data) &gt; 2) {</span></span>
<span id="cb166-43"><a href="#cb166-43" aria-hidden="true" tabindex="-1"></a><span class="co">#       phylosig_breadth_B2 &lt;- phylosig(matched_data_breadth$phy, matched_data_breadth$data[,1], method="K", test=TRUE, nsim=999)</span></span>
<span id="cb166-44"><a href="#cb166-44" aria-hidden="true" tabindex="-1"></a><span class="co">#       cat("  Blomberg's K for Niche Breadth (B2):\n")</span></span>
<span id="cb166-45"><a href="#cb166-45" aria-hidden="true" tabindex="-1"></a><span class="co">#       print(phylosig_breadth_B2)</span></span>
<span id="cb166-46"><a href="#cb166-46" aria-hidden="true" tabindex="-1"></a><span class="co">#     } else {</span></span>
<span id="cb166-47"><a href="#cb166-47" aria-hidden="true" tabindex="-1"></a><span class="co">#       cat("  Not enough matching species between tree and breadth data after NA removal.\n")</span></span>
<span id="cb166-48"><a href="#cb166-48" aria-hidden="true" tabindex="-1"></a><span class="co">#     }</span></span>
<span id="cb166-49"><a href="#cb166-49" aria-hidden="true" tabindex="-1"></a><span class="co">#   } else {</span></span>
<span id="cb166-50"><a href="#cb166-50" aria-hidden="true" tabindex="-1"></a><span class="co">#     cat("  Not enough species with valid niche breadth data and in the tree to test for phylogenetic signal.\n")</span></span>
<span id="cb166-51"><a href="#cb166-51" aria-hidden="true" tabindex="-1"></a><span class="co">#   }</span></span>
<span id="cb166-52"><a href="#cb166-52" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb166-53"><a href="#cb166-53" aria-hidden="true" tabindex="-1"></a><span class="co">#   # --- 2. Phylogenetic Signal in Suitability Shifts for each Future Scenario ---</span></span>
<span id="cb166-54"><a href="#cb166-54" aria-hidden="true" tabindex="-1"></a><span class="co">#   for (scenario_name in future_scenarios_to_load) {</span></span>
<span id="cb166-55"><a href="#cb166-55" aria-hidden="true" tabindex="-1"></a><span class="co">#     shift_col_name &lt;- paste0("shift_", gsub("ssp", "", scenario_name))</span></span>
<span id="cb166-56"><a href="#cb166-56" aria-hidden="true" tabindex="-1"></a><span class="co">#     cat("\n  Testing phylogenetic signal for Suitability Shift under scenario:", scenario_name, "(column:", shift_col_name, ")...\n")</span></span>
<span id="cb166-57"><a href="#cb166-57" aria-hidden="true" tabindex="-1"></a><span class="co">#     </span></span>
<span id="cb166-58"><a href="#cb166-58" aria-hidden="true" tabindex="-1"></a><span class="co">#     if (shift_col_name %in% names(fish_breadth_shifts_df)) {</span></span>
<span id="cb166-59"><a href="#cb166-59" aria-hidden="true" tabindex="-1"></a><span class="co">#       trait_data_shift &lt;- fish_breadth_shifts_df %&gt;%</span></span>
<span id="cb166-60"><a href="#cb166-60" aria-hidden="true" tabindex="-1"></a><span class="co">#         inner_join(fish_label_map_df, by = c("species_name_sanitized" = "Original_Genus_Species")) %&gt;%</span></span>
<span id="cb166-61"><a href="#cb166-61" aria-hidden="true" tabindex="-1"></a><span class="co">#         filter(short_code %in% pruned_anemonefish_tree_binary$tip.label) %&gt;%</span></span>
<span id="cb166-62"><a href="#cb166-62" aria-hidden="true" tabindex="-1"></a><span class="co">#         dplyr::select(short_code, suitability_shift = !!sym(shift_col_name)) %&gt;%</span></span>
<span id="cb166-63"><a href="#cb166-63" aria-hidden="true" tabindex="-1"></a><span class="co">#         filter(!is.na(suitability_shift)) %&gt;%</span></span>
<span id="cb166-64"><a href="#cb166-64" aria-hidden="true" tabindex="-1"></a><span class="co">#         distinct(short_code, .keep_all = TRUE)</span></span>
<span id="cb166-65"><a href="#cb166-65" aria-hidden="true" tabindex="-1"></a><span class="co">#         </span></span>
<span id="cb166-66"><a href="#cb166-66" aria-hidden="true" tabindex="-1"></a><span class="co">#       if (nrow(trait_data_shift) &gt; 2) {</span></span>
<span id="cb166-67"><a href="#cb166-67" aria-hidden="true" tabindex="-1"></a><span class="co">#         shift_values &lt;- trait_data_shift$suitability_shift</span></span>
<span id="cb166-68"><a href="#cb166-68" aria-hidden="true" tabindex="-1"></a><span class="co">#         names(shift_values) &lt;- trait_data_shift$short_code</span></span>
<span id="cb166-69"><a href="#cb166-69" aria-hidden="true" tabindex="-1"></a><span class="co">#         </span></span>
<span id="cb166-70"><a href="#cb166-70" aria-hidden="true" tabindex="-1"></a><span class="co">#         matched_data_shift &lt;- treedata(pruned_anemonefish_tree_binary, shift_values, sort=TRUE, warnings=FALSE)</span></span>
<span id="cb166-71"><a href="#cb166-71" aria-hidden="true" tabindex="-1"></a><span class="co">#         </span></span>
<span id="cb166-72"><a href="#cb166-72" aria-hidden="true" tabindex="-1"></a><span class="co">#         if(length(matched_data_shift$data) &gt; 2) {</span></span>
<span id="cb166-73"><a href="#cb166-73" aria-hidden="true" tabindex="-1"></a><span class="co">#           phylosig_shift &lt;- phylosig(matched_data_shift$phy, matched_data_shift$data[,1], method="K", test=TRUE, nsim=999)</span></span>
<span id="cb166-74"><a href="#cb166-74" aria-hidden="true" tabindex="-1"></a><span class="co">#           cat("  Blomberg's K for Suitability Shift (", scenario_name, "):\n")</span></span>
<span id="cb166-75"><a href="#cb166-75" aria-hidden="true" tabindex="-1"></a><span class="co">#           print(phylosig_shift)</span></span>
<span id="cb166-76"><a href="#cb166-76" aria-hidden="true" tabindex="-1"></a><span class="co">#         } else {</span></span>
<span id="cb166-77"><a href="#cb166-77" aria-hidden="true" tabindex="-1"></a><span class="co">#           cat("  Not enough matching species between tree and shift data for", scenario_name, "after NA removal.\n")</span></span>
<span id="cb166-78"><a href="#cb166-78" aria-hidden="true" tabindex="-1"></a><span class="co">#         }</span></span>
<span id="cb166-79"><a href="#cb166-79" aria-hidden="true" tabindex="-1"></a><span class="co">#       } else {</span></span>
<span id="cb166-80"><a href="#cb166-80" aria-hidden="true" tabindex="-1"></a><span class="co">#         cat("  Not enough species with valid suitability shift data for scenario", scenario_name, "and in the tree.\n")</span></span>
<span id="cb166-81"><a href="#cb166-81" aria-hidden="true" tabindex="-1"></a><span class="co">#       }</span></span>
<span id="cb166-82"><a href="#cb166-82" aria-hidden="true" tabindex="-1"></a><span class="co">#     } else {</span></span>
<span id="cb166-83"><a href="#cb166-83" aria-hidden="true" tabindex="-1"></a><span class="co">#       cat("  Shift data column '", shift_col_name, "' not found for scenario", scenario_name, ".\n")</span></span>
<span id="cb166-84"><a href="#cb166-84" aria-hidden="true" tabindex="-1"></a><span class="co">#     }</span></span>
<span id="cb166-85"><a href="#cb166-85" aria-hidden="true" tabindex="-1"></a><span class="co">#   }</span></span>
<span id="cb166-86"><a href="#cb166-86" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="variable-importance-comparison" class="level2">
<h2 class="anchored" data-anchor-id="variable-importance-comparison">Variable Importance Comparison</h2>
<p>This section loads the variable importance (VI) scores from the environmental-only SDMs for host sea anemones and anemonefish. It then compares the importance of each environmental predictor (PCA components if used) between the two guilds.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb167-4"><a href="#cb167-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb167-5"><a href="#cb167-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-6"><a href="#cb167-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine predictor suffix used for the environmental-only models</span></span>
<span id="cb167-7"><a href="#cb167-7" aria-hidden="true" tabindex="-1"></a>env_model_predictor_suffix <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(config<span class="sc">$</span>use_pca_predictors, <span class="st">"_pca"</span>, <span class="st">"_vif"</span>)</span>
<span id="cb167-8"><a href="#cb167-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-9"><a href="#cb167-9" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Load Host Anemone Variable Importance ---</span></span>
<span id="cb167-10"><a href="#cb167-10" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"--- Loading Host Anemone Variable Importance ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Loading Host Anemone Variable Importance ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>host_vi_subdir_name <span class="ot">&lt;-</span> config<span class="sc">$</span>model_output_subdir_map[[env_model_predictor_suffix]] <span class="sc">%||%</span> <span class="fu">ifelse</span>(config<span class="sc">$</span>use_pca_predictors, <span class="st">"anemone_pca"</span>, <span class="st">"anemone_vif"</span>) <span class="co"># Fallback</span></span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a>host_vi_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(config<span class="sc">$</span>target_vi_base, host_vi_subdir_name) <span class="co"># target_vi_base is data/output</span></span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb169-4"><a href="#cb169-4" aria-hidden="true" tabindex="-1"></a>host_vi_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> host_vi_dir, <span class="at">pattern =</span> <span class="st">"^VI_.*</span><span class="sc">\\</span><span class="st">.csv$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb169-5"><a href="#cb169-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(host_vi_files) <span class="sc">==</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> env_model_predictor_suffix <span class="sc">==</span> <span class="st">"_pca"</span>) { <span class="co"># Specific check for typical pca subfolder in target_vi_base</span></span>
<span id="cb169-6"><a href="#cb169-6" aria-hidden="true" tabindex="-1"></a>    host_vi_dir_alt <span class="ot">&lt;-</span> <span class="fu">file.path</span>(host_vi_dir, <span class="st">"anemone_pca"</span>)</span>
<span id="cb169-7"><a href="#cb169-7" aria-hidden="true" tabindex="-1"></a>    host_vi_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> host_vi_dir_alt, <span class="at">pattern =</span> <span class="st">"^VI_.*</span><span class="sc">\\</span><span class="st">.csv$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb169-8"><a href="#cb169-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">length</span>(host_vi_files)<span class="sc">&gt;</span><span class="dv">0</span>) host_vi_dir <span class="ot">&lt;-</span> host_vi_dir_alt <span class="co"># Update if found in alt</span></span>
<span id="cb169-9"><a href="#cb169-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb169-10"><a href="#cb169-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb169-11"><a href="#cb169-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb169-12"><a href="#cb169-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(host_vi_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb169-13"><a href="#cb169-13" aria-hidden="true" tabindex="-1"></a>  host_vi_data <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_df</span>(host_vi_files, </span>
<span id="cb169-14"><a href="#cb169-14" aria-hidden="true" tabindex="-1"></a>                <span class="sc">~</span>readr<span class="sc">::</span><span class="fu">read_csv</span>(.x, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb169-15"><a href="#cb169-15" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">mutate</span>(<span class="at">filename =</span> <span class="fu">basename</span>(.x))) <span class="sc">%&gt;%</span></span>
<span id="cb169-16"><a href="#cb169-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb169-17"><a href="#cb169-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">species_sanitized =</span> <span class="fu">str_remove</span>(<span class="fu">str_remove</span>(filename, <span class="st">"^VI_"</span>), <span class="st">"</span><span class="sc">\\</span><span class="st">.csv$"</span>),</span>
<span id="cb169-18"><a href="#cb169-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">guild =</span> <span class="st">"Host_Anemones"</span></span>
<span id="cb169-19"><a href="#cb169-19" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb169-20"><a href="#cb169-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rename</span>(<span class="at">Permutation_Importance =</span> Importance) <span class="co"># Match ant paper column if different</span></span>
<span id="cb169-21"><a href="#cb169-21" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb169-22"><a href="#cb169-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Loaded VI data for"</span>, <span class="fu">length</span>(<span class="fu">unique</span>(host_vi_data<span class="sc">$</span>species_sanitized)), <span class="st">"host species.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb169-23"><a href="#cb169-23" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb169-24"><a href="#cb169-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: No Variable Importance CSV files found for host anemones in:"</span>, host_vi_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb169-25"><a href="#cb169-25" aria-hidden="true" tabindex="-1"></a>  host_vi_data <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb169-26"><a href="#cb169-26" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded VI data for 9 host species.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Load Anemonefish (Environmental-Only Models) Variable Importance ---</span></span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- Loading Anemonefish (Env-Only) Variable Importance ---</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- Loading Anemonefish (Env-Only) Variable Importance ---</code></pre>
</div>
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>fish_vi_subdir_name <span class="ot">&lt;-</span> config<span class="sc">$</span>model_output_subdir_map[[env_model_predictor_suffix]] <span class="sc">%||%</span> <span class="fu">ifelse</span>(config<span class="sc">$</span>use_pca_predictors, <span class="st">"anemonefish_pca"</span>, <span class="st">"anemonefish_vif"</span>)</span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a>fish_vi_dir <span class="ot">&lt;-</span> <span class="fu">file.path</span>(config<span class="sc">$</span>target_vi_base, fish_vi_subdir_name)</span>
<span id="cb173-3"><a href="#cb173-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-4"><a href="#cb173-4" aria-hidden="true" tabindex="-1"></a>fish_vi_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> fish_vi_dir, <span class="at">pattern =</span> <span class="st">"^VI_.*</span><span class="sc">\\</span><span class="st">.csv$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb173-5"><a href="#cb173-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(fish_vi_files) <span class="sc">==</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> env_model_predictor_suffix <span class="sc">==</span> <span class="st">"_pca"</span>) { <span class="co"># Specific check</span></span>
<span id="cb173-6"><a href="#cb173-6" aria-hidden="true" tabindex="-1"></a>    fish_vi_dir_alt <span class="ot">&lt;-</span> <span class="fu">file.path</span>(fish_vi_dir, <span class="st">"anemonefish_pca"</span>)</span>
<span id="cb173-7"><a href="#cb173-7" aria-hidden="true" tabindex="-1"></a>    fish_vi_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> fish_vi_dir_alt, <span class="at">pattern =</span> <span class="st">"^VI_.*</span><span class="sc">\\</span><span class="st">.csv$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb173-8"><a href="#cb173-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">length</span>(fish_vi_files)<span class="sc">&gt;</span><span class="dv">0</span>) fish_vi_dir <span class="ot">&lt;-</span> fish_vi_dir_alt</span>
<span id="cb173-9"><a href="#cb173-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb173-10"><a href="#cb173-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-11"><a href="#cb173-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(fish_vi_files) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb173-12"><a href="#cb173-12" aria-hidden="true" tabindex="-1"></a>  fish_vi_data <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map_df</span>(fish_vi_files, </span>
<span id="cb173-13"><a href="#cb173-13" aria-hidden="true" tabindex="-1"></a>                <span class="sc">~</span>readr<span class="sc">::</span><span class="fu">read_csv</span>(.x, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb173-14"><a href="#cb173-14" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">mutate</span>(<span class="at">filename =</span> <span class="fu">basename</span>(.x))) <span class="sc">%&gt;%</span></span>
<span id="cb173-15"><a href="#cb173-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(</span>
<span id="cb173-16"><a href="#cb173-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">species_sanitized =</span> <span class="fu">str_remove</span>(<span class="fu">str_remove</span>(filename, <span class="st">"^VI_"</span>), <span class="st">"</span><span class="sc">\\</span><span class="st">.csv$"</span>),</span>
<span id="cb173-17"><a href="#cb173-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">guild =</span> <span class="st">"Anemonefish_EnvOnly"</span></span>
<span id="cb173-18"><a href="#cb173-18" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb173-19"><a href="#cb173-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rename</span>(<span class="at">Permutation_Importance =</span> Importance) <span class="co"># Match ant paper column if different</span></span>
<span id="cb173-20"><a href="#cb173-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Loaded VI data for"</span>, <span class="fu">length</span>(<span class="fu">unique</span>(fish_vi_data<span class="sc">$</span>species_sanitized)), <span class="st">"anemonefish (env-only) species.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb173-21"><a href="#cb173-21" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb173-22"><a href="#cb173-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Warning: No Variable Importance CSV files found for anemonefish (env-only) in:"</span>, fish_vi_dir, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb173-23"><a href="#cb173-23" aria-hidden="true" tabindex="-1"></a>  fish_vi_data <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb173-24"><a href="#cb173-24" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loaded VI data for 22 anemonefish (env-only) species.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Combine and Compare VI ---</span></span>
<span id="cb175-2"><a href="#cb175-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(host_vi_data) <span class="sc">&amp;&amp;</span> <span class="sc">!</span><span class="fu">is.null</span>(fish_vi_data) <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(host_vi_data) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> <span class="fu">nrow</span>(fish_vi_data) <span class="sc">&gt;</span> <span class="dv">0</span>) {</span>
<span id="cb175-3"><a href="#cb175-3" aria-hidden="true" tabindex="-1"></a>  combined_vi_data <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(host_vi_data, fish_vi_data) <span class="sc">%&gt;%</span></span>
<span id="cb175-4"><a href="#cb175-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(Permutation_Importance)) <span class="co"># Ensure importance values are valid</span></span>
<span id="cb175-5"><a href="#cb175-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb175-6"><a href="#cb175-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Ensure 'Variable' column exists (it should from SDMtune::varImp output)</span></span>
<span id="cb175-7"><a href="#cb175-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="sc">!</span><span class="st">"Variable"</span> <span class="sc">%in%</span> <span class="fu">names</span>(combined_vi_data)) {</span>
<span id="cb175-8"><a href="#cb175-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"The 'Variable' column is missing from the loaded VI data. Check the output of calculate_and_save_vi."</span>)</span>
<span id="cb175-9"><a href="#cb175-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb175-10"><a href="#cb175-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb175-11"><a href="#cb175-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Combined VI data for comparison:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb175-12"><a href="#cb175-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">head</span>(combined_vi_data))</span>
<span id="cb175-13"><a href="#cb175-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb175-14"><a href="#cb175-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Boxplot of VI by guild</span></span>
<span id="cb175-15"><a href="#cb175-15" aria-hidden="true" tabindex="-1"></a>  plot_vi_comparison <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(combined_vi_data, <span class="fu">aes</span>(<span class="at">x =</span> Variable, <span class="at">y =</span> Permutation_Importance, <span class="at">fill =</span> guild)) <span class="sc">+</span></span>
<span id="cb175-16"><a href="#cb175-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">preserve =</span> <span class="st">"single"</span>)) <span class="sc">+</span></span>
<span id="cb175-17"><a href="#cb175-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(</span>
<span id="cb175-18"><a href="#cb175-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">title =</span> <span class="st">"Variable Importance Comparison between Guilds"</span>,</span>
<span id="cb175-19"><a href="#cb175-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> <span class="st">"Predictor Variable"</span>,</span>
<span id="cb175-20"><a href="#cb175-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="st">"Permutation Importance (%)"</span></span>
<span id="cb175-21"><a href="#cb175-21" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb175-22"><a href="#cb175-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Host_Anemones"</span> <span class="ot">=</span> <span class="st">"coral"</span>, <span class="st">"Anemonefish_EnvOnly"</span> <span class="ot">=</span> <span class="st">"steelblue"</span>), <span class="at">name =</span> <span class="st">"Guild"</span>) <span class="sc">+</span></span>
<span id="cb175-23"><a href="#cb175-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb175-24"><a href="#cb175-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>), <span class="at">legend.position =</span> <span class="st">"top"</span>)</span>
<span id="cb175-25"><a href="#cb175-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb175-26"><a href="#cb175-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(plot_vi_comparison)</span>
<span id="cb175-27"><a href="#cb175-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb175-28"><a href="#cb175-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Perform t-tests for each predictor variable (e.g., each PC component if PCA was used)</span></span>
<span id="cb175-29"><a href="#cb175-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">--- T-tests for difference in Variable Importance between Guilds ---</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb175-30"><a href="#cb175-30" aria-hidden="true" tabindex="-1"></a>  predictor_vars_to_test <span class="ot">&lt;-</span> <span class="fu">unique</span>(combined_vi_data<span class="sc">$</span>Variable)</span>
<span id="cb175-31"><a href="#cb175-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb175-32"><a href="#cb175-32" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (var_name <span class="cf">in</span> predictor_vars_to_test) {</span>
<span id="cb175-33"><a href="#cb175-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="st">"  Testing variable:"</span>, var_name, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb175-34"><a href="#cb175-34" aria-hidden="true" tabindex="-1"></a>    data_for_ttest <span class="ot">&lt;-</span> combined_vi_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(Variable <span class="sc">==</span> var_name)</span>
<span id="cb175-35"><a href="#cb175-35" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb175-36"><a href="#cb175-36" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">nrow</span>(data_for_ttest <span class="sc">%&gt;%</span> <span class="fu">distinct</span>(guild)) <span class="sc">==</span> <span class="dv">2</span> <span class="sc">&amp;&amp;</span> </span>
<span id="cb175-37"><a href="#cb175-37" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(data_for_ttest <span class="sc">%&gt;%</span> <span class="fu">filter</span>(guild <span class="sc">==</span> <span class="st">"Host_Anemones"</span>) <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()) <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">&amp;&amp;</span></span>
<span id="cb175-38"><a href="#cb175-38" aria-hidden="true" tabindex="-1"></a>        <span class="fu">nrow</span>(data_for_ttest <span class="sc">%&gt;%</span> <span class="fu">filter</span>(guild <span class="sc">==</span> <span class="st">"Anemonefish_EnvOnly"</span>) <span class="sc">%&gt;%</span> <span class="fu">na.omit</span>()) <span class="sc">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb175-39"><a href="#cb175-39" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb175-40"><a href="#cb175-40" aria-hidden="true" tabindex="-1"></a>      ttest_vi_result <span class="ot">&lt;-</span> <span class="fu">tryCatch</span>({</span>
<span id="cb175-41"><a href="#cb175-41" aria-hidden="true" tabindex="-1"></a>        <span class="fu">t.test</span>(Permutation_Importance <span class="sc">~</span> guild, <span class="at">data =</span> data_for_ttest)</span>
<span id="cb175-42"><a href="#cb175-42" aria-hidden="true" tabindex="-1"></a>      }, <span class="at">error =</span> <span class="cf">function</span>(e){</span>
<span id="cb175-43"><a href="#cb175-43" aria-hidden="true" tabindex="-1"></a>        <span class="fu">cat</span>(<span class="st">"    Error in t-test for"</span>, var_name, <span class="st">":"</span>, e<span class="sc">$</span>message, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>); <span class="cn">NULL</span></span>
<span id="cb175-44"><a href="#cb175-44" aria-hidden="true" tabindex="-1"></a>      })</span>
<span id="cb175-45"><a href="#cb175-45" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(ttest_vi_result)) <span class="fu">print</span>(ttest_vi_result)</span>
<span id="cb175-46"><a href="#cb175-46" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb175-47"><a href="#cb175-47" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb175-48"><a href="#cb175-48" aria-hidden="true" tabindex="-1"></a>      <span class="fu">cat</span>(<span class="st">"    Skipping t-test for"</span>, var_name, <span class="st">"- insufficient data or only one guild present.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb175-49"><a href="#cb175-49" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb175-50"><a href="#cb175-50" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb175-51"><a href="#cb175-51" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb175-52"><a href="#cb175-52" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb175-53"><a href="#cb175-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Not enough variable importance data to compare between guilds.</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb175-54"><a href="#cb175-54" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Combined VI data for comparison:
# A tibble: 6 Ã— 5
  Variable Permutation_Importance filename               species_sanitized guild
  &lt;chr&gt;                     &lt;dbl&gt; &lt;chr&gt;                  &lt;chr&gt;             &lt;chr&gt;
1 PC1                        50.1 VI_Entacmaea_quadricoâ€¦ Entacmaea_quadriâ€¦ Hostâ€¦
2 PC2                        40.3 VI_Entacmaea_quadricoâ€¦ Entacmaea_quadriâ€¦ Hostâ€¦
3 PC4                         5.5 VI_Entacmaea_quadricoâ€¦ Entacmaea_quadriâ€¦ Hostâ€¦
4 PC3                         4.2 VI_Entacmaea_quadricoâ€¦ Entacmaea_quadriâ€¦ Hostâ€¦
5 PC2                        35.4 VI_Macrodactyla_doreeâ€¦ Macrodactyla_dorâ€¦ Hostâ€¦
6 PC1                        29.6 VI_Macrodactyla_doreeâ€¦ Macrodactyla_dorâ€¦ Hostâ€¦</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/variable-importance-comparison-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
--- T-tests for difference in Variable Importance between Guilds ---
  Testing variable: PC1 

    Welch Two Sample t-test

data:  Permutation_Importance by guild
t = -0.44029, df = 14.513, p-value = 0.6662
alternative hypothesis: true difference in means between group Anemonefish_EnvOnly and group Host_Anemones is not equal to 0
95 percent confidence interval:
 -21.80342  14.35595
sample estimates:
mean in group Anemonefish_EnvOnly       mean in group Host_Anemones 
                         27.53182                          31.25556 

  Testing variable: PC2 

    Welch Two Sample t-test

data:  Permutation_Importance by guild
t = 0.73366, df = 17.626, p-value = 0.4728
alternative hypothesis: true difference in means between group Anemonefish_EnvOnly and group Host_Anemones is not equal to 0
95 percent confidence interval:
 -10.88049  22.52998
sample estimates:
mean in group Anemonefish_EnvOnly       mean in group Host_Anemones 
                         33.81364                          27.98889 

  Testing variable: PC4 

    Welch Two Sample t-test

data:  Permutation_Importance by guild
t = 0.53242, df = 20.713, p-value = 0.6001
alternative hypothesis: true difference in means between group Anemonefish_EnvOnly and group Host_Anemones is not equal to 0
95 percent confidence interval:
 -9.214122 15.548465
sample estimates:
mean in group Anemonefish_EnvOnly       mean in group Host_Anemones 
                         22.82273                          19.65556 

  Testing variable: PC3 

    Welch Two Sample t-test

data:  Permutation_Importance by guild
t = -1.0046, df = 23.178, p-value = 0.3255
alternative hypothesis: true difference in means between group Anemonefish_EnvOnly and group Host_Anemones is not equal to 0
95 percent confidence interval:
 -16.076537   5.563406
sample estimates:
mean in group Anemonefish_EnvOnly       mean in group Host_Anemones 
                         15.85455                          21.11111 </code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>